{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GWAS summery stat data preprocess\n",
    "1.need to convert Markername to rsID by combining Chr and basepair coordinate in _lifted bim file and and replace Markername by matching\n",
    "2.calculate z-score using P-value and effect size using munge_sumstat in ldsc github repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MarkerName      Allele1 Allele2 Freq1   FreqSE  MinFreq MaxFreq Effect  StdErr  P-value Direction       HetISq  HetChiSq        HetDf   HetPVal freqSpan\n",
      "chr11:88249377  t       c       0.9905  0.0026  0.9882  0.9971  0.0238  0.0594  0.6889  -+?+++--+???-+++?       20      13.753  11      0.247   0.00890000000000002\n",
      "chr1:60320992   a       g       0.9375  0.0049  0.9221  0.9473  0.0185  0.021   0.3801  +-?--+?++??-+---+       20.1    15.022  12      0.2402  0.0252\n",
      "chr8:135908647  a       g       0.2112  0.007   0.1703  0.2316  -0.0054 0.0123  0.6598  --?+++?--??-----+       27.2    16.492  12      0.1697  0.0613\n",
      "chr16:77148858  a       g       0.9963  5e-04   0.9961  0.9984  -0.0228 0.1356  0.8663  -??-+??+-???----?       0       4.331   8       0.8261  0.00229999999999997\n",
      "chr11:97895884  c       g       0.0587  0.0035  0.0361  0.0685  -0.0219 0.0217  0.3148  +-?--+?--??+---++       0.2     12.029  12      0.4433  0.0324\n",
      "chr2:80117945   a       g       0.9979  7e-04   0.9968  0.9988  -0.4366 0.4506  0.3326  ?????+?--???---+?       0       5.236   6       0.5139  0.002\n",
      "chr18:814730    c       g       0.8893  0.0075  0.8639  0.9086  -0.0022 0.0176  0.9025  --?---?+-??+++--+       0       4.652   12      0.9686  0.0447\n",
      "chr19:32978300  a       g       0.0043  9e-04   0.0034  0.0075  0.0398  0.0991  0.6879  +-?--??--???+--+?       0       4.429   9       0.881   0.0041\n",
      "chr2:166277422  a       g       0.0023  0.0022  0.0011  0.0076  -0.1177 0.308   0.7023  ???+---?+????+???       21.5    6.372   5       0.2717  0.0065\n",
      "\u001b[K:\u001b[K20407145 t       c       0.9112  0.0071  0.8824  0.9202  -0.0278 0.0179  \u001b[7m/data/CARD/PD/summary_stats/META5_all.gz\u001b[m\u001b[K\u0007"
     ]
    }
   ],
   "source": [
    "#read meta5 file -- this is our GWAS summery stat data\n",
    "! zless /data/CARD/PD/summary_stats/META5_all.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy meta 5 file into data folder - run only once for download\n",
    "! cp -R /data/CARD/PD/summary_stats/META5_all.gz /data/songy4/TWAS/data/\n",
    "\n",
    "#unzip META5_all.gz - run only once for download\n",
    "! gzip -d META5_all.gz\n",
    "\n",
    "#make sumstat.csv which contains MarkerName, Allele1, Allele2, Effect, StdErr, P-value\n",
    "! awk '{print $1, $2, $3, $8, $9, $10}' data_folder/META5_all > data_folder/sumstat.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n"
     ]
    }
   ],
   "source": [
    "! awk '{print $1, $2, $3, $8, $9, $10}' data_folder/META5_all > data_folder/sumstat.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import library\n",
    "import subprocess\n",
    "import sys\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# set paths\n",
    "basedir = '/data/songy4/proteomics_196'\n",
    "datadir = f'{basedir}/data_folder'\n",
    "twasdir = f'/data/songy4/twas'\n",
    "fusiondir = f'{twasdir}/fusion_twas'\n",
    "geno_path = f'{datadir}/qc_proteomics196_csf_cardio'\n",
    "###Add _case_hg19_lifted or _control_hg19_lifted\n",
    "gene_list_path = f'{datadir}/PDBP_PPMI_proteomics_csf_cardio_gene_list.txt'\n",
    "pheno_path = f'{datadir}/PDBP_PPMI_proteomics_196_csf_cardio_npx_expression'\n",
    "###Add _case.txt or _control.txt\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_csf_cardio_coordinate.txt'\n",
    "covar_path = f'{datadir}/covariates_csf_cardio'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of META5_all: (11477547, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>FreqSE</th>\n",
       "      <th>MinFreq</th>\n",
       "      <th>MaxFreq</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Direction</th>\n",
       "      <th>HetISq</th>\n",
       "      <th>HetChiSq</th>\n",
       "      <th>HetDf</th>\n",
       "      <th>HetPVal</th>\n",
       "      <th>freqSpan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.9905</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.9882</td>\n",
       "      <td>0.9971</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>0.0594</td>\n",
       "      <td>0.6889</td>\n",
       "      <td>-+?+++--+???-+++?</td>\n",
       "      <td>20.0</td>\n",
       "      <td>13.753</td>\n",
       "      <td>11</td>\n",
       "      <td>0.2470</td>\n",
       "      <td>0.0089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9375</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.9221</td>\n",
       "      <td>0.9473</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.3801</td>\n",
       "      <td>+-?--+?++??-+---+</td>\n",
       "      <td>20.1</td>\n",
       "      <td>15.022</td>\n",
       "      <td>12</td>\n",
       "      <td>0.2402</td>\n",
       "      <td>0.0252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>0.0070</td>\n",
       "      <td>0.1703</td>\n",
       "      <td>0.2316</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.6598</td>\n",
       "      <td>--?+++?--??-----+</td>\n",
       "      <td>27.2</td>\n",
       "      <td>16.492</td>\n",
       "      <td>12</td>\n",
       "      <td>0.1697</td>\n",
       "      <td>0.0613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9963</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.9984</td>\n",
       "      <td>-0.0228</td>\n",
       "      <td>0.1356</td>\n",
       "      <td>0.8663</td>\n",
       "      <td>-??-+??+-???----?</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.331</td>\n",
       "      <td>8</td>\n",
       "      <td>0.8261</td>\n",
       "      <td>0.0023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:97895884</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0361</td>\n",
       "      <td>0.0685</td>\n",
       "      <td>-0.0219</td>\n",
       "      <td>0.0217</td>\n",
       "      <td>0.3148</td>\n",
       "      <td>+-?--+?--??+---++</td>\n",
       "      <td>0.2</td>\n",
       "      <td>12.029</td>\n",
       "      <td>12</td>\n",
       "      <td>0.4433</td>\n",
       "      <td>0.0324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MarkerName Allele1 Allele2   Freq1  FreqSE  MinFreq  MaxFreq  Effect  \\\n",
       "0  chr11:88249377       t       c  0.9905  0.0026   0.9882   0.9971  0.0238   \n",
       "1   chr1:60320992       a       g  0.9375  0.0049   0.9221   0.9473  0.0185   \n",
       "2  chr8:135908647       a       g  0.2112  0.0070   0.1703   0.2316 -0.0054   \n",
       "3  chr16:77148858       a       g  0.9963  0.0005   0.9961   0.9984 -0.0228   \n",
       "4  chr11:97895884       c       g  0.0587  0.0035   0.0361   0.0685 -0.0219   \n",
       "\n",
       "   StdErr  P-value          Direction  HetISq  HetChiSq  HetDf  HetPVal  \\\n",
       "0  0.0594   0.6889  -+?+++--+???-+++?    20.0    13.753     11   0.2470   \n",
       "1  0.0210   0.3801  +-?--+?++??-+---+    20.1    15.022     12   0.2402   \n",
       "2  0.0123   0.6598  --?+++?--??-----+    27.2    16.492     12   0.1697   \n",
       "3  0.1356   0.8663  -??-+??+-???----?     0.0     4.331      8   0.8261   \n",
       "4  0.0217   0.3148  +-?--+?--??+---++     0.2    12.029     12   0.4433   \n",
       "\n",
       "   freqSpan  \n",
       "0    0.0089  \n",
       "1    0.0252  \n",
       "2    0.0613  \n",
       "3    0.0023  \n",
       "4    0.0324  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "statsum = pd.read_csv(f\"{twasdir}/data_folder/META5_all\", sep='\\t')\n",
    "print('shape of META5_all:', statsum.shape)\n",
    "statsum.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4973527\n",
      "       MarkerName Allele1 Allele2   Freq1  FreqSE  MinFreq  MaxFreq  Effect  \\\n",
      "0  chr11:88249377       t       c  0.9905  0.0026   0.9882   0.9971  0.0238   \n",
      "1   chr1:60320992       a       g  0.9375  0.0049   0.9221   0.9473  0.0185   \n",
      "3  chr16:77148858       a       g  0.9963  0.0005   0.9961   0.9984 -0.0228   \n",
      "5   chr2:80117945       a       g  0.9979  0.0007   0.9968   0.9988 -0.4366   \n",
      "6    chr18:814730       c       g  0.8893  0.0075   0.8639   0.9086 -0.0022   \n",
      "\n",
      "   StdErr  P-value          Direction  HetISq  HetChiSq  HetDf  HetPVal  \\\n",
      "0  0.0594   0.6889  -+?+++--+???-+++?    20.0    13.753     11   0.2470   \n",
      "1  0.0210   0.3801  +-?--+?++??-+---+    20.1    15.022     12   0.2402   \n",
      "3  0.1356   0.8663  -??-+??+-???----?     0.0     4.331      8   0.8261   \n",
      "5  0.4506   0.3326  ?????+?--???---+?     0.0     5.236      6   0.5139   \n",
      "6  0.0176   0.9025  --?---?+-??+++--+     0.0     4.652     12   0.9686   \n",
      "\n",
      "   freqSpan  \n",
      "0    0.0089  \n",
      "1    0.0252  \n",
      "3    0.0023  \n",
      "5    0.0020  \n",
      "6    0.0447  \n"
     ]
    }
   ],
   "source": [
    "#count if Freq1 is bigger than 0.5 --> this is to check if minor allele is A1\n",
    "!cat {twasdir}/data_folder/META5_all | awk -F\"\\t\" '$4>0.5' | wc -l\n",
    "\n",
    "#check again how many Freq1 \n",
    "statsum_flip = statsum[statsum.Freq1 > 0.5]\n",
    "print(statsum_flip.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3693134, 16)\n",
      "Unique MarkerName in meta data: 7784413\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.380100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.659800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:97895884</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>-0.0219</td>\n",
       "      <td>0.0217</td>\n",
       "      <td>0.314800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18:814730</td>\n",
       "      <td>g</td>\n",
       "      <td>c</td>\n",
       "      <td>0.1107</td>\n",
       "      <td>0.0022</td>\n",
       "      <td>0.0176</td>\n",
       "      <td>0.902500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.120300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784408</th>\n",
       "      <td>chr18:50482478</td>\n",
       "      <td>g</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0203</td>\n",
       "      <td>0.0319</td>\n",
       "      <td>0.525300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784409</th>\n",
       "      <td>chr11:47906648</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0548</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.765700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784410</th>\n",
       "      <td>chr3:47855594</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.3501</td>\n",
       "      <td>-0.0348</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.000944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784411</th>\n",
       "      <td>chr3:194372962</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>0.0500</td>\n",
       "      <td>0.591900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7784412</th>\n",
       "      <td>chr19:34986930</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.2815</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.174900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7784413 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MarkerName Allele1 Allele2   Freq1  Effect  StdErr   P-value\n",
       "0          chr1:60320992       g       a  0.0625 -0.0185  0.0210  0.380100\n",
       "1         chr8:135908647       a       g  0.2112 -0.0054  0.0123  0.659800\n",
       "2         chr11:97895884       c       g  0.0587 -0.0219  0.0217  0.314800\n",
       "3           chr18:814730       g       c  0.1107  0.0022  0.0176  0.902500\n",
       "4        chr10:120407145       c       t  0.0888  0.0278  0.0179  0.120300\n",
       "...                  ...     ...     ...     ...     ...     ...       ...\n",
       "7784408   chr18:50482478       g       c  0.0278  0.0203  0.0319  0.525300\n",
       "7784409   chr11:47906648       c       t  0.0548  0.0062  0.0207  0.765700\n",
       "7784410    chr3:47855594       c       t  0.3501 -0.0348  0.0105  0.000944\n",
       "7784411   chr3:194372962       a       g  0.0166  0.0268  0.0500  0.591900\n",
       "7784412   chr19:34986930       a       c  0.2815  0.0209  0.0154  0.174900\n",
       "\n",
       "[7784413 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#flip Allele1 and Allele2 if Freq1 is bigger than .5 and change Freq1 to negative\n",
    "stat_sum = statsum.copy()\n",
    "cond = (stat_sum['Freq1'] > 0.5)\n",
    "stat_sum.loc[cond, ['Allele1', 'Allele2']] = stat_sum.loc[cond, ['Allele2', 'Allele1']].values\n",
    "#stat_sum_flip.Allele1, stat_sum_flip.Allele2 = np.where(stat_sum_flip.Freq1 > 0.5, [stat_sum_flip.Allele2, stat_sum_flip.Allele1], [stat_sum_flip.Allele1, stat_sum_flip.Allele2])\n",
    "\n",
    "#change Freq1\n",
    "stat_sum.loc[cond, ['Freq1']] = 1-stat_sum['Freq1']\n",
    "\n",
    "#change Effect value\n",
    "stat_sum.loc[cond, 'Effect'] *= -1\n",
    "\n",
    "#check again how many Freq1 are smaller than 0.01\n",
    "statsum_useless = stat_sum[stat_sum.Freq1 < 0.01]\n",
    "print(statsum_useless.shape)\n",
    "\n",
    "#drop two rows Freq1 is smaller than 0.01\n",
    "stat_sum.drop(stat_sum[stat_sum.Freq1 < 0.01].index, inplace=True)\n",
    "stat_sum = stat_sum.reset_index(drop=True)\n",
    "#stat_sum = stat_sum.drop(['index'], axis=1)\n",
    "\n",
    "#drop unwanted columns\n",
    "stat_sum.drop(stat_sum.iloc[:,10:16], axis=1, inplace=True)\n",
    "stat_sum.drop(stat_sum.iloc[:,4:7], axis=1, inplace=True)\n",
    "\n",
    "#check the unique number of MarkerName in meta\n",
    "print(\"Unique MarkerName in meta data:\", stat_sum['MarkerName'].nunique())\n",
    "\n",
    "stat_sum\n",
    "##no need to change  StdErr, P-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save stat_sum \n",
    "stat_sum.to_csv(r'./data_folder/sumstat.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shell_do(command, log=False, return_log=False):\n",
    "    print(f'Executing: {(\" \").join(command.split())}', file=sys.stderr)\n",
    "\n",
    "    res=subprocess.run(command.split(), stdout=subprocess.PIPE)\n",
    "\n",
    "    if log:\n",
    "        print(res.stdout.decode('utf-8'))\n",
    "    if return_log:\n",
    "        return(res.stdout.decode('utf-8'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read Bim file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of bim data: (23060234, 7)\n",
      "number of unique MarkerName in bim data: 23060188\n",
      "number of unique rsid in bim data: 23060234\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs145427775</td>\n",
       "      <td>0</td>\n",
       "      <td>10291</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:10291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs55998931</td>\n",
       "      <td>0</td>\n",
       "      <td>10492</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:10492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199896944</td>\n",
       "      <td>0</td>\n",
       "      <td>13504</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:13504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>0</td>\n",
       "      <td>14933</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:14933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs201855936</td>\n",
       "      <td>0</td>\n",
       "      <td>14948</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:14948</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr         rsid  kb    pos a1 a2  MarkerName\n",
       "0    1  rs145427775   0  10291  T  C  chr1:10291\n",
       "1    1   rs55998931   0  10492  T  C  chr1:10492\n",
       "2    1  rs199896944   0  13504  A  G  chr1:13504\n",
       "3    1  rs199856693   0  14933  A  G  chr1:14933\n",
       "4    1  rs201855936   0  14948  A  G  chr1:14948"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open bim file \n",
    "bim_df = pd.read_csv(f\"{geno_path}_case_hg19_lifted.bim\", sep='\\t', header=None,  names=('chr', 'rsid', 'kb', 'pos', 'a1', 'a2'))\n",
    "\n",
    "#combine chr and pos columns\n",
    "bim_df['MarkerName'] = str('chr') + bim_df.chr.astype(str) + ':' + bim_df.pos.astype(str)\n",
    "\n",
    "#check the shape of meta\n",
    "print(\"shape of bim data:\", bim_df.shape)\n",
    "\n",
    "#check the shape of meta\n",
    "print(\"number of unique MarkerName in bim data:\", bim_df['MarkerName'].nunique())\n",
    "print(\"number of unique rsid in bim data:\", bim_df['rsid'].nunique())\n",
    "bim_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique MarkerName in duplicated bim_df:  46\n",
      "Unique MarkerName in duplicated bim_df:  ['chr1:148609803' 'chr1:148561459' 'chr3:195218551' 'chr3:195216671'\n",
      " 'chr3:195216763' 'chr3:195218266' 'chr3:195216747' 'chr3:195229571'\n",
      " 'chr3:195229612' 'chr9:70948646' 'chr9:70933437' 'chr9:44409306'\n",
      " 'chr9:44093596' 'chr9:44065499' 'chr9:44060266' 'chr10:46914775'\n",
      " 'chr10:46911590' 'chr10:46906700' 'chr11:1093136' 'chr14:19819097'\n",
      " 'chr14:22629698' 'chr14:106478074' 'chr14:106478531' 'chr14:106478574'\n",
      " 'chr14:106478627' 'chr15:21164636' 'chr22:19172321' 'chr22:19198488'\n",
      " 'chr22:19257315' 'chr22:19263547' 'chr22:19299612' 'chr22:19300219'\n",
      " 'chr22:19310505' 'chr22:19390804' 'chr22:19497230' 'chr22:19503090'\n",
      " 'chr22:19555053' 'chr22:19593171' 'chr22:19623380' 'chr22:19643104'\n",
      " 'chr22:19669438' 'chr22:19719099' 'chr22:19738319' 'chr22:19739438'\n",
      " 'chr22:19786130' 'chr22:18872544']\n",
      "Unique allele 1 in duplicated bim_df:  ['A' 'T' 'C' '*' 'G' 'AAGAC']\n",
      "Unique allele 2 in duplicated bim_df:  ['G' 'C' 'T' 'ACC' 'TG' 'A' 'AC' 'ACAG' 'CCTCCTGCCCAG']\n"
     ]
    }
   ],
   "source": [
    "#check duplicated rows\n",
    "bim_df_dup = bim_df[bim_df['MarkerName'].duplicated(keep=False)]\n",
    "#check unique marker name and a1, a2 in duplicated rows\n",
    "print('Number of unique MarkerName in duplicated bim_df: ', bim_df_dup['MarkerName'].nunique())\n",
    "print('Unique MarkerName in duplicated bim_df: ', bim_df_dup['MarkerName'].unique())\n",
    "print('Unique allele 1 in duplicated bim_df: ', bim_df_dup['a1'].unique())\n",
    "print('Unique allele 2 in duplicated bim_df: ', bim_df_dup['a2'].unique())\n",
    "## there are 46 duplicated MarkerName with 2 different rsid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge meta data and bim file to create required format of GWAS sumstat "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.380100</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>0</td>\n",
       "      <td>60320992</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.659800</td>\n",
       "      <td>8</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>0</td>\n",
       "      <td>135908647</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.120300</td>\n",
       "      <td>10</td>\n",
       "      <td>rs4336949</td>\n",
       "      <td>0</td>\n",
       "      <td>120407145</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:29372878</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.822100</td>\n",
       "      <td>11</td>\n",
       "      <td>rs72638174</td>\n",
       "      <td>0</td>\n",
       "      <td>29372878</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3:164053059</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.0607</td>\n",
       "      <td>0.0488</td>\n",
       "      <td>0.214100</td>\n",
       "      <td>3</td>\n",
       "      <td>rs115079612</td>\n",
       "      <td>0</td>\n",
       "      <td>164053059</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478463</th>\n",
       "      <td>chr16:84525085</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>-0.0387</td>\n",
       "      <td>0.0310</td>\n",
       "      <td>0.211700</td>\n",
       "      <td>16</td>\n",
       "      <td>rs2927754</td>\n",
       "      <td>0</td>\n",
       "      <td>84525085</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478464</th>\n",
       "      <td>chr11:47906648</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0548</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.765700</td>\n",
       "      <td>11</td>\n",
       "      <td>rs148423893</td>\n",
       "      <td>0</td>\n",
       "      <td>47906648</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478465</th>\n",
       "      <td>chr3:47855594</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.3501</td>\n",
       "      <td>-0.0348</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.000944</td>\n",
       "      <td>3</td>\n",
       "      <td>rs62263573</td>\n",
       "      <td>0</td>\n",
       "      <td>47855594</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478466</th>\n",
       "      <td>chr3:194372962</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>0.0500</td>\n",
       "      <td>0.591900</td>\n",
       "      <td>3</td>\n",
       "      <td>rs149847890</td>\n",
       "      <td>0</td>\n",
       "      <td>194372962</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478467</th>\n",
       "      <td>chr19:34986930</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.2815</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.174900</td>\n",
       "      <td>19</td>\n",
       "      <td>rs13353264</td>\n",
       "      <td>0</td>\n",
       "      <td>34986930</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6478468 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MarkerName Allele1 Allele2   Freq1  Effect  StdErr   P-value  \\\n",
       "0          chr1:60320992       g       a  0.0625 -0.0185  0.0210  0.380100   \n",
       "1         chr8:135908647       a       g  0.2112 -0.0054  0.0123  0.659800   \n",
       "2        chr10:120407145       c       t  0.0888  0.0278  0.0179  0.120300   \n",
       "3         chr11:29372878       t       c  0.0592  0.0049  0.0218  0.822100   \n",
       "4         chr3:164053059       a       g  0.0146  0.0607  0.0488  0.214100   \n",
       "...                  ...     ...     ...     ...     ...     ...       ...   \n",
       "6478463   chr16:84525085       t       c  0.0268 -0.0387  0.0310  0.211700   \n",
       "6478464   chr11:47906648       c       t  0.0548  0.0062  0.0207  0.765700   \n",
       "6478465    chr3:47855594       c       t  0.3501 -0.0348  0.0105  0.000944   \n",
       "6478466   chr3:194372962       a       g  0.0166  0.0268  0.0500  0.591900   \n",
       "6478467   chr19:34986930       a       c  0.2815  0.0209  0.0154  0.174900   \n",
       "\n",
       "         chr         rsid  kb        pos a1 a2  \n",
       "0          1  rs116406626   0   60320992  G  A  \n",
       "1          8   rs11992603   0  135908647  A  G  \n",
       "2         10    rs4336949   0  120407145  C  T  \n",
       "3         11   rs72638174   0   29372878  T  C  \n",
       "4          3  rs115079612   0  164053059  A  G  \n",
       "...      ...          ...  ..        ... .. ..  \n",
       "6478463   16    rs2927754   0   84525085  T  C  \n",
       "6478464   11  rs148423893   0   47906648  C  T  \n",
       "6478465    3   rs62263573   0   47855594  C  T  \n",
       "6478466    3  rs149847890   0  194372962  A  G  \n",
       "6478467   19   rs13353264   0   34986930  A  C  \n",
       "\n",
       "[6478468 rows x 13 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge stat_sum with bim_df\n",
    "meta_bim = pd.merge(stat_sum, bim_df, on=['MarkerName'], how='inner')\n",
    "meta_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save stat_sum \n",
    "meta_bim.to_csv(r'./data_folder/meta_bim.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of unique rsid in duplicated meta_bim:  0\n",
      "number of unique rsid in meta_bim:  6478468\n",
      "number of unique MarkerName in meta_bim:  6478463\n"
     ]
    }
   ],
   "source": [
    "#check duplicated rsid\n",
    "meta_bim_dup = meta_bim[meta_bim['rsid'].duplicated(keep=False)]\n",
    "print('number of unique rsid in duplicated meta_bim: ', meta_bim_dup['rsid'].nunique())\n",
    "#check number of unique rsid just in case\n",
    "print('number of unique rsid in meta_bim: ', meta_bim.rsid.nunique())\n",
    "#check number of unique MarkerName just in case\n",
    "print('number of unique MarkerName in meta_bim: ', meta_bim.MarkerName.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>rsid</th>\n",
       "      <th>kb</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3011341</th>\n",
       "      <td>chr22:19503090</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>-0.0449</td>\n",
       "      <td>0.0637</td>\n",
       "      <td>0.48050</td>\n",
       "      <td>22</td>\n",
       "      <td>rs370330113</td>\n",
       "      <td>0</td>\n",
       "      <td>19503090</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3011342</th>\n",
       "      <td>chr22:19503090</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>-0.0449</td>\n",
       "      <td>0.0637</td>\n",
       "      <td>0.48050</td>\n",
       "      <td>22</td>\n",
       "      <td>rs138605059</td>\n",
       "      <td>0</td>\n",
       "      <td>19503090</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3214210</th>\n",
       "      <td>chr22:19172321</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0738</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>0.0183</td>\n",
       "      <td>0.52360</td>\n",
       "      <td>22</td>\n",
       "      <td>rs199687134</td>\n",
       "      <td>0</td>\n",
       "      <td>19172321</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3214211</th>\n",
       "      <td>chr22:19172321</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0738</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>0.0183</td>\n",
       "      <td>0.52360</td>\n",
       "      <td>22</td>\n",
       "      <td>rs1780641</td>\n",
       "      <td>0</td>\n",
       "      <td>19172321</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3279286</th>\n",
       "      <td>chr22:19497230</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0475</td>\n",
       "      <td>-0.0466</td>\n",
       "      <td>0.0263</td>\n",
       "      <td>0.07624</td>\n",
       "      <td>22</td>\n",
       "      <td>rs2521701</td>\n",
       "      <td>0</td>\n",
       "      <td>19497230</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3279287</th>\n",
       "      <td>chr22:19497230</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0475</td>\n",
       "      <td>-0.0466</td>\n",
       "      <td>0.0263</td>\n",
       "      <td>0.07624</td>\n",
       "      <td>22</td>\n",
       "      <td>rs117551893</td>\n",
       "      <td>0</td>\n",
       "      <td>19497230</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3950528</th>\n",
       "      <td>chr22:19390804</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0914</td>\n",
       "      <td>-0.0161</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>0.32640</td>\n",
       "      <td>22</td>\n",
       "      <td>rs141585725</td>\n",
       "      <td>0</td>\n",
       "      <td>19390804</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3950529</th>\n",
       "      <td>chr22:19390804</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0914</td>\n",
       "      <td>-0.0161</td>\n",
       "      <td>0.0164</td>\n",
       "      <td>0.32640</td>\n",
       "      <td>22</td>\n",
       "      <td>rs11089274</td>\n",
       "      <td>0</td>\n",
       "      <td>19390804</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5019123</th>\n",
       "      <td>chr22:19643104</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.1720</td>\n",
       "      <td>-0.0006</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>0.96150</td>\n",
       "      <td>22</td>\n",
       "      <td>rs145711952</td>\n",
       "      <td>0</td>\n",
       "      <td>19643104</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5019124</th>\n",
       "      <td>chr22:19643104</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.1720</td>\n",
       "      <td>-0.0006</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>0.96150</td>\n",
       "      <td>22</td>\n",
       "      <td>rs9618649</td>\n",
       "      <td>0</td>\n",
       "      <td>19643104</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             MarkerName Allele1 Allele2   Freq1  Effect  StdErr  P-value  chr  \\\n",
       "3011341  chr22:19503090       c       t  0.0111 -0.0449  0.0637  0.48050   22   \n",
       "3011342  chr22:19503090       c       t  0.0111 -0.0449  0.0637  0.48050   22   \n",
       "3214210  chr22:19172321       t       g  0.0738  0.0117  0.0183  0.52360   22   \n",
       "3214211  chr22:19172321       t       g  0.0738  0.0117  0.0183  0.52360   22   \n",
       "3279286  chr22:19497230       c       t  0.0475 -0.0466  0.0263  0.07624   22   \n",
       "3279287  chr22:19497230       c       t  0.0475 -0.0466  0.0263  0.07624   22   \n",
       "3950528  chr22:19390804       c       t  0.0914 -0.0161  0.0164  0.32640   22   \n",
       "3950529  chr22:19390804       c       t  0.0914 -0.0161  0.0164  0.32640   22   \n",
       "5019123  chr22:19643104       c       t  0.1720 -0.0006  0.0134  0.96150   22   \n",
       "5019124  chr22:19643104       c       t  0.1720 -0.0006  0.0134  0.96150   22   \n",
       "\n",
       "                rsid  kb       pos a1 a2  \n",
       "3011341  rs370330113   0  19503090  C  T  \n",
       "3011342  rs138605059   0  19503090  C  T  \n",
       "3214210  rs199687134   0  19172321  C  T  \n",
       "3214211    rs1780641   0  19172321  T  G  \n",
       "3279286    rs2521701   0  19497230  T  C  \n",
       "3279287  rs117551893   0  19497230  C  T  \n",
       "3950528  rs141585725   0  19390804  T  C  \n",
       "3950529   rs11089274   0  19390804  C  T  \n",
       "5019123  rs145711952   0  19643104  A  G  \n",
       "5019124    rs9618649   0  19643104  C  T  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print gwas summery stat with SNP that has duplicated MarkerName\n",
    "dup_marker = bim_df_dup['MarkerName'].unique().tolist()\n",
    "snp_stat_dup = meta_bim[meta_bim['MarkerName'].isin(dup_marker)]\n",
    "snp_stat_dup\n",
    "#5 duplicated (MarkerName) position with different rsid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numbers of rows in meta data: 6478468\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.3801</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>60320992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.6598</td>\n",
       "      <td>8</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>135908647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.1203</td>\n",
       "      <td>10</td>\n",
       "      <td>rs4336949</td>\n",
       "      <td>120407145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:29372878</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.8221</td>\n",
       "      <td>11</td>\n",
       "      <td>rs72638174</td>\n",
       "      <td>29372878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3:164053059</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.0607</td>\n",
       "      <td>0.0488</td>\n",
       "      <td>0.2141</td>\n",
       "      <td>3</td>\n",
       "      <td>rs115079612</td>\n",
       "      <td>164053059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        MarkerName A1 A2   Freq1  Effect  StdErr  P-value  chr          SNP  \\\n",
       "0    chr1:60320992  g  a  0.0625 -0.0185  0.0210   0.3801    1  rs116406626   \n",
       "1   chr8:135908647  a  g  0.2112 -0.0054  0.0123   0.6598    8   rs11992603   \n",
       "2  chr10:120407145  c  t  0.0888  0.0278  0.0179   0.1203   10    rs4336949   \n",
       "3   chr11:29372878  t  c  0.0592  0.0049  0.0218   0.8221   11   rs72638174   \n",
       "4   chr3:164053059  a  g  0.0146  0.0607  0.0488   0.2141    3  rs115079612   \n",
       "\n",
       "         pos  \n",
       "0   60320992  \n",
       "1  135908647  \n",
       "2  120407145  \n",
       "3   29372878  \n",
       "4  164053059  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop rows in meta if rsid is NaN \n",
    "#meta_df = snp_stat[snp_stat['SNP'].notna()]\n",
    "\n",
    "sumstat_df = meta_bim.copy()\n",
    "#change Allele1 and Allele2 to A1 and A2\n",
    "sumstat_df = sumstat_df.rename(columns={'Allele1':'A1', 'Allele2':'A2','rsid':'SNP'})\n",
    "sumstat_df = sumstat_df.drop(['a1', 'a2', 'kb'], axis=1)\n",
    "\n",
    "#reset index\n",
    "#meta_df = meta_df.reset_index(drop=True)\n",
    "\n",
    "#check numbers of rows in meta_df\n",
    "print(\"numbers of rows in meta data:\", sumstat_df.shape[0])\n",
    "\n",
    "sumstat_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>pos</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.380100</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>60320992</td>\n",
       "      <td>-0.880952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.659800</td>\n",
       "      <td>8</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>135908647</td>\n",
       "      <td>-0.439024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.120300</td>\n",
       "      <td>10</td>\n",
       "      <td>rs4336949</td>\n",
       "      <td>120407145</td>\n",
       "      <td>1.553073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:29372878</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.822100</td>\n",
       "      <td>11</td>\n",
       "      <td>rs72638174</td>\n",
       "      <td>29372878</td>\n",
       "      <td>0.224771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3:164053059</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.0607</td>\n",
       "      <td>0.0488</td>\n",
       "      <td>0.214100</td>\n",
       "      <td>3</td>\n",
       "      <td>rs115079612</td>\n",
       "      <td>164053059</td>\n",
       "      <td>1.243852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478463</th>\n",
       "      <td>chr16:84525085</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>-0.0387</td>\n",
       "      <td>0.0310</td>\n",
       "      <td>0.211700</td>\n",
       "      <td>16</td>\n",
       "      <td>rs2927754</td>\n",
       "      <td>84525085</td>\n",
       "      <td>-1.248387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478464</th>\n",
       "      <td>chr11:47906648</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0548</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.765700</td>\n",
       "      <td>11</td>\n",
       "      <td>rs148423893</td>\n",
       "      <td>47906648</td>\n",
       "      <td>0.299517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478465</th>\n",
       "      <td>chr3:47855594</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.3501</td>\n",
       "      <td>-0.0348</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.000944</td>\n",
       "      <td>3</td>\n",
       "      <td>rs62263573</td>\n",
       "      <td>47855594</td>\n",
       "      <td>-3.314286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478466</th>\n",
       "      <td>chr3:194372962</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>0.0500</td>\n",
       "      <td>0.591900</td>\n",
       "      <td>3</td>\n",
       "      <td>rs149847890</td>\n",
       "      <td>194372962</td>\n",
       "      <td>0.536000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478467</th>\n",
       "      <td>chr19:34986930</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.2815</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.174900</td>\n",
       "      <td>19</td>\n",
       "      <td>rs13353264</td>\n",
       "      <td>34986930</td>\n",
       "      <td>1.357143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6478468 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MarkerName A1 A2   Freq1  Effect  StdErr   P-value  chr  \\\n",
       "0          chr1:60320992  g  a  0.0625 -0.0185  0.0210  0.380100    1   \n",
       "1         chr8:135908647  a  g  0.2112 -0.0054  0.0123  0.659800    8   \n",
       "2        chr10:120407145  c  t  0.0888  0.0278  0.0179  0.120300   10   \n",
       "3         chr11:29372878  t  c  0.0592  0.0049  0.0218  0.822100   11   \n",
       "4         chr3:164053059  a  g  0.0146  0.0607  0.0488  0.214100    3   \n",
       "...                  ... .. ..     ...     ...     ...       ...  ...   \n",
       "6478463   chr16:84525085  t  c  0.0268 -0.0387  0.0310  0.211700   16   \n",
       "6478464   chr11:47906648  c  t  0.0548  0.0062  0.0207  0.765700   11   \n",
       "6478465    chr3:47855594  c  t  0.3501 -0.0348  0.0105  0.000944    3   \n",
       "6478466   chr3:194372962  a  g  0.0166  0.0268  0.0500  0.591900    3   \n",
       "6478467   chr19:34986930  a  c  0.2815  0.0209  0.0154  0.174900   19   \n",
       "\n",
       "                 SNP        pos         Z  \n",
       "0        rs116406626   60320992 -0.880952  \n",
       "1         rs11992603  135908647 -0.439024  \n",
       "2          rs4336949  120407145  1.553073  \n",
       "3         rs72638174   29372878  0.224771  \n",
       "4        rs115079612  164053059  1.243852  \n",
       "...              ...        ...       ...  \n",
       "6478463    rs2927754   84525085 -1.248387  \n",
       "6478464  rs148423893   47906648  0.299517  \n",
       "6478465   rs62263573   47855594 -3.314286  \n",
       "6478466  rs149847890  194372962  0.536000  \n",
       "6478467   rs13353264   34986930  1.357143  \n",
       "\n",
       "[6478468 rows x 11 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calculate z score with effect size and standard error (z = (score-mean)/se)\n",
    "\n",
    "def z_score(a, b):\n",
    "    if b == 0:\n",
    "        return 0.0\n",
    "    return float(a)/b\n",
    "\n",
    "sumstat_df['Z'] = np.vectorize(z_score)(sumstat_df['Effect'], sumstat_df['StdErr'])\n",
    "\n",
    "sumstat_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of meta data: (6478468, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs116406626</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>-0.880952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs11992603</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>-0.439024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4336949</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>1.553073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs72638174</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.224771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs115079612</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>1.243852</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SNP A1 A2   Freq1         Z\n",
       "0  rs116406626  g  a  0.0625 -0.880952\n",
       "1   rs11992603  a  g  0.2112 -0.439024\n",
       "2    rs4336949  c  t  0.0888  1.553073\n",
       "3   rs72638174  t  c  0.0592  0.224771\n",
       "4  rs115079612  a  g  0.0146  1.243852"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#function to move columns\n",
    "def movecol(df, cols_to_move=[], ref_col='', place='After'):\n",
    "    \n",
    "    cols = df.columns.tolist()\n",
    "    if place == 'After':\n",
    "        seg1 = cols[:list(cols).index(ref_col) + 1]\n",
    "        seg2 = cols_to_move\n",
    "    if place == 'Before':\n",
    "        seg1 = cols[:list(cols).index(ref_col)]\n",
    "        seg2 = cols_to_move + [ref_col]\n",
    "    \n",
    "    seg1 = [i for i in seg1 if i not in seg2]\n",
    "    seg3 = [i for i in cols if i not in seg1 + seg2]\n",
    "    \n",
    "    return(df[seg1 + seg2 + seg3])\n",
    "\n",
    "#rearrange Gene_Symbol, Chr, Start, Stop columns right nextg to index column \n",
    "sumstat = movecol(sumstat_df, cols_to_move=['SNP'], ref_col='MarkerName', place='After')\n",
    "\n",
    "#drop MarkerName, Effect, StdErr, P-value colunms \n",
    "sumstat = sumstat.drop(['MarkerName', 'Effect', 'StdErr', 'P-value', 'chr', 'pos'], axis=1)\n",
    "\n",
    "sumstat = sumstat.reset_index(drop=True)\n",
    "print(\"shape of meta data:\", sumstat.shape)\n",
    "sumstat.head()\n",
    "\n",
    "## this sumstat is the format we will use for our fusion analysis ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>pos</th>\n",
       "      <th>Z</th>\n",
       "      <th>chisq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>0.0625</td>\n",
       "      <td>-0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.380100</td>\n",
       "      <td>1</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>60320992</td>\n",
       "      <td>-0.880952</td>\n",
       "      <td>0.776077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.659800</td>\n",
       "      <td>8</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>135908647</td>\n",
       "      <td>-0.439024</td>\n",
       "      <td>0.192742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:120407145</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0888</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0179</td>\n",
       "      <td>0.120300</td>\n",
       "      <td>10</td>\n",
       "      <td>rs4336949</td>\n",
       "      <td>120407145</td>\n",
       "      <td>1.553073</td>\n",
       "      <td>2.412035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:29372878</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0592</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.822100</td>\n",
       "      <td>11</td>\n",
       "      <td>rs72638174</td>\n",
       "      <td>29372878</td>\n",
       "      <td>0.224771</td>\n",
       "      <td>0.050522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3:164053059</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0146</td>\n",
       "      <td>0.0607</td>\n",
       "      <td>0.0488</td>\n",
       "      <td>0.214100</td>\n",
       "      <td>3</td>\n",
       "      <td>rs115079612</td>\n",
       "      <td>164053059</td>\n",
       "      <td>1.243852</td>\n",
       "      <td>1.547169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478463</th>\n",
       "      <td>chr16:84525085</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>-0.0387</td>\n",
       "      <td>0.0310</td>\n",
       "      <td>0.211700</td>\n",
       "      <td>16</td>\n",
       "      <td>rs2927754</td>\n",
       "      <td>84525085</td>\n",
       "      <td>-1.248387</td>\n",
       "      <td>1.558470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478464</th>\n",
       "      <td>chr11:47906648</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.0548</td>\n",
       "      <td>0.0062</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.765700</td>\n",
       "      <td>11</td>\n",
       "      <td>rs148423893</td>\n",
       "      <td>47906648</td>\n",
       "      <td>0.299517</td>\n",
       "      <td>0.089710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478465</th>\n",
       "      <td>chr3:47855594</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>0.3501</td>\n",
       "      <td>-0.0348</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.000944</td>\n",
       "      <td>3</td>\n",
       "      <td>rs62263573</td>\n",
       "      <td>47855594</td>\n",
       "      <td>-3.314286</td>\n",
       "      <td>10.984490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478466</th>\n",
       "      <td>chr3:194372962</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0166</td>\n",
       "      <td>0.0268</td>\n",
       "      <td>0.0500</td>\n",
       "      <td>0.591900</td>\n",
       "      <td>3</td>\n",
       "      <td>rs149847890</td>\n",
       "      <td>194372962</td>\n",
       "      <td>0.536000</td>\n",
       "      <td>0.287296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6478467</th>\n",
       "      <td>chr19:34986930</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>0.2815</td>\n",
       "      <td>0.0209</td>\n",
       "      <td>0.0154</td>\n",
       "      <td>0.174900</td>\n",
       "      <td>19</td>\n",
       "      <td>rs13353264</td>\n",
       "      <td>34986930</td>\n",
       "      <td>1.357143</td>\n",
       "      <td>1.841837</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6478468 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              MarkerName A1 A2   Freq1  Effect  StdErr   P-value  chr  \\\n",
       "0          chr1:60320992  g  a  0.0625 -0.0185  0.0210  0.380100    1   \n",
       "1         chr8:135908647  a  g  0.2112 -0.0054  0.0123  0.659800    8   \n",
       "2        chr10:120407145  c  t  0.0888  0.0278  0.0179  0.120300   10   \n",
       "3         chr11:29372878  t  c  0.0592  0.0049  0.0218  0.822100   11   \n",
       "4         chr3:164053059  a  g  0.0146  0.0607  0.0488  0.214100    3   \n",
       "...                  ... .. ..     ...     ...     ...       ...  ...   \n",
       "6478463   chr16:84525085  t  c  0.0268 -0.0387  0.0310  0.211700   16   \n",
       "6478464   chr11:47906648  c  t  0.0548  0.0062  0.0207  0.765700   11   \n",
       "6478465    chr3:47855594  c  t  0.3501 -0.0348  0.0105  0.000944    3   \n",
       "6478466   chr3:194372962  a  g  0.0166  0.0268  0.0500  0.591900    3   \n",
       "6478467   chr19:34986930  a  c  0.2815  0.0209  0.0154  0.174900   19   \n",
       "\n",
       "                 SNP        pos         Z      chisq  \n",
       "0        rs116406626   60320992 -0.880952   0.776077  \n",
       "1         rs11992603  135908647 -0.439024   0.192742  \n",
       "2          rs4336949  120407145  1.553073   2.412035  \n",
       "3         rs72638174   29372878  0.224771   0.050522  \n",
       "4        rs115079612  164053059  1.243852   1.547169  \n",
       "...              ...        ...       ...        ...  \n",
       "6478463    rs2927754   84525085 -1.248387   1.558470  \n",
       "6478464  rs148423893   47906648  0.299517   0.089710  \n",
       "6478465   rs62263573   47855594 -3.314286  10.984490  \n",
       "6478466  rs149847890  194372962  0.536000   0.287296  \n",
       "6478467   rs13353264   34986930  1.357143   1.841837  \n",
       "\n",
       "[6478468 rows x 12 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calculate chisq\n",
    "stat_sum_chi = sumstat_df.copy()\n",
    "stat_sum_chi['chisq'] = stat_sum_chi['Z']**2\n",
    "stat_sum_chi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lambda from chisq: 1.171208791208791\n",
      "lambda from p value: 1.0206140350877193\n"
     ]
    }
   ],
   "source": [
    "#calculate lambda (median(z^2) / 0.455) using chisq\n",
    "lambda_stat_sum = ((stat_sum_chi['chisq'].median())/0.455)\n",
    "print(\"lambda from chisq:\",lambda_stat_sum)\n",
    "\n",
    "#calculate lambda (median(z^2) / 0.455) using pvalue\n",
    "lambda_statsum = ((stat_sum_chi['P-value'].median())/0.456)\n",
    "print(\"lambda from p value:\",lambda_statsum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save meta_df as txt -- run only once for download\n",
    "sumstat.to_csv(r'./data_folder/meta.txt', sep='\\t' ,index=False)\n",
    "#gwas sumary stat is ready to use"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Grab significant p value from sumstat \n",
    "## from here, run after 4_postprocess notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of significant GWAS data: (5891, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>chr</th>\n",
       "      <th>SNP</th>\n",
       "      <th>pos</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16:31054607</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0831</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>6.178000e-17</td>\n",
       "      <td>16</td>\n",
       "      <td>rs1108431</td>\n",
       "      <td>31054607</td>\n",
       "      <td>-8.393939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:121534732</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>5.506000e-10</td>\n",
       "      <td>10</td>\n",
       "      <td>rs196202</td>\n",
       "      <td>121534732</td>\n",
       "      <td>6.178947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5:102446462</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>-0.0603</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>4.716000e-09</td>\n",
       "      <td>5</td>\n",
       "      <td>rs257320</td>\n",
       "      <td>102446462</td>\n",
       "      <td>-5.854369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:31023273</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>-0.0852</td>\n",
       "      <td>0.0100</td>\n",
       "      <td>1.338000e-17</td>\n",
       "      <td>16</td>\n",
       "      <td>rs12447930</td>\n",
       "      <td>31023273</td>\n",
       "      <td>-8.520000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4:951040</td>\n",
       "      <td>g</td>\n",
       "      <td>t</td>\n",
       "      <td>-0.1074</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>1.054000e-28</td>\n",
       "      <td>4</td>\n",
       "      <td>rs3733344</td>\n",
       "      <td>951040</td>\n",
       "      <td>-11.072165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        MarkerName A1 A2  Effect  StdErr       P-value  chr         SNP  \\\n",
       "0   chr16:31054607  t  c -0.0831  0.0099  6.178000e-17   16   rs1108431   \n",
       "1  chr10:121534732  t  c  0.0587  0.0095  5.506000e-10   10    rs196202   \n",
       "2   chr5:102446462  a  c -0.0603  0.0103  4.716000e-09    5    rs257320   \n",
       "3   chr16:31023273  a  g -0.0852  0.0100  1.338000e-17   16  rs12447930   \n",
       "4      chr4:951040  g  t -0.1074  0.0097  1.054000e-28    4   rs3733344   \n",
       "\n",
       "         pos          Z  \n",
       "0   31054607  -8.393939  \n",
       "1  121534732   6.178947  \n",
       "2  102446462  -5.854369  \n",
       "3   31023273  -8.520000  \n",
       "4     951040 -11.072165  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##grab significant GWAS hits (p <= 5e-8)\n",
    "sig_meta = sumstat_df[sumstat_df['P-value']< 5e-8]\n",
    "#reset index\n",
    "sig_meta = sig_meta.reset_index(drop=True)\n",
    "print(\"shape of significant GWAS data:\", sig_meta.shape)\n",
    "\n",
    "sig_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [MarkerName, A1, A2, Effect, StdErr, P-value, SNP, Z]\n",
       "Index: []"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check if double marker are in sig_meta\n",
    "sig_meta_dup = sig_meta[sig_meta['MarkerName'].isin(double_marker)]\n",
    "sig_meta_dup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save sig_meta -- run only once\n",
    "sig_meta.to_csv(r'./data_folder/gwas_pos.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1682671</td>\n",
       "      <td>1711896</td>\n",
       "      <td>NADK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>11917521</td>\n",
       "      <td>11918988</td>\n",
       "      <td>NTproBNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>16268364</td>\n",
       "      <td>16302627</td>\n",
       "      <td>ZBTB17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>20301925</td>\n",
       "      <td>20306932</td>\n",
       "      <td>PLA2G2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>22148738</td>\n",
       "      <td>22263790</td>\n",
       "      <td>HSPG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>21</td>\n",
       "      <td>33031935</td>\n",
       "      <td>33041244</td>\n",
       "      <td>SOD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>21</td>\n",
       "      <td>43731777</td>\n",
       "      <td>43735761</td>\n",
       "      <td>TFF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>21</td>\n",
       "      <td>45192393</td>\n",
       "      <td>45196326</td>\n",
       "      <td>CSTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>21</td>\n",
       "      <td>46305868</td>\n",
       "      <td>46351904</td>\n",
       "      <td>ITGB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>21</td>\n",
       "      <td>46825052</td>\n",
       "      <td>46933634</td>\n",
       "      <td>COL18A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>343 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     X.Chr     start       end        ID\n",
       "0        1   1682671   1711896      NADK\n",
       "1        1  11917521  11918988  NTproBNP\n",
       "2        1  16268364  16302627    ZBTB17\n",
       "3        1  20301925  20306932   PLA2G2A\n",
       "4        1  22148738  22263790     HSPG2\n",
       "..     ...       ...       ...       ...\n",
       "338     21  33031935  33041244      SOD1\n",
       "339     21  43731777  43735761      TFF3\n",
       "340     21  45192393  45196326      CSTB\n",
       "341     21  46305868  46351904     ITGB2\n",
       "342     21  46825052  46933634   COL18A1\n",
       "\n",
       "[343 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#bring in coordinate data\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "coords"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the 90 parkinson's GWAS hits from Mike's paper (some are in the supplement)\n",
    "##https://www.thelancet.com/journals/laneur/article/PIIS1474-4422(19)30320-5/fulltext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs114138760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs35749011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76763715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs6658353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs11578699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>rs12456492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>rs8087969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>rs55818311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>rs77351827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>rs2248244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rsid\n",
       "0   rs114138760\n",
       "1    rs35749011\n",
       "2    rs76763715\n",
       "3     rs6658353\n",
       "4    rs11578699\n",
       "..          ...\n",
       "85   rs12456492\n",
       "86    rs8087969\n",
       "87   rs55818311\n",
       "88   rs77351827\n",
       "89    rs2248244\n",
       "\n",
       "[90 rows x 1 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read 90 parkinson's GWAS hits\n",
    "pd_hits = pd.read_csv(r\"./data_folder/PD_GRS_dbSNPs.tab\", sep='\\t', lineterminator='\\r', header=None,  names=('rsid', 'author', 'condition', 'name') )\n",
    "pd_hits.drop(pd_hits.iloc[:,1:4], axis=1, inplace=True)\n",
    "pd_hits['rsid'] = pd_hits['rsid'].str.replace('\\n', '')\n",
    "pd_hits['rsid'].replace('', np.nan, inplace=True)\n",
    "pd_hits.dropna(subset=['rsid'], inplace=True)\n",
    "pd_hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   chr        rsid  kb        pos a1 a2      MarkerName\n",
      "0    1  rs35749011   0  155135036  A  G  chr1:155135036\n",
      "1    1  rs76763715   0  155205634  C  T  chr1:155205634\n",
      "2    1  rs11578699   0  171719769  T  C  chr1:171719769\n",
      "3    1    rs823118   0  205723572  C  T  chr1:205723572\n",
      "4    1  rs11557080   0  205737739  A  G  chr1:205737739\n",
      "number of missing rsid:  14\n",
      "missing rsid:  ['rs114138760', 'rs6658353', 'rs76116224', 'rs2042477', 'rs11683001', 'rs11707416', 'rs34025766', 'rs6825004', 'rs26431', 'rs76949143', 'rs6476434', 'rs6497339', 'rs2904880', 'rs666463']\n"
     ]
    }
   ],
   "source": [
    "#merge rsid with pos from meta data\n",
    "pd_hits_pos_1 = pd.merge(bim_df, pd_hits, on=['rsid'], how='inner')\n",
    "print(pd_hits_pos_1.head())\n",
    "\n",
    "#get the rsid without pos\n",
    "mis_rsid = pd_hits[~pd_hits['rsid'].isin(pd_hits_pos_1['rsid'])]\n",
    "mis_rsid = mis_rsid.reset_index(drop=True)\n",
    "print(\"number of missing rsid: \", mis_rsid.rsid.nunique())\n",
    "print(\"missing rsid: \", mis_rsid.rsid.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs114138760</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs35749011</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155135036.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:155135036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>1.0</td>\n",
       "      <td>155205634.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:155205634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs6658353</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs11578699</td>\n",
       "      <td>1.0</td>\n",
       "      <td>171719769.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:171719769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>rs12456492</td>\n",
       "      <td>18.0</td>\n",
       "      <td>40673380.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr18:40673380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>rs8087969</td>\n",
       "      <td>18.0</td>\n",
       "      <td>48683589.0</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr18:48683589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>rs55818311</td>\n",
       "      <td>19.0</td>\n",
       "      <td>2341047.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr19:2341047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>rs77351827</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6006041.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr20:6006041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>rs2248244</td>\n",
       "      <td>21.0</td>\n",
       "      <td>38852361.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr21:38852361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rsid   chr          pos   a1   a2      MarkerName\n",
       "0   rs114138760   NaN          NaN  NaN  NaN             NaN\n",
       "1    rs35749011   1.0  155135036.0    A    G  chr1:155135036\n",
       "2    rs76763715   1.0  155205634.0    C    T  chr1:155205634\n",
       "3     rs6658353   NaN          NaN  NaN  NaN             NaN\n",
       "4    rs11578699   1.0  171719769.0    T    C  chr1:171719769\n",
       "..          ...   ...          ...  ...  ...             ...\n",
       "85   rs12456492  18.0   40673380.0    G    A  chr18:40673380\n",
       "86    rs8087969  18.0   48683589.0    G    T  chr18:48683589\n",
       "87   rs55818311  19.0    2341047.0    C    T   chr19:2341047\n",
       "88   rs77351827  20.0    6006041.0    T    C   chr20:6006041\n",
       "89    rs2248244  21.0   38852361.0    A    G  chr21:38852361\n",
       "\n",
       "[90 rows x 6 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge rsid with pos from meta data\n",
    "\n",
    "pd_hits_pos_2 = pd_hits.merge(bim_df, on=['rsid'], how='left') \n",
    "pd_hits_pos_2 = pd_hits_pos_2.drop(['kb'], axis=1)\n",
    "pd_hits_pos_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs114138760</td>\n",
       "      <td>1</td>\n",
       "      <td>154898185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:154898185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs35749011</td>\n",
       "      <td>1</td>\n",
       "      <td>155135036</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:155135036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>1</td>\n",
       "      <td>155205634</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:155205634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs6658353</td>\n",
       "      <td>1</td>\n",
       "      <td>161469054</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:161469054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs11578699</td>\n",
       "      <td>1</td>\n",
       "      <td>171719769</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:171719769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>rs12456492</td>\n",
       "      <td>18</td>\n",
       "      <td>40673380</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr18:40673380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>rs8087969</td>\n",
       "      <td>18</td>\n",
       "      <td>48683589</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr18:48683589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>rs55818311</td>\n",
       "      <td>19</td>\n",
       "      <td>2341047</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr19:2341047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>rs77351827</td>\n",
       "      <td>20</td>\n",
       "      <td>6006041</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr20:6006041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>rs2248244</td>\n",
       "      <td>21</td>\n",
       "      <td>38852361</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr21:38852361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rsid chr        pos a1 a2      MarkerName\n",
       "0   rs114138760   1  154898185  C  G  chr1:154898185\n",
       "1    rs35749011   1  155135036  A  G  chr1:155135036\n",
       "2    rs76763715   1  155205634  C  T  chr1:155205634\n",
       "3     rs6658353   1  161469054  C  G  chr1:161469054\n",
       "4    rs11578699   1  171719769  T  C  chr1:171719769\n",
       "..          ...  ..        ... .. ..             ...\n",
       "85   rs12456492  18   40673380  G  A  chr18:40673380\n",
       "86    rs8087969  18   48683589  G  T  chr18:48683589\n",
       "87   rs55818311  19    2341047  C  T   chr19:2341047\n",
       "88   rs77351827  20    6006041  T  C   chr20:6006041\n",
       "89    rs2248244  21   38852361  A  G  chr21:38852361\n",
       "\n",
       "[90 rows x 6 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##Fill in missing information manually by searching dbSNP NCBI\n",
    "#input data for rs114138760\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'chr'] = '1'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'pos'] = '154898185'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'a1'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'a2'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs114138760', 'MarkerName'] = 'chr1:154898185'\n",
    "\n",
    "#input data for rs6658353\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'chr'] = '1'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'pos'] = '161469054'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'a1'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'a2'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6658353', 'MarkerName'] = 'chr1:161469054'\n",
    "\n",
    "#input data for rs76116224\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'chr'] = '2'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'pos'] = '18147848'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'a2'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76116224', 'MarkerName'] = 'chr2:18147848'\n",
    "\n",
    "#input data for rs2042477\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'chr'] = '2'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'pos'] = '96000943'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'a2'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2042477', 'MarkerName'] = 'chr2:96000943'\n",
    "\n",
    "#input data for rs11683001\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'chr'] = '2'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'pos'] = '102396963'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'a1'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'a2'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11683001', 'MarkerName'] = 'chr2:102396963'\n",
    "\n",
    "#input data for rs11707416\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'chr'] = '3'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'pos'] = '151108965'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'a1'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'a2'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs11707416', 'MarkerName'] = 'chr3:151108965'\n",
    "\n",
    "#input data for rs34025766\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'chr'] = '4'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'pos'] = '17968811'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'a1'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'a2'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs34025766', 'MarkerName'] = 'chr4:17968811'\n",
    "\n",
    "#input data for rs6825004\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'chr'] = '4'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'pos'] = '77110365'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'a1'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'a2'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6825004', 'MarkerName'] = 'chr4:77110365'\n",
    "\n",
    "#input data for rs26431\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'chr'] = '5'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'pos'] = '102365794'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'a1'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'a2'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs26431', 'MarkerName'] = 'chr5:102365794'\n",
    "\n",
    "#input data for rs76949143\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'chr'] = '7'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'pos'] = '66009851'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'a1'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'a2'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs76949143', 'MarkerName'] = 'chr7:66009851'\n",
    "\n",
    "#input data for rs6476434\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'chr'] = '9'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'pos'] = '34046391'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'a2'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6476434', 'MarkerName'] = 'chr9:34046391'\n",
    "\n",
    "#input data for rs6497339\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'chr'] = '16'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'pos'] = '19277493'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'a2'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs6497339', 'MarkerName'] = 'chr16:19277493'\n",
    "\n",
    "#input data for rs2904880\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'chr'] = '16'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'pos'] = '28944396'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'a1'] = 'G'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'a2'] = 'C'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs2904880', 'MarkerName'] = 'chr16:28944396'\n",
    "\n",
    "#input data for rs666463\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'chr'] = '17'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'pos'] = '76425480'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'a1'] = 'T'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'a2'] = 'A'\n",
    "pd_hits_pos_2.loc[pd_hits_pos_2.rsid == 'rs666463', 'MarkerName'] = 'chr17:76425480'\n",
    "\n",
    "#convert column values to int\n",
    "col_case=pd_hits_pos_2.columns[1:3]\n",
    "pd_hits_pos_2.loc[:,col_case]=pd_hits_pos_2.loc[:,col_case].astype(np.int64)\n",
    "\n",
    "pd_hits_pos_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save pd_hits_pos_2 \n",
    "pd_hits_pos_2.to_csv(r'./data_folder/PD_GRS_dbSNPs_position.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>a1</th>\n",
       "      <th>a2</th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>pos_less</th>\n",
       "      <th>pos_more</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs114138760</td>\n",
       "      <td>1</td>\n",
       "      <td>154898185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:154898185</td>\n",
       "      <td>153898185</td>\n",
       "      <td>155898185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs35749011</td>\n",
       "      <td>1</td>\n",
       "      <td>155135036</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:155135036</td>\n",
       "      <td>154135036</td>\n",
       "      <td>156135036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>1</td>\n",
       "      <td>155205634</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:155205634</td>\n",
       "      <td>154205634</td>\n",
       "      <td>156205634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs6658353</td>\n",
       "      <td>1</td>\n",
       "      <td>161469054</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:161469054</td>\n",
       "      <td>160469054</td>\n",
       "      <td>162469054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs11578699</td>\n",
       "      <td>1</td>\n",
       "      <td>171719769</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:171719769</td>\n",
       "      <td>170719769</td>\n",
       "      <td>172719769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>rs12456492</td>\n",
       "      <td>18</td>\n",
       "      <td>40673380</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr18:40673380</td>\n",
       "      <td>39673380</td>\n",
       "      <td>41673380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>rs8087969</td>\n",
       "      <td>18</td>\n",
       "      <td>48683589</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chr18:48683589</td>\n",
       "      <td>47683589</td>\n",
       "      <td>49683589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>rs55818311</td>\n",
       "      <td>19</td>\n",
       "      <td>2341047</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr19:2341047</td>\n",
       "      <td>1341047</td>\n",
       "      <td>3341047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>rs77351827</td>\n",
       "      <td>20</td>\n",
       "      <td>6006041</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr20:6006041</td>\n",
       "      <td>5006041</td>\n",
       "      <td>7006041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>rs2248244</td>\n",
       "      <td>21</td>\n",
       "      <td>38852361</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr21:38852361</td>\n",
       "      <td>37852361</td>\n",
       "      <td>39852361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           rsid chr        pos a1 a2      MarkerName   pos_less   pos_more\n",
       "0   rs114138760   1  154898185  C  G  chr1:154898185  153898185  155898185\n",
       "1    rs35749011   1  155135036  A  G  chr1:155135036  154135036  156135036\n",
       "2    rs76763715   1  155205634  C  T  chr1:155205634  154205634  156205634\n",
       "3     rs6658353   1  161469054  C  G  chr1:161469054  160469054  162469054\n",
       "4    rs11578699   1  171719769  T  C  chr1:171719769  170719769  172719769\n",
       "..          ...  ..        ... .. ..             ...        ...        ...\n",
       "85   rs12456492  18   40673380  G  A  chr18:40673380   39673380   41673380\n",
       "86    rs8087969  18   48683589  G  T  chr18:48683589   47683589   49683589\n",
       "87   rs55818311  19    2341047  C  T   chr19:2341047    1341047    3341047\n",
       "88   rs77351827  20    6006041  T  C   chr20:6006041    5006041    7006041\n",
       "89    rs2248244  21   38852361  A  G  chr21:38852361   37852361   39852361\n",
       "\n",
       "[90 rows x 8 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create pos_less and pos_more which are position plus minus 1 megabase\n",
    "pd_hits_pos_2['pos_less']= pd_hits_pos_2['pos'] - 1000000\n",
    "pd_hits_pos_2['pos_more']= pd_hits_pos_2['pos'] + 1000000\n",
    "pd_hits_pos_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run each sample and combine the result at the end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in result_1:  85\n",
      "Non Null ID in result_1:  ['PAM', 'GPNMB', 'CTSB', 'CRHR1', 'CRHR1']\n",
      "Null ID in result_2:  89\n",
      "Non Null ID in result_2:  ['CRHR1']\n",
      "Null ID in result_3:  53\n",
      "Non Null ID in result_3:  ['IL6R', 'IL6R', 'IL6R', 'FCGR2A', 'PM20D1', 'PM20D1', 'IL1RL1', 'SPON2', 'SPON2', 'QDPR', 'ART3', 'ART3', 'ART3', 'PAM', 'BAG6', 'IL6', 'GUSB', 'CTSB', 'TNFRSF10C', 'CCL27', 'GRK5', 'GRK5', 'PRCP', 'CNTN1', 'CNTN1', 'DIABLO', 'LGALS3', 'ENTPD5', 'CORO1A', 'AOC3', 'SOST', 'SOST', 'CRHR1', 'CRHR1', 'CRHR1', 'CANT1', 'THOP1']\n"
     ]
    }
   ],
   "source": [
    "#csf_cardio\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_csf_cardio_coordinate.txt'\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "print('Null ID in result_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_1: ',result_1[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_2: ',result_2[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_3: ',result_3[['ID']].T.stack().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['IL6R',\n",
       " 'FCGR2A',\n",
       " 'PM20D1',\n",
       " 'IL1RL1',\n",
       " 'SPON2',\n",
       " 'QDPR',\n",
       " 'ART3',\n",
       " 'PAM',\n",
       " 'BAG6',\n",
       " 'IL6',\n",
       " 'GUSB',\n",
       " 'CTSB',\n",
       " 'TNFRSF10C',\n",
       " 'CCL27',\n",
       " 'GRK5',\n",
       " 'PRCP',\n",
       " 'CNTN1',\n",
       " 'DIABLO',\n",
       " 'LGALS3',\n",
       " 'ENTPD5',\n",
       " 'CORO1A',\n",
       " 'AOC3',\n",
       " 'SOST',\n",
       " 'CRHR1',\n",
       " 'CANT1',\n",
       " 'THOP1',\n",
       " 'GPNMB']"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab non duplicated values from 3 results\n",
    "\n",
    "one = result_1[['ID']].T.stack().tolist()\n",
    "two = result_2[['ID']].T.stack().tolist()\n",
    "three = result_3[['ID']].T.stack().tolist()\n",
    "\n",
    "three.extend(x for x in one if x not in three)\n",
    "three.extend(x for x in two if x not in three)\n",
    "\n",
    "csf_cardio_result = []\n",
    "for i in three:\n",
    "       if i not in csf_cardio_result:\n",
    "            csf_cardio_result.append(i)\n",
    "csf_cardio_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in result_1:  89\n",
      "Non Null ID in result_1:  ['TBC1D5']\n",
      "Null ID in result_2:  89\n",
      "Non Null ID in result_2:  ['TBC1D5']\n",
      "Null ID in result_3:  56\n",
      "Non Null ID in result_3:  ['PKLR', 'PKLR', 'PKLR', 'CD84', 'FASLG', 'NFASC', 'NFASC', 'PARP1', 'IL1R2', 'TBC1D5', 'DAG1', 'HCLS1', 'LAMP3', 'LAP3', 'CXCL9', 'CXCL9', 'CXCL9', 'CXCL14', 'HLA-E', 'MPIG6B', 'LAMA4', 'CCN2', 'IL6', 'B4GALT1', 'PRDX3', 'PRDX3', 'PGF', 'PRSS8', 'EIF5A', 'HEXIM1', 'HEXIM1', 'HEXIM1', 'HEXIM1', 'NCLN']\n"
     ]
    }
   ],
   "source": [
    "#csf_inf\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_csf_inf_coordinate.txt'\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "print('Null ID in result_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_1: ',result_1[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_2: ',result_2[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_3: ',result_3[['ID']].T.stack().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PKLR',\n",
       " 'CD84',\n",
       " 'FASLG',\n",
       " 'NFASC',\n",
       " 'PARP1',\n",
       " 'IL1R2',\n",
       " 'TBC1D5',\n",
       " 'DAG1',\n",
       " 'HCLS1',\n",
       " 'LAMP3',\n",
       " 'LAP3',\n",
       " 'CXCL9',\n",
       " 'CXCL14',\n",
       " 'HLA-E',\n",
       " 'MPIG6B',\n",
       " 'LAMA4',\n",
       " 'CCN2',\n",
       " 'IL6',\n",
       " 'B4GALT1',\n",
       " 'PRDX3',\n",
       " 'PGF',\n",
       " 'PRSS8',\n",
       " 'EIF5A',\n",
       " 'HEXIM1',\n",
       " 'NCLN']"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab non duplicated values from 3 results\n",
    "\n",
    "one = result_1[['ID']].T.stack().tolist()\n",
    "two = result_2[['ID']].T.stack().tolist()\n",
    "three = result_3[['ID']].T.stack().tolist()\n",
    "\n",
    "three.extend(x for x in one if x not in three)\n",
    "three.extend(x for x in two if x not in three)\n",
    "\n",
    "csf_inf_result = []\n",
    "for i in three:\n",
    "       if i not in csf_inf_result:\n",
    "            csf_inf_result.append(i)\n",
    "csf_inf_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in result_1:  86\n",
      "Non Null ID in result_1:  ['PMVK', 'BST1', 'SCARB2', 'BAG3']\n",
      "Null ID in result_2:  90\n",
      "Non Null ID in result_2:  []\n",
      "Null ID in result_3:  61\n",
      "Non Null ID in result_3:  ['PMVK', 'PMVK', 'PMVK', 'F11R', 'MYOC', 'IL1R1', 'BST1', 'NAAA', 'NAAA', 'NAAA', 'DDR1', 'TNXB', 'IL6', 'MSR1', 'SLC39A14', 'CCL19', 'BAG3', 'BAG3', 'CLEC14A', 'BMP4', 'SULT1A1', 'ITGAM', 'CLEC10A', 'DUSP3', 'DUSP3', 'FMNL1', 'FMNL1', 'MAPT', 'CHGB']\n"
     ]
    }
   ],
   "source": [
    "#csf_neuro\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_csf_neuro_coordinate.txt'\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "print('Null ID in result_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_1: ',result_1[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_2: ',result_2[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_3: ',result_3[['ID']].T.stack().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PMVK',\n",
       " 'F11R',\n",
       " 'MYOC',\n",
       " 'IL1R1',\n",
       " 'BST1',\n",
       " 'NAAA',\n",
       " 'DDR1',\n",
       " 'TNXB',\n",
       " 'IL6',\n",
       " 'MSR1',\n",
       " 'SLC39A14',\n",
       " 'CCL19',\n",
       " 'BAG3',\n",
       " 'CLEC14A',\n",
       " 'BMP4',\n",
       " 'SULT1A1',\n",
       " 'ITGAM',\n",
       " 'CLEC10A',\n",
       " 'DUSP3',\n",
       " 'FMNL1',\n",
       " 'MAPT',\n",
       " 'CHGB',\n",
       " 'SCARB2']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab non duplicated values from 3 results\n",
    "\n",
    "one = result_1[['ID']].T.stack().tolist()\n",
    "two = result_2[['ID']].T.stack().tolist()\n",
    "three = result_3[['ID']].T.stack().tolist()\n",
    "\n",
    "three.extend(x for x in one if x not in three)\n",
    "three.extend(x for x in two if x not in three)\n",
    "\n",
    "csf_neuro_result = []\n",
    "for i in three:\n",
    "       if i not in csf_neuro_result:\n",
    "            csf_neuro_result.append(i)\n",
    "csf_neuro_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in result_1:  90\n",
      "Non Null ID in result_1:  []\n",
      "Null ID in result_2:  88\n",
      "Non Null ID in result_2:  ['SLAMF6', 'SEPTIN9']\n",
      "Null ID in result_3:  61\n",
      "Non Null ID in result_3:  ['SCAMP3', 'SCAMP3', 'SCAMP3', 'SLAMF6', 'CNTN2', 'CNTN2', 'LEFTY2', 'IDUA', 'IDUA', 'CD38', 'TXNDC15', 'MOG', 'AIF1', 'VNN2', 'IL6', 'VPS37A', 'GFRA2', 'CDNF', 'ADM', 'PSMD9', 'SEZ6L2', 'STX4', 'STAT5B', 'PPY', 'PPY', 'GFAP', 'GFAP', 'CDC27', 'SEPTIN9']\n"
     ]
    }
   ],
   "source": [
    "#csf_onc\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_csf_onc_coordinate.txt'\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "print('Null ID in result_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_1: ',result_1[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_2: ',result_2[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_3: ',result_3[['ID']].T.stack().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SCAMP3',\n",
       " 'SLAMF6',\n",
       " 'CNTN2',\n",
       " 'LEFTY2',\n",
       " 'IDUA',\n",
       " 'CD38',\n",
       " 'TXNDC15',\n",
       " 'MOG',\n",
       " 'AIF1',\n",
       " 'VNN2',\n",
       " 'IL6',\n",
       " 'VPS37A',\n",
       " 'GFRA2',\n",
       " 'CDNF',\n",
       " 'ADM',\n",
       " 'PSMD9',\n",
       " 'SEZ6L2',\n",
       " 'STX4',\n",
       " 'STAT5B',\n",
       " 'PPY',\n",
       " 'GFAP',\n",
       " 'CDC27',\n",
       " 'SEPTIN9']"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab non duplicated values from 3 results\n",
    "\n",
    "one = result_1[['ID']].T.stack().tolist()\n",
    "two = result_2[['ID']].T.stack().tolist()\n",
    "three = result_3[['ID']].T.stack().tolist()\n",
    "\n",
    "three.extend(x for x in one if x not in three)\n",
    "three.extend(x for x in two if x not in three)\n",
    "\n",
    "csf_onc_result = []\n",
    "for i in three:\n",
    "       if i not in csf_onc_result:\n",
    "            csf_onc_result.append(i)\n",
    "csf_onc_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in result_1:  85\n",
      "Non Null ID in result_1:  ['PAM', 'GPNMB', 'CTSB', 'CRHR1', 'CRHR1']\n",
      "Null ID in result_2:  89\n",
      "Non Null ID in result_2:  ['CRHR1']\n",
      "Null ID in result_3:  53\n",
      "Non Null ID in result_3:  ['IL6R', 'IL6R', 'IL6R', 'FCGR2A', 'PM20D1', 'PM20D1', 'IL1RL1', 'SPON2', 'SPON2', 'QDPR', 'ART3', 'ART3', 'ART3', 'PAM', 'BAG6', 'IL6', 'GUSB', 'CTSB', 'DOK2', 'CCL27', 'GRK5', 'GRK5', 'PRCP', 'CNTN1', 'CNTN1', 'DIABLO', 'LGALS3', 'ENTPD5', 'CORO1A', 'AOC3', 'SOST', 'SOST', 'CRHR1', 'CRHR1', 'CRHR1', 'CANT1', 'THOP1']\n"
     ]
    }
   ],
   "source": [
    "#plasma_cardio\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_plasma_cardio_coordinate.txt'\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "print('Null ID in result_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_1: ',result_1[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_2: ',result_2[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_3: ',result_3[['ID']].T.stack().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['IL6R',\n",
       " 'FCGR2A',\n",
       " 'PM20D1',\n",
       " 'IL1RL1',\n",
       " 'SPON2',\n",
       " 'QDPR',\n",
       " 'ART3',\n",
       " 'PAM',\n",
       " 'BAG6',\n",
       " 'IL6',\n",
       " 'GUSB',\n",
       " 'CTSB',\n",
       " 'DOK2',\n",
       " 'CCL27',\n",
       " 'GRK5',\n",
       " 'PRCP',\n",
       " 'CNTN1',\n",
       " 'DIABLO',\n",
       " 'LGALS3',\n",
       " 'ENTPD5',\n",
       " 'CORO1A',\n",
       " 'AOC3',\n",
       " 'SOST',\n",
       " 'CRHR1',\n",
       " 'CANT1',\n",
       " 'THOP1',\n",
       " 'GPNMB']"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab non duplicated values from 3 results\n",
    "\n",
    "one = result_1[['ID']].T.stack().tolist()\n",
    "two = result_2[['ID']].T.stack().tolist()\n",
    "three = result_3[['ID']].T.stack().tolist()\n",
    "\n",
    "three.extend(x for x in one if x not in three)\n",
    "three.extend(x for x in two if x not in three)\n",
    "\n",
    "plasma_cardio_result = []\n",
    "for i in three:\n",
    "       if i not in plasma_cardio_result:\n",
    "            plasma_cardio_result.append(i)\n",
    "plasma_cardio_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in result_1:  89\n",
      "Non Null ID in result_1:  ['TBC1D5']\n",
      "Null ID in result_2:  89\n",
      "Non Null ID in result_2:  ['TBC1D5']\n",
      "Null ID in result_3:  55\n",
      "Non Null ID in result_3:  ['PKLR', 'PKLR', 'PKLR', 'CD84', 'FASLG', 'NFASC', 'NFASC', 'PARP1', 'IL1R2', 'TBC1D5', 'DAG1', 'HCLS1', 'LAMP3', 'LAP3', 'CXCL9', 'CXCL9', 'CXCL9', 'CXCL14', 'HLA-E', 'MPIG6B', 'LAMA4', 'CCN2', 'IL6', 'B4GALT1', 'PRDX3', 'PRDX3', 'PGF', 'LAT', 'PRSS8', 'EIF5A', 'HEXIM1', 'HEXIM1', 'HEXIM1', 'HEXIM1', 'NCLN']\n"
     ]
    }
   ],
   "source": [
    "#plasma_inf\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_plasma_inf_coordinate.txt'\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "print('Null ID in result_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_1: ',result_1[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_2: ',result_2[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_3: ',result_3[['ID']].T.stack().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PKLR',\n",
       " 'CD84',\n",
       " 'FASLG',\n",
       " 'NFASC',\n",
       " 'PARP1',\n",
       " 'IL1R2',\n",
       " 'TBC1D5',\n",
       " 'DAG1',\n",
       " 'HCLS1',\n",
       " 'LAMP3',\n",
       " 'LAP3',\n",
       " 'CXCL9',\n",
       " 'CXCL14',\n",
       " 'HLA-E',\n",
       " 'MPIG6B',\n",
       " 'LAMA4',\n",
       " 'CCN2',\n",
       " 'IL6',\n",
       " 'B4GALT1',\n",
       " 'PRDX3',\n",
       " 'PGF',\n",
       " 'LAT',\n",
       " 'PRSS8',\n",
       " 'EIF5A',\n",
       " 'HEXIM1',\n",
       " 'NCLN']"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab non duplicated values from 3 results\n",
    "\n",
    "one = result_1[['ID']].T.stack().tolist()\n",
    "two = result_2[['ID']].T.stack().tolist()\n",
    "three = result_3[['ID']].T.stack().tolist()\n",
    "\n",
    "three.extend(x for x in one if x not in three)\n",
    "three.extend(x for x in two if x not in three)\n",
    "\n",
    "plasma_inf_result = []\n",
    "for i in three:\n",
    "       if i not in plasma_inf_result:\n",
    "            plasma_inf_result.append(i)\n",
    "plasma_inf_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in result_1:  86\n",
      "Non Null ID in result_1:  ['PMVK', 'BST1', 'SCARB2', 'BAG3']\n",
      "Null ID in result_2:  89\n",
      "Non Null ID in result_2:  ['KRT14']\n",
      "Null ID in result_3:  61\n",
      "Non Null ID in result_3:  ['PMVK', 'PMVK', 'PMVK', 'F11R', 'MYOC', 'IL1R1', 'BST1', 'NAAA', 'NAAA', 'NAAA', 'DDR1', 'TNXB', 'IL6', 'MSR1', 'SLC39A14', 'CCL19', 'BAG3', 'BAG3', 'CLEC14A', 'BMP4', 'SULT1A1', 'ITGAM', 'CLEC10A', 'KRT14', 'DUSP3', 'DUSP3', 'FMNL1', 'FMNL1', 'CHGB']\n"
     ]
    }
   ],
   "source": [
    "#plasma_neuro\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_plasma_neuro_coordinate.txt'\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "print('Null ID in result_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_1: ',result_1[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_2: ',result_2[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_3: ',result_3[['ID']].T.stack().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PMVK',\n",
       " 'F11R',\n",
       " 'MYOC',\n",
       " 'IL1R1',\n",
       " 'BST1',\n",
       " 'NAAA',\n",
       " 'DDR1',\n",
       " 'TNXB',\n",
       " 'IL6',\n",
       " 'MSR1',\n",
       " 'SLC39A14',\n",
       " 'CCL19',\n",
       " 'BAG3',\n",
       " 'CLEC14A',\n",
       " 'BMP4',\n",
       " 'SULT1A1',\n",
       " 'ITGAM',\n",
       " 'CLEC10A',\n",
       " 'KRT14',\n",
       " 'DUSP3',\n",
       " 'FMNL1',\n",
       " 'CHGB',\n",
       " 'SCARB2']"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab non duplicated values from 3 results\n",
    "\n",
    "one = result_1[['ID']].T.stack().tolist()\n",
    "two = result_2[['ID']].T.stack().tolist()\n",
    "three = result_3[['ID']].T.stack().tolist()\n",
    "\n",
    "three.extend(x for x in one if x not in three)\n",
    "three.extend(x for x in two if x not in three)\n",
    "\n",
    "plasma_neuro_result = []\n",
    "for i in three:\n",
    "       if i not in plasma_neuro_result:\n",
    "            plasma_neuro_result.append(i)\n",
    "plasma_neuro_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Null ID in result_1:  90\n",
      "Non Null ID in result_1:  []\n",
      "Null ID in result_2:  88\n",
      "Non Null ID in result_2:  ['SLAMF6', 'SEPTIN9']\n",
      "Null ID in result_3:  57\n",
      "Non Null ID in result_3:  ['SCAMP3', 'SCAMP3', 'SCAMP3', 'SLAMF6', 'CNTN2', 'CNTN2', 'LEFTY2', 'IDUA', 'IDUA', 'CD38', 'USO1', 'USO1', 'USO1', 'TXNDC15', 'MOG', 'AIF1', 'VNN2', 'IL6', 'VPS37A', 'GFRA2', 'NUDT2', 'CDNF', 'ADM', 'PSMD9', 'SEZ6L2', 'STX4', 'STAT5B', 'PPY', 'PPY', 'GFAP', 'GFAP', 'CDC27', 'SEPTIN9']\n"
     ]
    }
   ],
   "source": [
    "#plasma_onc\n",
    "coord_path = f'{datadir}/PDBP_PPMI_proteomics_plasma_onc_coordinate.txt'\n",
    "coords = pd.read_csv(coord_path, sep='\\t')\n",
    "\n",
    "#grab gene ID from coords file if position is within start and end in coords ---> coords['start'] <= row['pos'] <= coords['end']\n",
    "\n",
    "from numpy import nan\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & ((coords['start'] <= row['pos']) & (coords['end'] >= row['pos']))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_1 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['end']) >= row['pos_more']) | ((coords['start'] <= row['pos_less']) & (coords['end'] >= row['pos_less'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_2 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "def match_labels(row):\n",
    "    curr_df = coords[(coords['X.Chr'] == row['chr']) & (((coords['start'] <= row['pos_more']) & (coords['start'] >= row['pos_less'])) | ((coords['end'] >= row['pos_less']) & (coords['end'] <= row['pos_more'])))]\n",
    "    try:\n",
    "        row['ID'] = curr_df['ID'].iloc[0]\n",
    "    except:\n",
    "        row['ID'] = np.NaN\n",
    "\n",
    "    return row\n",
    "\n",
    "result_3 = pd_hits_pos_2.apply(lambda x:match_labels(x),axis=1)\n",
    "\n",
    "print('Null ID in result_1: ',result_1.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_1: ',result_1[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_2: ',result_2.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_2: ',result_2[['ID']].T.stack().tolist())\n",
    "\n",
    "print('Null ID in result_3: ',result_3.ID.isnull().sum())\n",
    "\n",
    "#get non null values\n",
    "print('Non Null ID in result_3: ',result_3[['ID']].T.stack().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SCAMP3',\n",
       " 'SLAMF6',\n",
       " 'CNTN2',\n",
       " 'LEFTY2',\n",
       " 'IDUA',\n",
       " 'CD38',\n",
       " 'USO1',\n",
       " 'TXNDC15',\n",
       " 'MOG',\n",
       " 'AIF1',\n",
       " 'VNN2',\n",
       " 'IL6',\n",
       " 'VPS37A',\n",
       " 'GFRA2',\n",
       " 'NUDT2',\n",
       " 'CDNF',\n",
       " 'ADM',\n",
       " 'PSMD9',\n",
       " 'SEZ6L2',\n",
       " 'STX4',\n",
       " 'STAT5B',\n",
       " 'PPY',\n",
       " 'GFAP',\n",
       " 'CDC27',\n",
       " 'SEPTIN9']"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab non duplicated values from 3 results\n",
    "\n",
    "one = result_1[['ID']].T.stack().tolist()\n",
    "two = result_2[['ID']].T.stack().tolist()\n",
    "three = result_3[['ID']].T.stack().tolist()\n",
    "\n",
    "three.extend(x for x in one if x not in three)\n",
    "three.extend(x for x in two if x not in three)\n",
    "\n",
    "plasma_onc_result = []\n",
    "for i in three:\n",
    "       if i not in plasma_onc_result:\n",
    "            plasma_onc_result.append(i)\n",
    "plasma_onc_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>csf_cardio</th>\n",
       "      <th>csf_inf</th>\n",
       "      <th>csf_neuro</th>\n",
       "      <th>csf_onc</th>\n",
       "      <th>plasma_cardio</th>\n",
       "      <th>plasma_inf</th>\n",
       "      <th>plasma_neuro</th>\n",
       "      <th>plasma_onc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL6R</td>\n",
       "      <td>PKLR</td>\n",
       "      <td>PMVK</td>\n",
       "      <td>SCAMP3</td>\n",
       "      <td>IL6R</td>\n",
       "      <td>PKLR</td>\n",
       "      <td>PMVK</td>\n",
       "      <td>SCAMP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FCGR2A</td>\n",
       "      <td>CD84</td>\n",
       "      <td>F11R</td>\n",
       "      <td>SLAMF6</td>\n",
       "      <td>FCGR2A</td>\n",
       "      <td>CD84</td>\n",
       "      <td>F11R</td>\n",
       "      <td>SLAMF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PM20D1</td>\n",
       "      <td>FASLG</td>\n",
       "      <td>MYOC</td>\n",
       "      <td>CNTN2</td>\n",
       "      <td>PM20D1</td>\n",
       "      <td>FASLG</td>\n",
       "      <td>MYOC</td>\n",
       "      <td>CNTN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IL1RL1</td>\n",
       "      <td>NFASC</td>\n",
       "      <td>IL1R1</td>\n",
       "      <td>LEFTY2</td>\n",
       "      <td>IL1RL1</td>\n",
       "      <td>NFASC</td>\n",
       "      <td>IL1R1</td>\n",
       "      <td>LEFTY2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPON2</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>BST1</td>\n",
       "      <td>IDUA</td>\n",
       "      <td>SPON2</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>BST1</td>\n",
       "      <td>IDUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>QDPR</td>\n",
       "      <td>IL1R2</td>\n",
       "      <td>NAAA</td>\n",
       "      <td>CD38</td>\n",
       "      <td>QDPR</td>\n",
       "      <td>IL1R2</td>\n",
       "      <td>NAAA</td>\n",
       "      <td>CD38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ART3</td>\n",
       "      <td>TBC1D5</td>\n",
       "      <td>DDR1</td>\n",
       "      <td>TXNDC15</td>\n",
       "      <td>ART3</td>\n",
       "      <td>TBC1D5</td>\n",
       "      <td>DDR1</td>\n",
       "      <td>USO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PAM</td>\n",
       "      <td>DAG1</td>\n",
       "      <td>TNXB</td>\n",
       "      <td>MOG</td>\n",
       "      <td>PAM</td>\n",
       "      <td>DAG1</td>\n",
       "      <td>TNXB</td>\n",
       "      <td>TXNDC15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BAG6</td>\n",
       "      <td>HCLS1</td>\n",
       "      <td>IL6</td>\n",
       "      <td>AIF1</td>\n",
       "      <td>BAG6</td>\n",
       "      <td>HCLS1</td>\n",
       "      <td>IL6</td>\n",
       "      <td>MOG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>IL6</td>\n",
       "      <td>LAMP3</td>\n",
       "      <td>MSR1</td>\n",
       "      <td>VNN2</td>\n",
       "      <td>IL6</td>\n",
       "      <td>LAMP3</td>\n",
       "      <td>MSR1</td>\n",
       "      <td>AIF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GUSB</td>\n",
       "      <td>LAP3</td>\n",
       "      <td>SLC39A14</td>\n",
       "      <td>IL6</td>\n",
       "      <td>GUSB</td>\n",
       "      <td>LAP3</td>\n",
       "      <td>SLC39A14</td>\n",
       "      <td>VNN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CTSB</td>\n",
       "      <td>CXCL9</td>\n",
       "      <td>CCL19</td>\n",
       "      <td>VPS37A</td>\n",
       "      <td>CTSB</td>\n",
       "      <td>CXCL9</td>\n",
       "      <td>CCL19</td>\n",
       "      <td>IL6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TNFRSF10C</td>\n",
       "      <td>CXCL14</td>\n",
       "      <td>BAG3</td>\n",
       "      <td>GFRA2</td>\n",
       "      <td>DOK2</td>\n",
       "      <td>CXCL14</td>\n",
       "      <td>BAG3</td>\n",
       "      <td>VPS37A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CCL27</td>\n",
       "      <td>HLA-E</td>\n",
       "      <td>CLEC14A</td>\n",
       "      <td>CDNF</td>\n",
       "      <td>CCL27</td>\n",
       "      <td>HLA-E</td>\n",
       "      <td>CLEC14A</td>\n",
       "      <td>GFRA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GRK5</td>\n",
       "      <td>MPIG6B</td>\n",
       "      <td>BMP4</td>\n",
       "      <td>ADM</td>\n",
       "      <td>GRK5</td>\n",
       "      <td>MPIG6B</td>\n",
       "      <td>BMP4</td>\n",
       "      <td>NUDT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PRCP</td>\n",
       "      <td>LAMA4</td>\n",
       "      <td>SULT1A1</td>\n",
       "      <td>PSMD9</td>\n",
       "      <td>PRCP</td>\n",
       "      <td>LAMA4</td>\n",
       "      <td>SULT1A1</td>\n",
       "      <td>CDNF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CNTN1</td>\n",
       "      <td>CCN2</td>\n",
       "      <td>ITGAM</td>\n",
       "      <td>SEZ6L2</td>\n",
       "      <td>CNTN1</td>\n",
       "      <td>CCN2</td>\n",
       "      <td>ITGAM</td>\n",
       "      <td>ADM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>DIABLO</td>\n",
       "      <td>IL6</td>\n",
       "      <td>CLEC10A</td>\n",
       "      <td>STX4</td>\n",
       "      <td>DIABLO</td>\n",
       "      <td>IL6</td>\n",
       "      <td>CLEC10A</td>\n",
       "      <td>PSMD9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>LGALS3</td>\n",
       "      <td>B4GALT1</td>\n",
       "      <td>DUSP3</td>\n",
       "      <td>STAT5B</td>\n",
       "      <td>LGALS3</td>\n",
       "      <td>B4GALT1</td>\n",
       "      <td>KRT14</td>\n",
       "      <td>SEZ6L2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENTPD5</td>\n",
       "      <td>PRDX3</td>\n",
       "      <td>FMNL1</td>\n",
       "      <td>PPY</td>\n",
       "      <td>ENTPD5</td>\n",
       "      <td>PRDX3</td>\n",
       "      <td>DUSP3</td>\n",
       "      <td>STX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CORO1A</td>\n",
       "      <td>PGF</td>\n",
       "      <td>MAPT</td>\n",
       "      <td>GFAP</td>\n",
       "      <td>CORO1A</td>\n",
       "      <td>PGF</td>\n",
       "      <td>FMNL1</td>\n",
       "      <td>STAT5B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AOC3</td>\n",
       "      <td>PRSS8</td>\n",
       "      <td>CHGB</td>\n",
       "      <td>CDC27</td>\n",
       "      <td>AOC3</td>\n",
       "      <td>LAT</td>\n",
       "      <td>CHGB</td>\n",
       "      <td>PPY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SOST</td>\n",
       "      <td>EIF5A</td>\n",
       "      <td>SCARB2</td>\n",
       "      <td>SEPTIN9</td>\n",
       "      <td>SOST</td>\n",
       "      <td>PRSS8</td>\n",
       "      <td>SCARB2</td>\n",
       "      <td>GFAP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CRHR1</td>\n",
       "      <td>HEXIM1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRHR1</td>\n",
       "      <td>EIF5A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDC27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CANT1</td>\n",
       "      <td>NCLN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CANT1</td>\n",
       "      <td>HEXIM1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SEPTIN9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>THOP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>THOP1</td>\n",
       "      <td>NCLN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GPNMB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GPNMB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   csf_cardio  csf_inf csf_neuro  csf_onc plasma_cardio plasma_inf  \\\n",
       "0        IL6R     PKLR      PMVK   SCAMP3          IL6R       PKLR   \n",
       "1      FCGR2A     CD84      F11R   SLAMF6        FCGR2A       CD84   \n",
       "2      PM20D1    FASLG      MYOC    CNTN2        PM20D1      FASLG   \n",
       "3      IL1RL1    NFASC     IL1R1   LEFTY2        IL1RL1      NFASC   \n",
       "4       SPON2    PARP1      BST1     IDUA         SPON2      PARP1   \n",
       "5        QDPR    IL1R2      NAAA     CD38          QDPR      IL1R2   \n",
       "6        ART3   TBC1D5      DDR1  TXNDC15          ART3     TBC1D5   \n",
       "7         PAM     DAG1      TNXB      MOG           PAM       DAG1   \n",
       "8        BAG6    HCLS1       IL6     AIF1          BAG6      HCLS1   \n",
       "9         IL6    LAMP3      MSR1     VNN2           IL6      LAMP3   \n",
       "10       GUSB     LAP3  SLC39A14      IL6          GUSB       LAP3   \n",
       "11       CTSB    CXCL9     CCL19   VPS37A          CTSB      CXCL9   \n",
       "12  TNFRSF10C   CXCL14      BAG3    GFRA2          DOK2     CXCL14   \n",
       "13      CCL27    HLA-E   CLEC14A     CDNF         CCL27      HLA-E   \n",
       "14       GRK5   MPIG6B      BMP4      ADM          GRK5     MPIG6B   \n",
       "15       PRCP    LAMA4   SULT1A1    PSMD9          PRCP      LAMA4   \n",
       "16      CNTN1     CCN2     ITGAM   SEZ6L2         CNTN1       CCN2   \n",
       "17     DIABLO      IL6   CLEC10A     STX4        DIABLO        IL6   \n",
       "18     LGALS3  B4GALT1     DUSP3   STAT5B        LGALS3    B4GALT1   \n",
       "19     ENTPD5    PRDX3     FMNL1      PPY        ENTPD5      PRDX3   \n",
       "20     CORO1A      PGF      MAPT     GFAP        CORO1A        PGF   \n",
       "21       AOC3    PRSS8      CHGB    CDC27          AOC3        LAT   \n",
       "22       SOST    EIF5A    SCARB2  SEPTIN9          SOST      PRSS8   \n",
       "23      CRHR1   HEXIM1       NaN      NaN         CRHR1      EIF5A   \n",
       "24      CANT1     NCLN       NaN      NaN         CANT1     HEXIM1   \n",
       "25      THOP1      NaN       NaN      NaN         THOP1       NCLN   \n",
       "26      GPNMB      NaN       NaN      NaN         GPNMB        NaN   \n",
       "\n",
       "   plasma_neuro plasma_onc  \n",
       "0          PMVK     SCAMP3  \n",
       "1          F11R     SLAMF6  \n",
       "2          MYOC      CNTN2  \n",
       "3         IL1R1     LEFTY2  \n",
       "4          BST1       IDUA  \n",
       "5          NAAA       CD38  \n",
       "6          DDR1       USO1  \n",
       "7          TNXB    TXNDC15  \n",
       "8           IL6        MOG  \n",
       "9          MSR1       AIF1  \n",
       "10     SLC39A14       VNN2  \n",
       "11        CCL19        IL6  \n",
       "12         BAG3     VPS37A  \n",
       "13      CLEC14A      GFRA2  \n",
       "14         BMP4      NUDT2  \n",
       "15      SULT1A1       CDNF  \n",
       "16        ITGAM        ADM  \n",
       "17      CLEC10A      PSMD9  \n",
       "18        KRT14     SEZ6L2  \n",
       "19        DUSP3       STX4  \n",
       "20        FMNL1     STAT5B  \n",
       "21         CHGB        PPY  \n",
       "22       SCARB2       GFAP  \n",
       "23          NaN      CDC27  \n",
       "24          NaN    SEPTIN9  \n",
       "25          NaN        NaN  \n",
       "26          NaN        NaN  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read proteomics result \n",
    "\n",
    "#read gwas significant\n",
    "\n",
    "#put the result into a new dataframe \n",
    "csf_cardio = pd.DataFrame({'csf_cardio':csf_cardio_result})\n",
    "csf_inf = pd.DataFrame({'csf_inf':csf_inf_result})\n",
    "csf_neuro = pd.DataFrame({'csf_neuro':csf_neuro_result})\n",
    "csf_onc = pd.DataFrame({'csf_onc':csf_onc_result})\n",
    "plasma_cardio = pd.DataFrame({'plasma_cardio':plasma_cardio_result})\n",
    "plasma_inf = pd.DataFrame({'plasma_inf':plasma_inf_result})\n",
    "plasma_neuro = pd.DataFrame({'plasma_neuro':plasma_neuro_result})\n",
    "plasma_onc = pd.DataFrame({'plasma_onc':plasma_onc_result})\n",
    "\n",
    "proteomics_result = pd.concat([csf_cardio, csf_inf, csf_neuro, csf_onc, plasma_cardio, plasma_inf, plasma_neuro, plasma_onc], axis=1) \n",
    "proteomics_result \n",
    "#these are genes +/- 1MB from the 90 parkinson's GWAS hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteomics_result.to_csv(r'./data_folder/proteomics_pd_hits.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.8",
   "language": "python",
   "name": "py3.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
