{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import sys\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shell_do(command, log=False, return_log=False):\n",
    "    print(f'Executing: {(\" \").join(command.split())}', file=sys.stderr)\n",
    "\n",
    "    res=subprocess.run(command.split(), stdout=subprocess.PIPE)\n",
    "\n",
    "    if log:\n",
    "        print(res.stdout.decode('utf-8'))\n",
    "    if return_log:\n",
    "        return(res.stdout.decode('utf-8'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set paths\n",
    "basedir = '/data/songy4/proteomics_196'\n",
    "datadir = f'{basedir}/data_folder'\n",
    "twasdir = f'/data/songy4/twas'\n",
    "fusiondir = f'{twasdir}/fusion_twas'\n",
    "sumstat_path = f'{datadir}/meta.txt'\n",
    "top_dir = f'{basedir}/output/top_csf_cardio_case'\n",
    "anal_dir = f'{basedir}/output/analysis_csf_cardio_case'\n",
    "fusion_ldref_basename = f'{fusiondir}/LDREF/1000G.EUR'\n",
    "fusion_post_script = f'{fusiondir}/FUSION.post_process.R'\n",
    "\n",
    "!mkdir --parents output/top_csf_cardio_case\n",
    "!mkdir --parents output/analysis_csf_cardio_case\n",
    "\n",
    "#!for i in {1..22}; do mkdir --parents output/analysis_csf_cardio_case/PD.$i; done;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of rows in PD.1.dat: 7\n",
      "number of rows in PD.2.dat: 2\n",
      "number of rows in PD.3.dat: 0\n",
      "number of rows in PD.4.dat: 1\n",
      "number of rows in PD.5.dat: 3\n",
      "number of rows in PD.6.dat: 1\n",
      "number of rows in PD.7.dat: 3\n",
      "number of rows in PD.8.dat: 0\n",
      "number of rows in PD.9.dat: 0\n",
      "number of rows in PD.10.dat: 3\n",
      "number of rows in PD.11.dat: 1\n",
      "number of rows in PD.12.dat: 0\n",
      "number of rows in PD.13.dat: 1\n",
      "number of rows in PD.14.dat: 1\n",
      "number of rows in PD.15.dat: 1\n",
      "number of rows in PD.16.dat: 1\n",
      "number of rows in PD.17.dat: 0\n",
      "number of rows in PD.18.dat: 0\n",
      "number of rows in PD.19.dat: 3\n",
      "number of rows in PD.20.dat: 0\n",
      "number of rows in PD.21.dat: 0\n",
      "number of rows in PD.22.dat: 2\n",
      "number of rows in dat files [7, 2, 0, 1, 3, 1, 3, 0, 0, 3, 1, 0, 1, 1, 1, 1, 0, 0, 3, 0, 0, 2]\n"
     ]
    }
   ],
   "source": [
    "#grab all gene ID from to.analysis.joint_included.dat files\n",
    "import glob\n",
    "\n",
    "def read_id(file):\n",
    "    return pd.read_csv(file, delim_whitespace=1, usecols=[1])\n",
    "\n",
    "files = glob.glob(f'{basedir}/output/pd_csf_cardio_case/PD.*.dat')\n",
    "\n",
    "gene_ls = []\n",
    "for i in range(1, 23):\n",
    "    pd_i = \"/data/songy4/proteomics_196/output/pd_csf_cardio_case/PD.\" + str(i) + \".dat\"\n",
    "    if pd_i in files:\n",
    "        pd_i = pd.read_csv(f\"{basedir}/output/pd_csf_cardio_case/PD.{i}.dat\", sep='\\t', index_col=False)\n",
    "        print('number of rows in PD.' + str(i) + '.dat:', pd_i.shape[0])\n",
    "        gene_ls.append(pd_i.shape[0])\n",
    "    else:\n",
    "        print('number of rows in PD.' + str(i) + '.dat:', '0')\n",
    "        gene_ls.append(0)\n",
    "print('number of rows in dat files', gene_ls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from each .dat file in pd, grab rows where TWAS.P column has smaller value than 0.05/number of rows and save them as .top file in top folder\n",
    "#grab the i number in dat files\n",
    "file_name = [s.replace('/data/songy4/proteomics_196/output/pd_csf_cardio_case/PD.', '') for s in files]\n",
    "file_name = [s.replace('.dat', '') for s in file_name]\n",
    "\n",
    "for i in file_name:\n",
    "    pd_i = pd.read_csv(f\"{basedir}/output/pd_csf_cardio_case/PD.{i}.dat\", sep='\\t', index_col=False)\n",
    "    pd_i['TWAS.P'] = pd.to_numeric(pd_i['TWAS.P'], errors='coerce')\n",
    "    pd_i['PANEL'] = pd_i['PANEL'].astype(str).str[3].fillna('NA')\n",
    "    value_i = 0.05/pd_i.shape[0]\n",
    "    df_i = pd.DataFrame(columns = pd_i.columns.tolist())\n",
    "    for index, row in pd_i.iterrows():\n",
    "        if row['TWAS.P'] < value_i:\n",
    "            df_i.loc[index] = row\n",
    "    df_i.to_csv(f'./output/top_csf_cardio_case/PD.{i}.top', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#use row numbers (number of genes which is ID) to divide 0.05  ---> this is manula way\n",
    "\n",
    "#!cat ./output/pd/PD.1.dat | awk 'NR == 1 || $NF < 0.05/1036' > ./output/top/PD.1.top\n",
    "#!cat ./output/pd/PD.2.dat | awk 'NR == 1 || $NF < 0.05/667' > ./output/top/PD.2.top\n",
    "#!cat ./output/pd/PD.3.dat | awk 'NR == 1 || $NF < 0.05/567' > ./output/top/PD.3.top\n",
    "#!cat ./output/pd/PD.4.dat | awk 'NR == 1 || $NF < 0.05/411' > ./output/top/PD.4.top\n",
    "#!cat ./output/pd/PD.5.dat | awk 'NR == 1 || $NF < 0.05/465' > ./output/top/PD.5.top\n",
    "#!cat ./output/pd/PD.6.dat | awk 'NR == 1 || $NF < 0.05/467' > ./output/top/PD.6.top\n",
    "#!cat ./output/pd/PD.7.dat | awk 'NR == 1 || $NF < 0.05/498' > ./output/top/PD.7.top\n",
    "#!cat ./output/pd/PD.8.dat | awk 'NR == 1 || $NF < 0.05/341' > ./output/top/PD.8.top\n",
    "#!cat ./output/pd/PD.9.dat | awk 'NR == 1 || $NF < 0.05/386' > ./output/top/PD.9.top\n",
    "#!cat ./output/pd/PD.10.dat | awk 'NR == 1 || $NF < 0.05/389' > ./output/top/PD.10.top\n",
    "#!cat ./output/pd/PD.11.dat | awk 'NR == 1 || $NF < 0.05/575' > ./output/top/PD.11.top\n",
    "#!cat ./output/pd/PD.12.dat | awk 'NR == 1 || $NF < 0.05/504' > ./output/top/PD.12.top\n",
    "#!cat ./output/pd/PD.13.dat | awk 'NR == 1 || $NF < 0.05/178' > ./output/top/PD.13.top\n",
    "#!cat ./output/pd/PD.14.dat | awk 'NR == 1 || $NF < 0.05/301' > ./output/top/PD.14.top\n",
    "#!cat ./output/pd/PD.15.dat | awk 'NR == 1 || $NF < 0.05/320' > ./output/top/PD.15.top\n",
    "#!cat ./output/pd/PD.16.dat | awk 'NR == 1 || $NF < 0.05/406' > ./output/top/PD.16.top\n",
    "#!cat ./output/pd/PD.17.dat | awk 'NR == 1 || $NF < 0.05/520' > ./output/top/PD.17.top\n",
    "#!cat ./output/pd/PD.18.dat | awk 'NR == 1 || $NF < 0.05/147' > ./output/top/PD.18.top\n",
    "#!cat ./output/pd/PD.19.dat | awk 'NR == 1 || $NF < 0.05/692' > ./output/top/PD.19.top\n",
    "#!cat ./output/pd/PD.20.dat | awk 'NR == 1 || $NF < 0.05/250' > ./output/top/PD.20.top\n",
    "#!cat ./output/pd/PD.21.dat | awk 'NR == 1 || $NF < 0.05/112' > ./output/top/PD.21.top\n",
    "#!cat ./output/pd/PD.22.dat | awk 'NR == 1 || $NF < 0.05/246' > ./output/top/PD.22.top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.1.top --out output/analysis_csf_cardio_case/PD.1.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 1 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.10.top --out output/analysis_csf_cardio_case/PD.10.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 10 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.11.top --out output/analysis_csf_cardio_case/PD.11.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 11 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.13.top --out output/analysis_csf_cardio_case/PD.13.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 13 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.14.top --out output/analysis_csf_cardio_case/PD.14.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 14 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.15.top --out output/analysis_csf_cardio_case/PD.15.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 15 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.16.top --out output/analysis_csf_cardio_case/PD.16.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 16 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.19.top --out output/analysis_csf_cardio_case/PD.19.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 19 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.2.top --out output/analysis_csf_cardio_case/PD.2.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 2 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.22.top --out output/analysis_csf_cardio_case/PD.22.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 22 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.4.top --out output/analysis_csf_cardio_case/PD.4.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 4 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.5.top --out output/analysis_csf_cardio_case/PD.5.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 5 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.6.top --out output/analysis_csf_cardio_case/PD.6.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 6 --plot --locus_win 100000\n",
      "Executing: Rscript /data/songy4/twas/fusion_twas/FUSION.post_process.R --sumstats /data/songy4/proteomics_196/data_folder/meta.txt --input output/top_csf_cardio_case/PD.7.top --out output/analysis_csf_cardio_case/PD.7.top.analysis --ref_ld_chr /data/songy4/twas/fusion_twas/LDREF/1000G.EUR. --chr 7 --plot --locus_win 100000\n"
     ]
    }
   ],
   "source": [
    "#fusion post process \n",
    "for i in file_name:\n",
    "    anal_dir = f'output/analysis_csf_cardio_case'\n",
    "    top_dir = f'output/top_csf_cardio_case'\n",
    "    fusion_post_cmd_i = f'\\\n",
    "    Rscript {fusion_post_script} \\\n",
    "    --sumstats {sumstat_path} \\\n",
    "    --input {top_dir}/PD.{i}.top \\\n",
    "    --out {anal_dir}/PD.{i}.top.analysis \\\n",
    "    --ref_ld_chr {fusion_ldref_basename}. \\\n",
    "    --chr {i} \\\n",
    "    --plot --locus_win 100000'\n",
    "    shell_do(fusion_post_cmd_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#grab all gene ID from to.analysis.joint_included.dat files\n",
    "import glob\n",
    "\n",
    "def read_id(file):\n",
    "    return pd.read_csv(file, delim_whitespace=1, usecols=[1])\n",
    "\n",
    "files = glob.glob(f'{anal_dir}/PD.*.joint_included.dat')\n",
    "    \n",
    "csf_cardio_case_df = pd.concat([read_id(file) for file in files], axis=0)\n",
    "\n",
    "csf_cardio_case_df.to_csv(r'./data_folder/pdbp_ppmi_proteomics196_csf_cardio_result_case.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of gwas: (439, 2)\n",
      "shape of csf_cardio_case: (4, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FCGR2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PM20D1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LGALS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PON2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID\n",
       "0  FCGR2A\n",
       "1  PM20D1\n",
       "2  LGALS3\n",
       "3    PON2"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#compare twas_genes_case.csv, twas_genes_control.csv and gwas_genes.csv\n",
    "gwas = pd.read_csv(f\"{datadir}/gwas.csv\")\n",
    "csf_cardio_case = pd.read_csv(r\"./data_folder/pdbp_ppmi_proteomics196_csf_cardio_result_case.csv\", sep=' ')\n",
    "#twas_cont = pd.read_csv(r\"./data_folder/twas_genes_control.csv\", sep=' ')\n",
    "\n",
    "print('shape of gwas:', gwas.shape)\n",
    "print('shape of csf_cardio_case:', csf_cardio_case.shape)\n",
    "#print('shape of twas control:', twas_cont.shape)\n",
    "csf_cardio_case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of csf_cardio_case genes in GWAS genes: 1\n",
      "number of csf_cardio_case genes not in GWAS genes: 3\n"
     ]
    }
   ],
   "source": [
    "#TWAS case & GWAS significant hits\n",
    "csf_cardio_case_gwas = csf_cardio_case[csf_cardio_case['ID'].isin(gwas['Gene'])].reset_index(drop=True)\n",
    "print(\"number of csf_cardio_case genes in GWAS genes:\", csf_cardio_case_gwas.shape[0])\n",
    "\n",
    "#TWAS case & GWAS significant hits\n",
    "#twas_cont_gwas = twas_cont[twas_cont['ID'].isin(gwas['ID'])].reset_index(drop=True)\n",
    "#print(\"number of TWAS control genes in GWAS genes:\", twas_cont_gwas.shape[0])\n",
    "\n",
    "#TWAS case & GWAS significant hits\n",
    "csf_cardio_case_not_gwas = csf_cardio_case[~csf_cardio_case['ID'].isin(gwas['Gene'])].reset_index(drop=True)\n",
    "print(\"number of csf_cardio_case genes not in GWAS genes:\", csf_cardio_case_not_gwas.shape[0])\n",
    "\n",
    "#TWAS case & GWAS significant hits\n",
    "#twas_cont_not_gwas = twas_cont[~twas_cont['ID'].isin(gwas['ID'])].reset_index(drop=True)\n",
    "#print(\"number of TWAS control genes not in GWAS genes:\", twas_cont_not_gwas.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>P0</th>\n",
       "      <th>P1</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>BEST.GWAS.Z</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>EQTL.R2</th>\n",
       "      <th>EQTL.Z</th>\n",
       "      <th>EQTL.GWAS.Z</th>\n",
       "      <th>NSNP</th>\n",
       "      <th>NWGT</th>\n",
       "      <th>MODEL</th>\n",
       "      <th>MODELCV.R2</th>\n",
       "      <th>MODELCV.PV</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FCGR2A</td>\n",
       "      <td>1</td>\n",
       "      <td>161475220</td>\n",
       "      <td>161493803</td>\n",
       "      <td>0.759</td>\n",
       "      <td>rs1801274</td>\n",
       "      <td>-6.79</td>\n",
       "      <td>rs12722986</td>\n",
       "      <td>0.1395</td>\n",
       "      <td>-6.21</td>\n",
       "      <td>-4.951</td>\n",
       "      <td>442</td>\n",
       "      <td>15</td>\n",
       "      <td>lasso</td>\n",
       "      <td>0.26</td>\n",
       "      <td>1.600000e-07</td>\n",
       "      <td>4.676231</td>\n",
       "      <td>2.920000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PM20D1</td>\n",
       "      <td>1</td>\n",
       "      <td>205797150</td>\n",
       "      <td>205819260</td>\n",
       "      <td>0.587</td>\n",
       "      <td>rs823114</td>\n",
       "      <td>-11.31</td>\n",
       "      <td>rs12565968</td>\n",
       "      <td>0.1471</td>\n",
       "      <td>4.05</td>\n",
       "      <td>-7.559</td>\n",
       "      <td>386</td>\n",
       "      <td>15</td>\n",
       "      <td>lasso</td>\n",
       "      <td>0.19</td>\n",
       "      <td>1.200000e-05</td>\n",
       "      <td>-8.463696</td>\n",
       "      <td>2.590000e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LGALS3</td>\n",
       "      <td>14</td>\n",
       "      <td>55590828</td>\n",
       "      <td>55612126</td>\n",
       "      <td>0.441</td>\n",
       "      <td>rs8018800</td>\n",
       "      <td>-8.22</td>\n",
       "      <td>rs17672376</td>\n",
       "      <td>0.1910</td>\n",
       "      <td>-5.48</td>\n",
       "      <td>-2.340</td>\n",
       "      <td>431</td>\n",
       "      <td>15</td>\n",
       "      <td>lasso</td>\n",
       "      <td>0.28</td>\n",
       "      <td>3.300000e-08</td>\n",
       "      <td>4.360000</td>\n",
       "      <td>1.280000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PON2</td>\n",
       "      <td>7</td>\n",
       "      <td>95034175</td>\n",
       "      <td>95064510</td>\n",
       "      <td>0.460</td>\n",
       "      <td>rs7778623</td>\n",
       "      <td>-4.76</td>\n",
       "      <td>rs6973380</td>\n",
       "      <td>0.4133</td>\n",
       "      <td>-6.33</td>\n",
       "      <td>-4.190</td>\n",
       "      <td>475</td>\n",
       "      <td>1</td>\n",
       "      <td>top1</td>\n",
       "      <td>0.41</td>\n",
       "      <td>3.800000e-12</td>\n",
       "      <td>4.190000</td>\n",
       "      <td>2.800000e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID CHR         P0         P1    HSQ BEST.GWAS.ID  BEST.GWAS.Z  \\\n",
       "0  FCGR2A   1  161475220  161493803  0.759    rs1801274        -6.79   \n",
       "1  PM20D1   1  205797150  205819260  0.587     rs823114       -11.31   \n",
       "2  LGALS3  14   55590828   55612126  0.441    rs8018800        -8.22   \n",
       "3    PON2   7   95034175   95064510  0.460    rs7778623        -4.76   \n",
       "\n",
       "      EQTL.ID  EQTL.R2  EQTL.Z  EQTL.GWAS.Z NSNP NWGT  MODEL  MODELCV.R2  \\\n",
       "0  rs12722986   0.1395   -6.21       -4.951  442   15  lasso        0.26   \n",
       "1  rs12565968   0.1471    4.05       -7.559  386   15  lasso        0.19   \n",
       "2  rs17672376   0.1910   -5.48       -2.340  431   15  lasso        0.28   \n",
       "3   rs6973380   0.4133   -6.33       -4.190  475    1   top1        0.41   \n",
       "\n",
       "     MODELCV.PV    TWAS.Z        TWAS.P  \n",
       "0  1.600000e-07  4.676231  2.920000e-06  \n",
       "1  1.200000e-05 -8.463696  2.590000e-17  \n",
       "2  3.300000e-08  4.360000  1.280000e-05  \n",
       "3  3.800000e-12  4.190000  2.800000e-05  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab all gene ID from to.analysis.joint_included.dat files\n",
    "import glob\n",
    "\n",
    "def read_id(file):\n",
    "    return pd.read_csv(file, delim_whitespace=1)\n",
    "\n",
    "files = glob.glob(f'{top_dir}/*.top')\n",
    "    \n",
    "case_top = pd.concat([read_id(file) for file in files], axis=0).reset_index(drop=True)\n",
    "case_top.drop(case_top.iloc[:,0:2], axis=1, inplace=True)\n",
    "csf_cardio_case_top = case_top.merge(csf_cardio_case, on='ID', how='inner').reset_index(drop=True)\n",
    "csf_cardio_case_top\n",
    "#twas_case_top.to_csv(r'./data_folder/twas_case_top_all.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of significant csf_cardio_case data: (4, 18)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>P0</th>\n",
       "      <th>P1</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>BEST.GWAS.Z</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>EQTL.R2</th>\n",
       "      <th>EQTL.Z</th>\n",
       "      <th>EQTL.GWAS.Z</th>\n",
       "      <th>NSNP</th>\n",
       "      <th>NWGT</th>\n",
       "      <th>MODEL</th>\n",
       "      <th>MODELCV.R2</th>\n",
       "      <th>MODELCV.PV</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FCGR2A</td>\n",
       "      <td>1</td>\n",
       "      <td>161475220</td>\n",
       "      <td>161493803</td>\n",
       "      <td>0.759</td>\n",
       "      <td>rs1801274</td>\n",
       "      <td>-6.79</td>\n",
       "      <td>rs12722986</td>\n",
       "      <td>0.1395</td>\n",
       "      <td>-6.21</td>\n",
       "      <td>-4.951</td>\n",
       "      <td>442</td>\n",
       "      <td>15</td>\n",
       "      <td>lasso</td>\n",
       "      <td>0.26</td>\n",
       "      <td>1.600000e-07</td>\n",
       "      <td>4.676231</td>\n",
       "      <td>2.920000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PM20D1</td>\n",
       "      <td>1</td>\n",
       "      <td>205797150</td>\n",
       "      <td>205819260</td>\n",
       "      <td>0.587</td>\n",
       "      <td>rs823114</td>\n",
       "      <td>-11.31</td>\n",
       "      <td>rs12565968</td>\n",
       "      <td>0.1471</td>\n",
       "      <td>4.05</td>\n",
       "      <td>-7.559</td>\n",
       "      <td>386</td>\n",
       "      <td>15</td>\n",
       "      <td>lasso</td>\n",
       "      <td>0.19</td>\n",
       "      <td>1.200000e-05</td>\n",
       "      <td>-8.463696</td>\n",
       "      <td>2.590000e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LGALS3</td>\n",
       "      <td>14</td>\n",
       "      <td>55590828</td>\n",
       "      <td>55612126</td>\n",
       "      <td>0.441</td>\n",
       "      <td>rs8018800</td>\n",
       "      <td>-8.22</td>\n",
       "      <td>rs17672376</td>\n",
       "      <td>0.1910</td>\n",
       "      <td>-5.48</td>\n",
       "      <td>-2.340</td>\n",
       "      <td>431</td>\n",
       "      <td>15</td>\n",
       "      <td>lasso</td>\n",
       "      <td>0.28</td>\n",
       "      <td>3.300000e-08</td>\n",
       "      <td>4.360000</td>\n",
       "      <td>1.280000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PON2</td>\n",
       "      <td>7</td>\n",
       "      <td>95034175</td>\n",
       "      <td>95064510</td>\n",
       "      <td>0.460</td>\n",
       "      <td>rs7778623</td>\n",
       "      <td>-4.76</td>\n",
       "      <td>rs6973380</td>\n",
       "      <td>0.4133</td>\n",
       "      <td>-6.33</td>\n",
       "      <td>-4.190</td>\n",
       "      <td>475</td>\n",
       "      <td>1</td>\n",
       "      <td>top1</td>\n",
       "      <td>0.41</td>\n",
       "      <td>3.800000e-12</td>\n",
       "      <td>4.190000</td>\n",
       "      <td>2.800000e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID CHR         P0         P1    HSQ BEST.GWAS.ID  BEST.GWAS.Z  \\\n",
       "0  FCGR2A   1  161475220  161493803  0.759    rs1801274        -6.79   \n",
       "1  PM20D1   1  205797150  205819260  0.587     rs823114       -11.31   \n",
       "2  LGALS3  14   55590828   55612126  0.441    rs8018800        -8.22   \n",
       "3    PON2   7   95034175   95064510  0.460    rs7778623        -4.76   \n",
       "\n",
       "      EQTL.ID  EQTL.R2  EQTL.Z  EQTL.GWAS.Z NSNP NWGT  MODEL  MODELCV.R2  \\\n",
       "0  rs12722986   0.1395   -6.21       -4.951  442   15  lasso        0.26   \n",
       "1  rs12565968   0.1471    4.05       -7.559  386   15  lasso        0.19   \n",
       "2  rs17672376   0.1910   -5.48       -2.340  431   15  lasso        0.28   \n",
       "3   rs6973380   0.4133   -6.33       -4.190  475    1   top1        0.41   \n",
       "\n",
       "     MODELCV.PV    TWAS.Z        TWAS.P  \n",
       "0  1.600000e-07  4.676231  2.920000e-06  \n",
       "1  1.200000e-05 -8.463696  2.590000e-17  \n",
       "2  3.300000e-08  4.360000  1.280000e-05  \n",
       "3  3.800000e-12  4.190000  2.800000e-05  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##grab significant GWAS hits (P-value threshold to 0.05/N-tests per datatype)\n",
    "sig_csf_cardio_case_top = csf_cardio_case_top[csf_cardio_case_top['TWAS.P']< 1.40E-04]\n",
    "#reset index\n",
    "sig_csf_cardio_case_top = sig_csf_cardio_case_top.reset_index(drop=True)\n",
    "print(\"shape of significant csf_cardio_case data:\", sig_csf_cardio_case_top.shape)\n",
    "\n",
    "sig_csf_cardio_case_top.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-8-ac9a63c4c9a4>:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  case_top['GWAS.hit'] = case_top.ID.isin(csf_cardio_case_gwas.ID).astype(int)\n",
      "<ipython-input-8-ac9a63c4c9a4>:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  case_top['GWAS.hit'] = case_top['GWAS.hit'].map({0: 'no', 1: 'yes'})\n"
     ]
    }
   ],
   "source": [
    "from scipy import stats\n",
    "#remove some columns\n",
    "case_top = sig_csf_cardio_case_top[['ID', 'CHR', 'EQTL.ID', 'EQTL.Z', 'TWAS.Z', 'TWAS.P']]\n",
    "#add EQTL.P\n",
    "#case_top['EQTL.P'] = case_top.stats.norm.cdf(-Z)\n",
    "#add GWAS.hit column 1 if ID is in twas_case_gwas 0 if not\n",
    "case_top['GWAS.hit'] = case_top.ID.isin(csf_cardio_case_gwas.ID).astype(int)\n",
    "#convert 0 to no and 1 to yes\n",
    "case_top['GWAS.hit'] = case_top['GWAS.hit'].map({0: 'no', 1: 'yes'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "case_top data types: \n",
      " ID           object\n",
      "CHR          object\n",
      "EQTL.ID      object\n",
      "EQTL.Z      float64\n",
      "TWAS.Z      float64\n",
      "TWAS.P      float64\n",
      "GWAS.hit     object\n",
      "dtype: object\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-9-20fcd3ca053a>:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  case_top['CHR'] = case_top['CHR'].astype(int)\n"
     ]
    }
   ],
   "source": [
    "#check data type for the dataframe\n",
    "print(\"case_top data types: \\n\", case_top.dtypes)\n",
    "#convert CHR from object data type to int\n",
    "case_top['CHR'] = case_top['CHR'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "case_top data types: \n",
      " ID           object\n",
      "CHR           int64\n",
      "EQTL.ID      object\n",
      "EQTL.Z      float64\n",
      "TWAS.Z      float64\n",
      "TWAS.P      float64\n",
      "GWAS.hit     object\n",
      "dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>EQTL.Z</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>GWAS.hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FCGR2A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12722986</td>\n",
       "      <td>-6.21</td>\n",
       "      <td>4.676231</td>\n",
       "      <td>2.920000e-06</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PM20D1</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12565968</td>\n",
       "      <td>4.05</td>\n",
       "      <td>-8.463696</td>\n",
       "      <td>2.590000e-17</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PON2</td>\n",
       "      <td>7</td>\n",
       "      <td>rs6973380</td>\n",
       "      <td>-6.33</td>\n",
       "      <td>4.190000</td>\n",
       "      <td>2.800000e-05</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LGALS3</td>\n",
       "      <td>14</td>\n",
       "      <td>rs17672376</td>\n",
       "      <td>-5.48</td>\n",
       "      <td>4.360000</td>\n",
       "      <td>1.280000e-05</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID  CHR     EQTL.ID  EQTL.Z    TWAS.Z        TWAS.P GWAS.hit\n",
       "0  FCGR2A    1  rs12722986   -6.21  4.676231  2.920000e-06      yes\n",
       "1  PM20D1    1  rs12565968    4.05 -8.463696  2.590000e-17       no\n",
       "2    PON2    7   rs6973380   -6.33  4.190000  2.800000e-05       no\n",
       "3  LGALS3   14  rs17672376   -5.48  4.360000  1.280000e-05       no"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check data type for the dataframe\n",
    "print(\"case_top data types: \\n\", case_top.dtypes)\n",
    "#sort the rows by CHR order\n",
    "case_top = case_top.sort_values('CHR').reset_index(drop=True)\n",
    "case_top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "no    1\n",
       "Name: GWAS.hit, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check if GWAS.hit has 40 yes and 60 no\n",
    "pd.value_counts(case_top['GWAS.hit'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save twas_cont_top and cont_top --run only once\n",
    "#twas_case_top.to_csv(r'./data_folder/twas_case_top_all.csv',index=False)\n",
    "sig_csf_cardio_case_top.to_csv(r'./data_folder/significant_csf_cardio_case_top_all.csv',index=False)\n",
    "case_top.to_csv(r'./data_folder/csf_cardio_case_top.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert QTL to SMR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno_path = f'{datadir}/qc_proteomics196_csf_cardio'\n",
    "sumstat_path = f'{datadir}/meta.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!smr --bfile {geno_path}_case_hg19_lifted --gwas-summary mygwas.ma --beqtl-summary myeqtl --out mysmr --thread-num 10 \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.8",
   "language": "python",
   "name": "py3.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
