{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2f19c7dc-ba59-4b30-a9f9-81c72b12971f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: tables in /usr/local/Anaconda/envs/py3.8/lib/python3.8/site-packages (3.6.1)\n",
      "Requirement already satisfied: numpy>=1.9.3 in /usr/local/Anaconda/envs/py3.8/lib/python3.8/site-packages (from tables) (1.19.5)\n",
      "Requirement already satisfied: numexpr>=2.6.2 in /usr/local/Anaconda/envs/py3.8/lib/python3.8/site-packages (from tables) (2.7.2)\n"
     ]
    }
   ],
   "source": [
    "#upload some libraries\n",
    "import os\n",
    "import pandas.testing as tm\n",
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import sys\n",
    "import subprocess\n",
    "import statsmodels.api as sm\n",
    "from scipy import stats\n",
    "import statsmodels.formula.api as smf\n",
    "#patsy builtins are for . in transcriptom ID\n",
    "from patsy.builtins import *\n",
    "import statistics\n",
    "import umap.umap_ as umap\n",
    "from joblib import dump, load\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import style\n",
    "import seaborn as sns\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "! pip install --upgrade tables\n",
    "import tables\n",
    "\n",
    "# set paths\n",
    "basedir = '/data/songy4/tes'\n",
    "datadir = f'{basedir}/data_folder'\n",
    "workdir = f'{datadir}/output'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a34bb26-3244-443d-904f-277aac40d8dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#download the renormalized datasets ---- run it only once\n",
    "#TElocal output files normalized with limma for a different timepoint, weren't adjusted for any covariates\n",
    "! cp -R /data/LNG/Frank/TE/limma/baseline_GENE_TE_norm_adjusted.cntTable /{datadir}/\n",
    "! cp -R /data/LNG/Frank/TE/limma/v04_GENE_TE_norm_adjusted.cntTable /{datadir}/\n",
    "! cp -R /data/LNG/Frank/TE/limma/v06_GENE_TE_norm_adjusted.cntTable /{datadir}/\n",
    "! cp -R /data/LNG/Frank/TE/limma/v08_GENE_TE_norm_adjusted.cntTable /{datadir}/\n",
    "\n",
    "#TE location data\n",
    "! cp -R /data/LNG/Frank/TE/hg38_rmsk_TE.gtf.locInd.locations /{datadir}/\n",
    "\n",
    "#gene locations\n",
    "! cp -R /data/LNG/Frank/TE/hg38_gene_annotation.gtf.locations /{datadir}/\n",
    "\n",
    "#plate information (one of covariates among phenotype (= case vs control), sex, age, plate, expression pcs 1 to 5)\n",
    "! cp -R /data/LNG/Frank/TE/TElocal_replicates/megaMetaTable.csv /{datadir}/\n",
    "\n",
    "#genomic data (plink files)\n",
    "! cp -R /data/LNG/Frank/TE/hits_plus_tags_AMPPD_Euros.bed /{datadir}/\n",
    "! cp -R /data/LNG/Frank/TE/hits_plus_tags_AMPPD_Euros.bim /{datadir}/\n",
    "! cp -R /data/LNG/Frank/TE/hits_plus_tags_AMPPD_Euros.fam /{datadir}/\n",
    "! cp -R /data/LNG/Frank/TE/hits_plus_tags_AMPPD_Euros.log /{datadir}/\n",
    "! cp -R /data/LNG/Frank/TE/hits_plus_tags_AMPPD_Euros.nosex /{datadir}/\n",
    "\n",
    "#PD GWAS sum stats\n",
    "! cp -R /data/LNG/Frank/meta5_sumstats_harmonized_hg19_hg38.csv /{datadir}/\n",
    "\n",
    "#pheno\n",
    "! cp -R /data/LNG/Frank/TE/DEMOG_DIAG.csv /{datadir}/\n",
    "\n",
    "#/data/CARD/PD/summary_stats --> PD GWAS sum stats were based on files from here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d91ce1f5-3df3-4753-9388-9c288459432e",
   "metadata": {},
   "source": [
    "Covariate data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b33475a9-577e-4999-9c27-d59cbb474903",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Anaconda/envs/py3.8/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (8,9,10,11,12) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of covariate data: (4891, 138)\n",
      "  HudAlphaSampleName  Description Alternate MRN Specimen Bar Code  \\\n",
      "0       5104-SL-0001       373111   ST-00009911        0000373111   \n",
      "1       5104-SL-0002  PP0016-0526   ST-00000049       PP0016-0526   \n",
      "2       5104-SL-0003      3125874   ST-00001277        0003125874   \n",
      "3       5104-SL-0004      3179818   ST-00020262        0003179818   \n",
      "4       5104-SL-0005      3164601   ST-00021309        0003164601   \n",
      "\n",
      "   Specimen Type  Quantity (ug)  Concentration (ng/ul)  Concentration Flag  \\\n",
      "0  Nucleic Acids            1.0                 115.90                   1   \n",
      "1  Nucleic Acids            1.0                  80.07                   1   \n",
      "2  Nucleic Acids            1.0                 219.70                   1   \n",
      "3  Nucleic Acids            1.0                 100.06                   1   \n",
      "4  Nucleic Acids            1.0                 124.40                   1   \n",
      "\n",
      "  Submitted Volume (ul) Normalization Volume (30ng/ul)  ...  \\\n",
      "0           8.628127696                     29.3718723  ...   \n",
      "1           12.48907206                    25.51092794  ...   \n",
      "2           4.551661356                    33.44833864  ...   \n",
      "3           9.994003598                     28.0059964  ...   \n",
      "4           8.038585209                    29.96141479  ...   \n",
      "\n",
      "  PCT_R1_TRANSCRIPT_STRAND_READS MEDIAN_5PRIME_BIAS PCT_USABLE_BASES  \\\n",
      "0                         1.8975           1.019442          47.9359   \n",
      "1                         1.8572           0.965040          39.6130   \n",
      "2                         4.5679           0.997003          34.8096   \n",
      "3                         1.6215           1.013765          45.3119   \n",
      "4                         1.9322           0.985280          34.8793   \n",
      "\n",
      "   pct_globin  pct_rRNA  genes detected at count > 0  \\\n",
      "0    0.002246  0.001102                      31660.0   \n",
      "1    0.000907  0.000517                      34466.0   \n",
      "2    0.003358  0.002375                      40292.0   \n",
      "3    0.004553  0.000982                      32349.0   \n",
      "4    0.002511  0.000682                      30530.0   \n",
      "\n",
      "  genes detected at count > 10  genes detected at count > 100  \\\n",
      "0                      21917.0                        14454.0   \n",
      "1                      21018.0                        13525.0   \n",
      "2                      21224.0                        12627.0   \n",
      "3                      20163.0                        12868.0   \n",
      "4                      20580.0                        13213.0   \n",
      "\n",
      "   genes detected at count > 1000 genes detected at count > 10000  \n",
      "0                          7358.0                           866.0  \n",
      "1                          6522.0                           778.0  \n",
      "2                          5281.0                           491.0  \n",
      "3                          5687.0                           680.0  \n",
      "4                          5849.0                           512.0  \n",
      "\n",
      "[5 rows x 138 columns]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Quantity (ug)</th>\n",
       "      <th>Concentration (ng/ul)</th>\n",
       "      <th>Concentration Flag</th>\n",
       "      <th>260/280 Ratio</th>\n",
       "      <th>260/230 Ratio</th>\n",
       "      <th>RIN Value</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>Plate</th>\n",
       "      <th>PATNO</th>\n",
       "      <th>SITE</th>\n",
       "      <th>...</th>\n",
       "      <th>PCT_R1_TRANSCRIPT_STRAND_READS</th>\n",
       "      <th>MEDIAN_5PRIME_BIAS</th>\n",
       "      <th>PCT_USABLE_BASES</th>\n",
       "      <th>pct_globin</th>\n",
       "      <th>pct_rRNA</th>\n",
       "      <th>genes detected at count &gt; 0</th>\n",
       "      <th>genes detected at count &gt; 10</th>\n",
       "      <th>genes detected at count &gt; 100</th>\n",
       "      <th>genes detected at count &gt; 1000</th>\n",
       "      <th>genes detected at count &gt; 10000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>4891.000000</td>\n",
       "      <td>4861.000000</td>\n",
       "      <td>4891.000000</td>\n",
       "      <td>4784.000000</td>\n",
       "      <td>4781.000000</td>\n",
       "      <td>4837.000000</td>\n",
       "      <td>4891.000000</td>\n",
       "      <td>4891.000000</td>\n",
       "      <td>4841.000000</td>\n",
       "      <td>4782.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4878.000000</td>\n",
       "      <td>4878.000000</td>\n",
       "      <td>4878.000000</td>\n",
       "      <td>4879.000000</td>\n",
       "      <td>4879.000000</td>\n",
       "      <td>4877.000000</td>\n",
       "      <td>4877.000000</td>\n",
       "      <td>4877.000000</td>\n",
       "      <td>4877.000000</td>\n",
       "      <td>4877.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.022042</td>\n",
       "      <td>82.121765</td>\n",
       "      <td>0.939890</td>\n",
       "      <td>2.095542</td>\n",
       "      <td>1.309453</td>\n",
       "      <td>7.702791</td>\n",
       "      <td>45.749744</td>\n",
       "      <td>106.505009</td>\n",
       "      <td>20481.995042</td>\n",
       "      <td>124.598703</td>\n",
       "      <td>...</td>\n",
       "      <td>5.209730</td>\n",
       "      <td>0.995533</td>\n",
       "      <td>39.153704</td>\n",
       "      <td>0.009166</td>\n",
       "      <td>0.029934</td>\n",
       "      <td>36876.692434</td>\n",
       "      <td>21645.114210</td>\n",
       "      <td>12340.213656</td>\n",
       "      <td>4696.244002</td>\n",
       "      <td>354.919008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.105439</td>\n",
       "      <td>47.209438</td>\n",
       "      <td>0.237716</td>\n",
       "      <td>0.445532</td>\n",
       "      <td>5.070301</td>\n",
       "      <td>1.646776</td>\n",
       "      <td>27.042168</td>\n",
       "      <td>99.314465</td>\n",
       "      <td>22379.247093</td>\n",
       "      <td>119.956888</td>\n",
       "      <td>...</td>\n",
       "      <td>6.655015</td>\n",
       "      <td>0.072550</td>\n",
       "      <td>9.050471</td>\n",
       "      <td>0.214597</td>\n",
       "      <td>0.618656</td>\n",
       "      <td>8083.571934</td>\n",
       "      <td>4620.730462</td>\n",
       "      <td>1252.296439</td>\n",
       "      <td>1169.621370</td>\n",
       "      <td>180.757908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.600000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.240000</td>\n",
       "      <td>-267.800000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1009.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.050000</td>\n",
       "      <td>0.033756</td>\n",
       "      <td>0.028400</td>\n",
       "      <td>0.000135</td>\n",
       "      <td>0.000250</td>\n",
       "      <td>597.000000</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.050000</td>\n",
       "      <td>0.790000</td>\n",
       "      <td>7.200000</td>\n",
       "      <td>22.000000</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>3422.000000</td>\n",
       "      <td>28.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.052500</td>\n",
       "      <td>0.964074</td>\n",
       "      <td>35.539100</td>\n",
       "      <td>0.001585</td>\n",
       "      <td>0.002980</td>\n",
       "      <td>30951.000000</td>\n",
       "      <td>19401.000000</td>\n",
       "      <td>11882.000000</td>\n",
       "      <td>4152.000000</td>\n",
       "      <td>237.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>72.200000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.090000</td>\n",
       "      <td>1.360000</td>\n",
       "      <td>8.100000</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>27.000000</td>\n",
       "      <td>3905.000000</td>\n",
       "      <td>73.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.665150</td>\n",
       "      <td>1.003158</td>\n",
       "      <td>40.182800</td>\n",
       "      <td>0.002378</td>\n",
       "      <td>0.005781</td>\n",
       "      <td>33340.000000</td>\n",
       "      <td>20189.000000</td>\n",
       "      <td>12428.000000</td>\n",
       "      <td>4823.000000</td>\n",
       "      <td>337.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>103.900000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.130000</td>\n",
       "      <td>1.750000</td>\n",
       "      <td>8.800000</td>\n",
       "      <td>69.000000</td>\n",
       "      <td>212.000000</td>\n",
       "      <td>41432.000000</td>\n",
       "      <td>289.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.505450</td>\n",
       "      <td>1.039712</td>\n",
       "      <td>44.599175</td>\n",
       "      <td>0.003741</td>\n",
       "      <td>0.012073</td>\n",
       "      <td>41407.000000</td>\n",
       "      <td>21470.000000</td>\n",
       "      <td>12916.000000</td>\n",
       "      <td>5412.000000</td>\n",
       "      <td>451.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.000000</td>\n",
       "      <td>701.100000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>26.820000</td>\n",
       "      <td>185.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>96.000000</td>\n",
       "      <td>225.000000</td>\n",
       "      <td>92834.000000</td>\n",
       "      <td>335.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>95.589300</td>\n",
       "      <td>1.224844</td>\n",
       "      <td>68.048300</td>\n",
       "      <td>13.544165</td>\n",
       "      <td>38.354647</td>\n",
       "      <td>56464.000000</td>\n",
       "      <td>49526.000000</td>\n",
       "      <td>21136.000000</td>\n",
       "      <td>8145.000000</td>\n",
       "      <td>1438.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Quantity (ug)  Concentration (ng/ul)  Concentration Flag  \\\n",
       "count    4891.000000            4861.000000         4891.000000   \n",
       "mean        1.022042              82.121765            0.939890   \n",
       "std         0.105439              47.209438            0.237716   \n",
       "min         1.000000               4.600000            0.000000   \n",
       "25%         1.000000              50.000000            1.000000   \n",
       "50%         1.000000              72.200000            1.000000   \n",
       "75%         1.000000             103.900000            1.000000   \n",
       "max         2.000000             701.100000            1.000000   \n",
       "\n",
       "       260/280 Ratio  260/230 Ratio    RIN Value     POSITION        Plate  \\\n",
       "count    4784.000000    4781.000000  4837.000000  4891.000000  4891.000000   \n",
       "mean        2.095542       1.309453     7.702791    45.749744   106.505009   \n",
       "std         0.445532       5.070301     1.646776    27.042168    99.314465   \n",
       "min         0.240000    -267.800000     0.000000     1.000000     1.000000   \n",
       "25%         2.050000       0.790000     7.200000    22.000000    13.000000   \n",
       "50%         2.090000       1.360000     8.100000    45.000000    27.000000   \n",
       "75%         2.130000       1.750000     8.800000    69.000000   212.000000   \n",
       "max        26.820000     185.000000    10.000000    96.000000   225.000000   \n",
       "\n",
       "              PATNO         SITE  ...  PCT_R1_TRANSCRIPT_STRAND_READS  \\\n",
       "count   4841.000000  4782.000000  ...                     4878.000000   \n",
       "mean   20481.995042   124.598703  ...                        5.209730   \n",
       "std    22379.247093   119.956888  ...                        6.655015   \n",
       "min     1009.000000     1.000000  ...                        1.050000   \n",
       "25%     3422.000000    28.000000  ...                        2.052500   \n",
       "50%     3905.000000    73.000000  ...                        2.665150   \n",
       "75%    41432.000000   289.000000  ...                        5.505450   \n",
       "max    92834.000000   335.000000  ...                       95.589300   \n",
       "\n",
       "       MEDIAN_5PRIME_BIAS  PCT_USABLE_BASES   pct_globin     pct_rRNA  \\\n",
       "count         4878.000000       4878.000000  4879.000000  4879.000000   \n",
       "mean             0.995533         39.153704     0.009166     0.029934   \n",
       "std              0.072550          9.050471     0.214597     0.618656   \n",
       "min              0.033756          0.028400     0.000135     0.000250   \n",
       "25%              0.964074         35.539100     0.001585     0.002980   \n",
       "50%              1.003158         40.182800     0.002378     0.005781   \n",
       "75%              1.039712         44.599175     0.003741     0.012073   \n",
       "max              1.224844         68.048300    13.544165    38.354647   \n",
       "\n",
       "       genes detected at count > 0  genes detected at count > 10  \\\n",
       "count                  4877.000000                   4877.000000   \n",
       "mean                  36876.692434                  21645.114210   \n",
       "std                    8083.571934                   4620.730462   \n",
       "min                     597.000000                     14.000000   \n",
       "25%                   30951.000000                  19401.000000   \n",
       "50%                   33340.000000                  20189.000000   \n",
       "75%                   41407.000000                  21470.000000   \n",
       "max                   56464.000000                  49526.000000   \n",
       "\n",
       "       genes detected at count > 100  genes detected at count > 1000  \\\n",
       "count                    4877.000000                     4877.000000   \n",
       "mean                    12340.213656                     4696.244002   \n",
       "std                      1252.296439                     1169.621370   \n",
       "min                         4.000000                        0.000000   \n",
       "25%                     11882.000000                     4152.000000   \n",
       "50%                     12428.000000                     4823.000000   \n",
       "75%                     12916.000000                     5412.000000   \n",
       "max                     21136.000000                     8145.000000   \n",
       "\n",
       "       genes detected at count > 10000  \n",
       "count                      4877.000000  \n",
       "mean                        354.919008  \n",
       "std                         180.757908  \n",
       "min                           0.000000  \n",
       "25%                         237.000000  \n",
       "50%                         337.000000  \n",
       "75%                         451.000000  \n",
       "max                        1438.000000  \n",
       "\n",
       "[8 rows x 98 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open pheno file megaMetaTable.csv\n",
    "covariate = pd.read_csv(f\"{datadir}/megaMetaTable.csv\", index_col=False)\n",
    "\n",
    "print('shape of covariate data:', covariate.shape)\n",
    "print(covariate.head())\n",
    "\n",
    "covariate.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3a65a5ec-0bee-4657-97d1-ab2dce0dea83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted Volume (ul):  <class 'str'>\n",
      "Normalization Volume (30ng/ul):  <class 'str'>\n",
      "Total Volume (ul):  <class 'str'>\n",
      "Input RNASeq (ng):  <class 'str'>\n",
      "Input miRNA (ng):  <class 'str'>\n",
      "Column names: \n",
      " ['HudAlphaSampleName', 'Description', 'Alternate MRN', 'Specimen Bar Code', 'Specimen Type', 'Quantity (ug)', 'Concentration (ng/ul)', 'Concentration Flag', 'Submitted Volume (ul)', 'Normalization Volume (30ng/ul)', 'Total Volume (ul)', 'Input RNASeq (ng)', 'Input miRNA (ng)', '260/280 Ratio', '260/230 Ratio', 'RIN Value', 'BOX', 'POSITION', 'Plate', 'Phase', 'PoolAssign', 'HudAlphaLibraryID', 'sampleid', 'DateAnalyzed', 'PATNO', 'SEX', 'DIAGNOSIS', 'CLINICAL_EVENT', 'SITE', 'CNO', 'QCflagIR3', 'VISIT', 'PATNO_VISIT', 'VISIT_POST_MED_USE', 'VISIT_PRIOR_MED_USE', 'months_between_collection_firstMedUse', 'hours_between_PDdose_collection', 'age_at_consent', 'age_at_diagnosis', 'time_since_diagnosis', 'ageBin', 'Disease_Status', 'Disease_Status_byVisit', 'Study_Arm', 'Most_likely_primary_diagnosis', 'Genetic_Status', 'Hispanic_or_Latino', 'AmericanIndian_or_AlaskaNative', 'Asian', 'Black_or_AfricanAmerican', 'Hawaiian_or_OtherPacificIslander', 'White', 'Race_not_specified', 'EXCLUDED', 'P3GRP', 'Prodromal_group_type', 'Monocytes', 'Eosinophils', 'Basophils', 'Platelets', 'Neutrophils (%)', 'Lymphocytes (%)', 'Monocytes (%)', 'Eosinophils (%)', 'Basophils (%)', 'RBC', 'Hemoglobin', 'RBC Morphology', 'Neutrophils', 'Lymphocytes', 'Atypical Lymphocytes', 'PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6', 'PC7', 'PC8', 'PC9', 'PC10', 'Sample', 'deletion_length', 'insertion_length', 'mismatch_rate', 'multimapped_percent', 'uniquely_mapped_percent', 'num_annotated_splices', 'insertion_rate', 'unmapped_other_percent', 'multimapped', 'num_splices', 'avg_input_read_length', 'multimapped_toomany', 'total_reads', 'num_GCAG_splices', 'unmapped_tooshort', 'unmapped_other', 'unmapped_mismatches_percent', 'num_noncanonical_splices', 'num_ATAC_splices', 'multimapped_toomany_percent', 'avg_mapped_read_length', 'unmapped_mismatches', 'deletion_rate', 'uniquely_mapped', 'unmapped_tooshort_percent', 'num_GTAG_splices', 'PF_MISMATCH_RATE', 'PF_READS_ALIGNED', 'TOTAL_READS', 'PF_HQ_ERROR_RATE', 'PCT_CHIMERAS', 'PF_INDEL_RATE', 'PCT_PF_READS_ALIGNED', 'PCT_ADAPTER', 'MEAN_INSERT_SIZE', 'PCT_UTR_BASES', 'PCT_INTERGENIC_BASES', 'MEDIAN_CV_COVERAGE', 'PCT_CODING_BASES', 'MEDIAN_3PRIME_BIAS', 'NUM_UNEXPLAINED_READS', 'PCT_RIBOSOMAL_BASES', 'PCT_INTRONIC_BASES', 'PCT_R2_TRANSCRIPT_STRAND_READS', 'PCT_MRNA_BASES', 'MEDIAN_5PRIME_TO_3PRIME_BIAS', 'PCT_R1_TRANSCRIPT_STRAND_READS', 'MEDIAN_5PRIME_BIAS', 'PCT_USABLE_BASES', 'pct_globin', 'pct_rRNA', 'genes detected at count > 0', 'genes detected at count > 10', 'genes detected at count > 100', 'genes detected at count > 1000', 'genes detected at count > 10000']\n"
     ]
    }
   ],
   "source": [
    "#check the datatype for column 8 to 12\n",
    "for col in covariate.columns[8:13]:\n",
    "    print(str(col) + ': ', type(covariate[col].iat[0]))\n",
    "\n",
    "#check header\n",
    "print('Column names: \\n',covariate.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3113bd31-6f52-4e95-ac1a-a9a8a8cfd143",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique values in White column: \n",
      " ['Yes' 'No' nan 'Unknown or not reported']\n",
      "Null value in each column: \n",
      " Plate                                  0\n",
      "PATNO                                 50\n",
      "SEX                                    0\n",
      "DIAGNOSIS                              0\n",
      "age_at_consent                       110\n",
      "age_at_diagnosis                    3493\n",
      "Hispanic_or_Latino                   117\n",
      "AmericanIndian_or_AlaskaNative       117\n",
      "Asian                                117\n",
      "Black_or_AfricanAmerican             117\n",
      "Hawaiian_or_OtherPacificIslander     117\n",
      "White                                117\n",
      "Race_not_specified                   117\n",
      "dtype: int64\n",
      "Value counts in White column: \n",
      " Yes                        4564\n",
      "No                          195\n",
      "Unknown or not reported      15\n",
      "Name: White, dtype: int64\n",
      "cov_df shape:  (4564, 5)\n",
      "    PATNO     SEX DIAGNOSIS  age_at_consent White\n",
      "0  3174.0    Male        PD       51.162218   Yes\n",
      "1  3201.0  Female   Control       64.832307   Yes\n",
      "2  3057.0  Female   Control       60.000000   Yes\n",
      "3  4126.0    Male        PD       56.161533   Yes\n",
      "4  4117.0  Female        PD       60.161533   Yes\n",
      "Number of unique values in PATNO column:  1541\n"
     ]
    }
   ],
   "source": [
    "#pick the columns needed\n",
    "covariate_df = covariate[['Plate','PATNO','SEX', 'DIAGNOSIS', 'age_at_consent', 'age_at_diagnosis', 'Hispanic_or_Latino','AmericanIndian_or_AlaskaNative','Asian','Black_or_AfricanAmerican','Hawaiian_or_OtherPacificIslander','White','Race_not_specified']]\n",
    "\n",
    "#check the unique value in white column\n",
    "print('Unique values in White column: \\n', covariate_df.White.unique())\n",
    "\n",
    "#check nan numbers in each column \n",
    "print('Null value in each column: \\n', covariate_df.isna().sum())\n",
    "\n",
    "#check how many each value in white column\n",
    "print('Value counts in White column: \\n',covariate_df.White.value_counts())\n",
    "\n",
    "#drop rows where NaN\n",
    "covariate_df = covariate_df[covariate_df.White.notna()]\n",
    "#drop rows if not Yes\n",
    "cov_white = covariate_df[covariate_df.White.str.contains(\"Yes\")]\n",
    "\n",
    "#drop some columns \n",
    "cov_df = cov_white[['PATNO', 'SEX', 'DIAGNOSIS', 'age_at_consent', 'White']]\n",
    "print('cov_df shape: ', cov_df.shape)\n",
    "print(cov_df.head())\n",
    "\n",
    "#count unique number of patient ID\n",
    "print('Number of unique values in PATNO column: ',cov_df['PATNO'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "287e01d9-4121-4c26-9f9a-d79db8c35de0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of patient ID that has duplicate:  1541\n",
      "number of patient ID that has no duplicate:  449\n",
      "        PATNO   SEX  DIAGNOSIS  age_at_consent White\n",
      "947   92834.0  Male  Prodromal       66.748802   Yes\n",
      "4049  92834.0  Male  Prodromal       66.748802   Yes\n",
      "4061  92834.0  Male  Prodromal       66.748802   Yes\n",
      "       PATNO   SEX DIAGNOSIS  age_at_consent White\n",
      "872   3001.0  Male        PD       65.084189   Yes\n",
      "902   3001.0  Male        PD       65.084189   Yes\n",
      "912   3001.0  Male        PD       65.084189   Yes\n",
      "942   3001.0  Male        PD       65.084189   Yes\n",
      "2680  3001.0  Male        PD       65.084189   Yes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATNO</th>\n",
       "      <th>SEX</th>\n",
       "      <th>DIAGNOSIS</th>\n",
       "      <th>age_at_consent</th>\n",
       "      <th>White</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3000.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>Control</td>\n",
       "      <td>69.998631</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3001.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>PD</td>\n",
       "      <td>65.084189</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3002.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>PD</td>\n",
       "      <td>68.161533</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3003.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>PD</td>\n",
       "      <td>57.160849</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3004.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Control</td>\n",
       "      <td>60.161533</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>85242.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Prodromal</td>\n",
       "      <td>63.915127</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>90456.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Prodromal</td>\n",
       "      <td>82.496920</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1538</th>\n",
       "      <td>91837.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Prodromal</td>\n",
       "      <td>74.086242</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1539</th>\n",
       "      <td>92490.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Prodromal</td>\n",
       "      <td>72.914442</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1540</th>\n",
       "      <td>92834.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>Prodromal</td>\n",
       "      <td>66.748802</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1541 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        PATNO     SEX  DIAGNOSIS  age_at_consent White\n",
       "0      3000.0  Female    Control       69.998631   Yes\n",
       "1      3001.0    Male         PD       65.084189   Yes\n",
       "2      3002.0  Female         PD       68.161533   Yes\n",
       "3      3003.0  Female         PD       57.160849   Yes\n",
       "4      3004.0    Male    Control       60.161533   Yes\n",
       "...       ...     ...        ...             ...   ...\n",
       "1536  85242.0    Male  Prodromal       63.915127   Yes\n",
       "1537  90456.0    Male  Prodromal       82.496920   Yes\n",
       "1538  91837.0    Male  Prodromal       74.086242   Yes\n",
       "1539  92490.0    Male  Prodromal       72.914442   Yes\n",
       "1540  92834.0    Male  Prodromal       66.748802   Yes\n",
       "\n",
       "[1541 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##Check how many patient ID are duplicates and the duplicated records are identical or different\n",
    "#count how many times patient number is duplicated\n",
    "size = cov_df.groupby(cov_df['PATNO'].tolist(),as_index=False).size()\n",
    "print('number of patient ID that has duplicate: ', len(size))\n",
    "\n",
    "#check how many IDs are only count as 1\n",
    "size_1 = size.query(\"size==1\")\n",
    "print('number of patient ID that has no duplicate: ', len(size_1))\n",
    "\n",
    "#print a couple IDs that have duplicated to see what is going on\n",
    "\n",
    "size_2 = cov_df.query(\"PATNO==92834.0\")\n",
    "print(size_2.head())\n",
    "\n",
    "#print a couple IDs that have duplicated to see what is going on\n",
    "\n",
    "size_3 = cov_df.query(\"PATNO==3001.0\")\n",
    "print(size_3.head())\n",
    "\n",
    "#groupby patient number\n",
    "df_1 = cov_df.copy()\n",
    "df_1 = df_1.groupby(['PATNO']).first().reset_index()\n",
    "df_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "13846e5b-23ff-4767-b84c-ed3ba5c3a7eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        ID MALE    PHENO   AGE\n",
      "0  PP-3000    0  Control  70.0\n",
      "1  PP-3001    1       PD  65.0\n",
      "2  PP-3002    0       PD  68.0\n",
      "3  PP-3003    0       PD  57.0\n",
      "4  PP-3004    1  Control  60.0\n",
      "Unique values in PHENO:  ['Control' 'PD' 'SWEDD' 'Prodromal' 'Genetic Registry' 'Genetic Cohort']\n",
      "Count of each value in PHENO:  Genetic Registry    424\n",
      "Genetic Cohort      412\n",
      "PD                  403\n",
      "Control             184\n",
      "SWEDD                60\n",
      "Prodromal            58\n",
      "Name: PHENO, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#drop plate and age_at_diagnosis columns\n",
    "cov = df_1[['PATNO','SEX', 'DIAGNOSIS','age_at_consent']]\n",
    "#rename columns\n",
    "cov.columns = ['ID', 'MALE', 'PHENO','AGE']\n",
    "#drop rows where ID is NaN\n",
    "cov = cov[cov['ID'].notna()]\n",
    "\n",
    "#remove .0 from ID\n",
    "cov['ID'] = cov['ID'].apply(lambda f: format(f, '.0f'))\n",
    "\n",
    "#add pp. on PATNO\n",
    "cov['ID'] = 'PP-' + cov['ID'].astype(str)\n",
    "\n",
    "#convert Female to 0, Male to 1\n",
    "cov['MALE'] = cov['MALE'].map({'Female': 0, 'Male': 1})\n",
    "\n",
    "#remove .0 from MALE\n",
    "cov['MALE'] = cov['MALE'].apply(lambda f: format(f, '.0f'))\n",
    "\n",
    "#round AGE values\n",
    "cov.AGE = cov.AGE.round()\n",
    "                      \n",
    "print(cov.head())\n",
    "\n",
    "#check PHENO choices\n",
    "print('Unique values in PHENO: ', cov.PHENO.unique())\n",
    "\n",
    "#count each value in PHENO column\n",
    "print('Count of each value in PHENO: ', cov.PHENO.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "229e8d59-7dba-4fd3-89d2-3344f02fea73",
   "metadata": {},
   "source": [
    "Pheno data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "eecbf394-c7ff-4fba-9454-3b4a7b8e7184",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of pheno data: (2154, 14)\n",
      "Unique values in RECRUIT: ['PD' 'HC' 'SWEDD' 'PRODROMA' 'GENUN' 'REGPD' 'REGUN' 'GENPD']\n",
      "Unique values in YEARSEDUC: [18. 19. 16. 14. 13. 12. 20. 23. 17. 15. 21. 11. 24. 26. 25.  9. 10. 22.\n",
      "  8.  5. nan  3.  0.  7.  4.  6.  2. 30. 27.  1. 28.]\n",
      "Unique values in FAMILY_HISTORY: [ 0.  1. nan]\n",
      "Unique values in DX_INIT: ['iPD' 'NoDisease' 'ET' 'OtherD' 'Psychogenic' 'DLB' 'PrdrNM' 'PrdrM'\n",
      " 'Hemi_PKS' 'VaPS']\n",
      "null value in each column: \n",
      " PATNO              0\n",
      "RECRUIT            0\n",
      "FEMALE             0\n",
      "YEARSEDUC          1\n",
      "FAMILY_HISTORY    42\n",
      "DX_INIT            0\n",
      "dtype: int64\n",
      "   PATNO RECRUIT  FEMALE  YEARSEDUC  FAMILY_HISTORY    DX_INIT\n",
      "0   3400      PD       1       18.0             0.0        iPD\n",
      "1   3401      HC       1       19.0             0.0  NoDisease\n",
      "2   3402   SWEDD       0       16.0             0.0        iPD\n",
      "3   3403      PD       0       18.0             0.0        iPD\n",
      "4   3404      HC       1       14.0             0.0  NoDisease\n",
      "Value counts in DX_INIT: iPD            1040\n",
      "NoDisease       929\n",
      "PrdrNM          125\n",
      "OtherD           27\n",
      "ET               15\n",
      "PrdrM            11\n",
      "DLB               3\n",
      "Psychogenic       2\n",
      "Hemi_PKS          1\n",
      "VaPS              1\n",
      "Name: DX_INIT, dtype: int64\n",
      "Number of unique values in PATNO: 2154\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-11-bdd1b7868e1a>:29: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pheno_df['ID'] = 'PP-' + pheno_df['ID'].astype(str)\n",
      "<ipython-input-11-bdd1b7868e1a>:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pheno_df['EDUCATION'] = pheno_df['EDUCATION'].apply(lambda f: format(f, '.0f'))\n",
      "<ipython-input-11-bdd1b7868e1a>:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pheno_df['FAMILY_HISTORY'] = pheno_df['FAMILY_HISTORY'].apply(lambda f: format(f, '.0f'))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>FEMALE</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP-3400</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>iPD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP-3401</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>NoDisease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP-3402</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>iPD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PP-3403</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>iPD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PP-3404</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>NoDisease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2149</th>\n",
       "      <td>PP-40750</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>iPD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2150</th>\n",
       "      <td>PP-42381</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>NoDisease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2151</th>\n",
       "      <td>PP-41513</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>NoDisease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2152</th>\n",
       "      <td>PP-40767</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>iPD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153</th>\n",
       "      <td>PP-43086</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>iPD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2154 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID  FEMALE EDUCATION FAMILY_HISTORY      PHENO\n",
       "0      PP-3400       1        18              0        iPD\n",
       "1      PP-3401       1        19              0  NoDisease\n",
       "2      PP-3402       0        16              0        iPD\n",
       "3      PP-3403       0        18              0        iPD\n",
       "4      PP-3404       1        14              0  NoDisease\n",
       "...        ...     ...       ...            ...        ...\n",
       "2149  PP-40750       0        16              1        iPD\n",
       "2150  PP-42381       1        20              1  NoDisease\n",
       "2151  PP-41513       0        21              1  NoDisease\n",
       "2152  PP-40767       0         6              1        iPD\n",
       "2153  PP-43086       1        17              1        iPD\n",
       "\n",
       "[2154 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open pheno file DEMOG_DIAG.csv\n",
    "pheno = pd.read_csv(f\"{datadir}/DEMOG_DIAG.csv\", index_col=False )\n",
    "\n",
    "print('shape of pheno data:', pheno.shape)\n",
    "\n",
    "#grab PATNO, RECRUIT, FEMALE, YEARSEDUC, FAMILY_HISTORY, DX_INIT\n",
    "pheno_info = pheno[['PATNO','RECRUIT', 'FEMALE', 'YEARSEDUC', 'FAMILY_HISTORY', 'DX_INIT']]\n",
    "\n",
    "#check unique values from each column\n",
    "print('Unique values in RECRUIT:',pheno.RECRUIT.unique())\n",
    "print('Unique values in YEARSEDUC:',pheno.YEARSEDUC.unique())\n",
    "print('Unique values in FAMILY_HISTORY:',pheno.FAMILY_HISTORY.unique())\n",
    "print('Unique values in DX_INIT:',pheno.DX_INIT.unique())\n",
    "\n",
    "#check nan numbers in each column \n",
    "print('null value in each column: \\n', pheno_info.isna().sum())\n",
    "print(pheno_info.head())\n",
    "\n",
    "#count each value in PHENO column\n",
    "print('Value counts in DX_INIT:',pheno_info.DX_INIT.value_counts())\n",
    "print('Number of unique values in PATNO:', pheno_info.PATNO.nunique())\n",
    "\n",
    "#drop RECRUIT column\n",
    "pheno_df = pheno_info[['PATNO', 'FEMALE', 'YEARSEDUC', 'FAMILY_HISTORY', 'DX_INIT']]\n",
    "#rename columns\n",
    "pheno_df.columns = ['ID', 'FEMALE', 'EDUCATION','FAMILY_HISTORY', 'PHENO']\n",
    "\n",
    "#add pp. on ID\n",
    "pheno_df['ID'] = 'PP-' + pheno_df['ID'].astype(str)\n",
    "\n",
    "#remove .0 from EDUCATION and FAMILY HISTORY\n",
    "pheno_df['EDUCATION'] = pheno_df['EDUCATION'].apply(lambda f: format(f, '.0f'))\n",
    "pheno_df['FAMILY_HISTORY'] = pheno_df['FAMILY_HISTORY'].apply(lambda f: format(f, '.0f'))\n",
    "                 \n",
    "pheno_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "acefdc83-0c34-4030-ab80-e932e7bab389",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of PD in pheno data: 1040\n",
      "Number of Control in pheno data: 929\n",
      "shape of pheno_df: (1969, 5)\n",
      "        ID  FEMALE EDUCATION FAMILY_HISTORY      PHENO\n",
      "0  PP-3400       1        18              0        iPD\n",
      "1  PP-3401       1        19              0  NoDisease\n",
      "2  PP-3402       0        16              0        iPD\n",
      "3  PP-3403       0        18              0        iPD\n",
      "4  PP-3404       1        14              0  NoDisease\n",
      "Value counts in pheno data: iPD          1040\n",
      "NoDisease     929\n",
      "Name: PHENO, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#count PD and Control from Diagnosis\n",
    "PD_pheno = pheno_df[pheno_df['PHENO'].str.contains('iPD')]\n",
    "print(\"Number of PD in pheno data:\" , PD_pheno['PHENO'].count())\n",
    "\n",
    "Control_pheno = pheno_df[pheno_df['PHENO'].str.contains('NoDisease')]\n",
    "print(\"Number of Control in pheno data:\" , Control_pheno['PHENO'].count())\n",
    "\n",
    "#only keep the rows contain iPD and NoDisease in DX_INIT\n",
    "word_list = [\"iPD\", \"NoDisease\"]\n",
    "pheno_df = pheno_df[pheno_df['PHENO'].str.contains('|'.join(word_list))]\n",
    "pheno_df = pheno_df.reset_index(drop=True)\n",
    "print(\"shape of pheno_df:\", pheno_df.shape)\n",
    "print(pheno_df.head())\n",
    "\n",
    "#count each value in PHENO column\n",
    "print(\"Value counts in pheno data:\" , pheno_df.PHENO.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "820bca75-f36b-431b-a4ed-a3f6c48ad5e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of cov_case: (834, 5)\n",
      "        ID EDUCATION FAMILY_HISTORY MALE   AGE\n",
      "0  PP-3400        18              0    0  39.0\n",
      "1  PP-3402        16              0    1  46.0\n",
      "2  PP-3403        18              0    1  69.0\n",
      "3  PP-3406        18              0    1  35.0\n",
      "4  PP-3407        16              0    1  66.0\n",
      "shape of cov_cont: (626, 5)\n",
      "        ID EDUCATION FAMILY_HISTORY MALE   AGE\n",
      "0  PP-3401        19              0    0  56.0\n",
      "1  PP-3404        14              0    0  56.0\n",
      "2  PP-3405        18              0    0  63.0\n",
      "3  PP-3100        16              0    0  69.0\n",
      "4  PP-3151        13              0    1  58.0\n",
      "Value counts in pheno_id data: 1    834\n",
      "0    626\n",
      "Name: PHENO, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#merge pheno_df with cov\n",
    "cov = cov.drop(['PHENO'], axis=1)\n",
    "pheno_df = pheno_df.drop(['FEMALE'], axis=1)\n",
    "pheno_cov = pheno_df.merge(cov, on='ID', how='inner')\n",
    "\n",
    "#convert PD to 1, Control to 0\n",
    "pheno_id = pheno_cov.copy()\n",
    "pheno_id['PHENO'] = pheno_id['PHENO'].map({'NoDisease': 0, 'iPD': 1})\n",
    "\n",
    "#leave only PHENO=1 to only include only cases, PHENO=0 for controls\n",
    "cov_case = pheno_id.loc[pheno_id['PHENO'] == 1]\n",
    "cov_cont = pheno_id.loc[pheno_id['PHENO'] == 0]\n",
    "\n",
    "#drop Pheno column\n",
    "cov_case = cov_case.drop(['PHENO'], axis=1)\n",
    "cov_cont = cov_cont.drop(['PHENO'], axis=1)\n",
    "\n",
    "#reset index\n",
    "cov_case = cov_case.reset_index(drop=True)\n",
    "cov_cont = cov_cont.reset_index(drop=True)\n",
    "\n",
    "print(\"shape of cov_case:\", cov_case.shape)\n",
    "print(cov_case.head())\n",
    "\n",
    "print(\"shape of cov_cont:\", cov_cont.shape)\n",
    "print(cov_cont.head())\n",
    "\n",
    "#count each value in PHENO column\n",
    "print(\"Value counts in pheno_id data:\" , pheno_id.PHENO.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "696ed71a-c3ab-49fc-9ade-2f490425448a",
   "metadata": {},
   "source": [
    "Check plink files and make 10 genomic PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c501d6ce-0bb8-48c6-b278-2bc4c1e9ed25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2927\n",
      "BF-1001 BF-1001 0 0 0 -9\n",
      "BF-1002 BF-1002 0 0 0 -9\n",
      "BF-1003 BF-1003 0 0 0 -9\n",
      "BF-1004 BF-1004 0 0 0 -9\n",
      "BF-1005 BF-1005 0 0 0 -9\n",
      "BF-1006 BF-1006 0 0 0 -9\n",
      "BF-1008 BF-1008 0 0 0 -9\n",
      "BF-1009 BF-1009 0 0 0 -9\n",
      "BF-1010 BF-1010 0 0 0 -9\n",
      "BF-1011 BF-1011 0 0 0 -9\n",
      "305138\n",
      "1\trs200683566\t0\t13838\tT\tC\n",
      "1\trs370886505\t0\t14397\tC\tCTGT\n",
      "1\trs375086259\t0\t14653\tT\tC\n",
      "1\trs79585140\t0\t14907\tG\tA\n",
      "1\trs199856693\t0\t14933\tA\tG\n",
      "1\trs71252250\t0\t15118\tG\tA\n",
      "1\trs201635489\t0\t15447\tG\tA\n",
      "1\trs201026389\t0\t16125\tG\tT\n",
      "1\trs78588380\t0\t16257\tC\tG\n",
      "1\trs200736374\t0\t16288\tG\tC\n"
     ]
    }
   ],
   "source": [
    "#check number of IDs in original .fam file\n",
    "!cat {datadir}/hits_plus_tags_AMPPD_Euros.fam | wc -l\n",
    "#check .fam file\n",
    "!head /{datadir}/hits_plus_tags_AMPPD_Euros.fam\n",
    "\n",
    "#check number of IDs in original .bim file\n",
    "!cat {datadir}/hits_plus_tags_AMPPD_Euros.bim | wc -l\n",
    "#check .bim file\n",
    "!head /{datadir}/hits_plus_tags_AMPPD_Euros.bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93b7db1e-007f-40e3-ae2b-a5e6277cb44b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract genetic PCs from expression data \n",
    "!plink --bfile /{datadir}/hits_plus_tags_AMPPD_Euros --pca 10 --out /{datadir}/PCA_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f7bc9897-9868-4cf3-b1b7-a6fec3e41348",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>PC1_gen</th>\n",
       "      <th>PC2_gen</th>\n",
       "      <th>PC3_gen</th>\n",
       "      <th>PC4_gen</th>\n",
       "      <th>PC5_gen</th>\n",
       "      <th>PC6_gen</th>\n",
       "      <th>PC7_gen</th>\n",
       "      <th>PC8_gen</th>\n",
       "      <th>PC9_gen</th>\n",
       "      <th>PC10_gen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BF-1001</td>\n",
       "      <td>-0.013383</td>\n",
       "      <td>0.026178</td>\n",
       "      <td>-0.012396</td>\n",
       "      <td>0.004339</td>\n",
       "      <td>-0.005004</td>\n",
       "      <td>0.007554</td>\n",
       "      <td>-0.002626</td>\n",
       "      <td>0.022212</td>\n",
       "      <td>0.011273</td>\n",
       "      <td>-0.004516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BF-1002</td>\n",
       "      <td>-0.011969</td>\n",
       "      <td>0.014639</td>\n",
       "      <td>-0.026859</td>\n",
       "      <td>-0.000687</td>\n",
       "      <td>-0.003197</td>\n",
       "      <td>-0.005286</td>\n",
       "      <td>-0.000583</td>\n",
       "      <td>0.015597</td>\n",
       "      <td>0.002706</td>\n",
       "      <td>-0.007646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BF-1003</td>\n",
       "      <td>-0.010549</td>\n",
       "      <td>0.021005</td>\n",
       "      <td>0.014955</td>\n",
       "      <td>-0.000793</td>\n",
       "      <td>-0.001582</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.002457</td>\n",
       "      <td>-0.000282</td>\n",
       "      <td>-0.002610</td>\n",
       "      <td>0.001135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BF-1004</td>\n",
       "      <td>-0.012569</td>\n",
       "      <td>0.023211</td>\n",
       "      <td>-0.006541</td>\n",
       "      <td>0.003494</td>\n",
       "      <td>-0.011989</td>\n",
       "      <td>-0.000027</td>\n",
       "      <td>-0.004490</td>\n",
       "      <td>0.037969</td>\n",
       "      <td>0.016873</td>\n",
       "      <td>-0.008314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BF-1005</td>\n",
       "      <td>-0.010763</td>\n",
       "      <td>0.011671</td>\n",
       "      <td>-0.020893</td>\n",
       "      <td>-0.012810</td>\n",
       "      <td>-0.007227</td>\n",
       "      <td>-0.003492</td>\n",
       "      <td>-0.005223</td>\n",
       "      <td>-0.016002</td>\n",
       "      <td>-0.023424</td>\n",
       "      <td>0.000279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2922</th>\n",
       "      <td>PP-85242</td>\n",
       "      <td>-0.009681</td>\n",
       "      <td>-0.012745</td>\n",
       "      <td>0.022996</td>\n",
       "      <td>-0.000655</td>\n",
       "      <td>-0.001486</td>\n",
       "      <td>-0.002943</td>\n",
       "      <td>0.009491</td>\n",
       "      <td>-0.003265</td>\n",
       "      <td>0.019764</td>\n",
       "      <td>0.008159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2923</th>\n",
       "      <td>PP-90456</td>\n",
       "      <td>0.032165</td>\n",
       "      <td>0.001445</td>\n",
       "      <td>0.039731</td>\n",
       "      <td>-0.004689</td>\n",
       "      <td>-0.011589</td>\n",
       "      <td>0.001291</td>\n",
       "      <td>-0.004769</td>\n",
       "      <td>-0.004276</td>\n",
       "      <td>-0.007312</td>\n",
       "      <td>0.010096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2924</th>\n",
       "      <td>PP-91837</td>\n",
       "      <td>-0.012103</td>\n",
       "      <td>-0.001567</td>\n",
       "      <td>-0.020199</td>\n",
       "      <td>0.001134</td>\n",
       "      <td>-0.008514</td>\n",
       "      <td>-0.004052</td>\n",
       "      <td>-0.009519</td>\n",
       "      <td>0.014668</td>\n",
       "      <td>-0.027068</td>\n",
       "      <td>-0.003649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2925</th>\n",
       "      <td>PP-92490</td>\n",
       "      <td>-0.017216</td>\n",
       "      <td>0.001212</td>\n",
       "      <td>0.002467</td>\n",
       "      <td>-0.016545</td>\n",
       "      <td>-0.027899</td>\n",
       "      <td>-0.001444</td>\n",
       "      <td>-0.004236</td>\n",
       "      <td>-0.005391</td>\n",
       "      <td>-0.020077</td>\n",
       "      <td>0.000391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2926</th>\n",
       "      <td>PP-92834</td>\n",
       "      <td>-0.011086</td>\n",
       "      <td>-0.010031</td>\n",
       "      <td>0.019623</td>\n",
       "      <td>-0.004961</td>\n",
       "      <td>-0.003557</td>\n",
       "      <td>0.001757</td>\n",
       "      <td>-0.000242</td>\n",
       "      <td>0.007486</td>\n",
       "      <td>-0.012965</td>\n",
       "      <td>0.002003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2927 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID   PC1_gen   PC2_gen   PC3_gen   PC4_gen   PC5_gen   PC6_gen  \\\n",
       "0      BF-1001 -0.013383  0.026178 -0.012396  0.004339 -0.005004  0.007554   \n",
       "1      BF-1002 -0.011969  0.014639 -0.026859 -0.000687 -0.003197 -0.005286   \n",
       "2      BF-1003 -0.010549  0.021005  0.014955 -0.000793 -0.001582  0.008044   \n",
       "3      BF-1004 -0.012569  0.023211 -0.006541  0.003494 -0.011989 -0.000027   \n",
       "4      BF-1005 -0.010763  0.011671 -0.020893 -0.012810 -0.007227 -0.003492   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "2922  PP-85242 -0.009681 -0.012745  0.022996 -0.000655 -0.001486 -0.002943   \n",
       "2923  PP-90456  0.032165  0.001445  0.039731 -0.004689 -0.011589  0.001291   \n",
       "2924  PP-91837 -0.012103 -0.001567 -0.020199  0.001134 -0.008514 -0.004052   \n",
       "2925  PP-92490 -0.017216  0.001212  0.002467 -0.016545 -0.027899 -0.001444   \n",
       "2926  PP-92834 -0.011086 -0.010031  0.019623 -0.004961 -0.003557  0.001757   \n",
       "\n",
       "       PC7_gen   PC8_gen   PC9_gen  PC10_gen  \n",
       "0    -0.002626  0.022212  0.011273 -0.004516  \n",
       "1    -0.000583  0.015597  0.002706 -0.007646  \n",
       "2     0.002457 -0.000282 -0.002610  0.001135  \n",
       "3    -0.004490  0.037969  0.016873 -0.008314  \n",
       "4    -0.005223 -0.016002 -0.023424  0.000279  \n",
       "...        ...       ...       ...       ...  \n",
       "2922  0.009491 -0.003265  0.019764  0.008159  \n",
       "2923 -0.004769 -0.004276 -0.007312  0.010096  \n",
       "2924 -0.009519  0.014668 -0.027068 -0.003649  \n",
       "2925 -0.004236 -0.005391 -0.020077  0.000391  \n",
       "2926 -0.000242  0.007486 -0.012965  0.002003  \n",
       "\n",
       "[2927 rows x 11 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read genetic PCs\n",
    "pcs_df = pd.read_csv(f\"{datadir}/PCA_1.eigenvec\", sep=' ', usecols=[1,2,3,4,5,6,7,8,9,10,11], names=['ID','PC1_gen','PC2_gen','PC3_gen','PC4_gen','PC5_gen','PC6_gen','PC7_gen','PC8_gen','PC9_gen','PC10_gen'])\n",
    "pcs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ecb906bb-e4ad-4974-b5ff-b097f5b3df8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null value in each column in case: \n",
      " ID                  0\n",
      "EDUCATION           0\n",
      "FAMILY_HISTORY     13\n",
      "MALE                0\n",
      "AGE                 0\n",
      "PC1_gen           101\n",
      "PC2_gen           101\n",
      "PC3_gen           101\n",
      "PC4_gen           101\n",
      "PC5_gen           101\n",
      "PC6_gen           101\n",
      "PC7_gen           101\n",
      "PC8_gen           101\n",
      "PC9_gen           101\n",
      "PC10_gen          101\n",
      "dtype: int64\n",
      "null value in each column in control: \n",
      " ID                 0\n",
      "EDUCATION          0\n",
      "FAMILY_HISTORY     7\n",
      "MALE               0\n",
      "AGE                0\n",
      "PC1_gen           91\n",
      "PC2_gen           91\n",
      "PC3_gen           91\n",
      "PC4_gen           91\n",
      "PC5_gen           91\n",
      "PC6_gen           91\n",
      "PC7_gen           91\n",
      "PC8_gen           91\n",
      "PC9_gen           91\n",
      "PC10_gen          91\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#merge pcs_df into covariate_df\n",
    "covar_df_case = cov_case.merge(pcs_df, on='ID', how='left')\n",
    "covar_df_cont = cov_cont.merge(pcs_df, on='ID', how='left')\n",
    "\n",
    "#convert nan in FAMILY_HISTORY to null\n",
    "covar_df_case['FAMILY_HISTORY'] = covar_df_case['FAMILY_HISTORY'].replace('nan',np.nan)\n",
    "covar_df_cont['FAMILY_HISTORY'] = covar_df_cont['FAMILY_HISTORY'].replace('nan',np.nan)\n",
    "\n",
    "#check nan numbers in each column \n",
    "print('null value in each column in case: \\n', covar_df_case.isna().sum())\n",
    "print('null value in each column in control: \\n', covar_df_cont.isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "59f0a590-cbcf-437f-aca8-f1be32462b6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of covar_df_case (733, 16)\n",
      "       FID      IID EDUCATION FAMILY_HISTORY MALE   AGE   PC1_gen   PC2_gen  \\\n",
      "0  PP-3400  PP-3400        18              0    0  39.0 -0.004662 -0.023299   \n",
      "1  PP-3402  PP-3402        16              0    1  46.0  0.011903 -0.023524   \n",
      "2  PP-3403  PP-3403        18              0    1  69.0 -0.014213 -0.025636   \n",
      "3  PP-3406  PP-3406        18              0    1  35.0 -0.008313 -0.010611   \n",
      "4  PP-3407  PP-3407        16              0    1  66.0  0.033813 -0.012696   \n",
      "\n",
      "    PC3_gen   PC4_gen   PC5_gen   PC6_gen   PC7_gen   PC8_gen   PC9_gen  \\\n",
      "0  0.006330 -0.001153  0.003980 -0.002586 -0.000425 -0.007179 -0.020155   \n",
      "1 -0.001647  0.005045 -0.003863 -0.003192 -0.003624  0.000327  0.012216   \n",
      "2 -0.002261 -0.003854 -0.017373  0.002293 -0.002750  0.023918  0.020800   \n",
      "3  0.017661 -0.010679 -0.007702 -0.004243 -0.004153  0.002486  0.002382   \n",
      "4  0.020515 -0.000225 -0.008233 -0.003591  0.002511 -0.004819  0.002946   \n",
      "\n",
      "   PC10_gen  \n",
      "0 -0.000930  \n",
      "1 -0.006821  \n",
      "2 -0.003866  \n",
      "3  0.008385  \n",
      "4  0.009602  \n",
      "shape of covar_df_cont (535, 16)\n",
      "       FID      IID EDUCATION FAMILY_HISTORY MALE   AGE   PC1_gen   PC2_gen  \\\n",
      "0  PP-3401  PP-3401        19              0    0  56.0  0.034271 -0.023656   \n",
      "1  PP-3404  PP-3404        14              0    0  56.0 -0.011506 -0.020343   \n",
      "2  PP-3405  PP-3405        18              0    0  63.0 -0.011204 -0.014210   \n",
      "3  PP-3100  PP-3100        16              0    0  69.0 -0.012570 -0.010427   \n",
      "4  PP-3151  PP-3151        13              0    1  58.0 -0.011135 -0.016579   \n",
      "\n",
      "    PC3_gen   PC4_gen   PC5_gen   PC6_gen   PC7_gen   PC8_gen   PC9_gen  \\\n",
      "0  0.016260  0.000452 -0.005504 -0.000337 -0.000150  0.011953  0.014463   \n",
      "1  0.016984  0.003306 -0.011024 -0.004369 -0.006111  0.024201  0.000526   \n",
      "2 -0.000746 -0.004401 -0.019775 -0.003406 -0.012721  0.017960 -0.013703   \n",
      "3  0.008777  0.000524 -0.002120 -0.001913 -0.010488  0.003645 -0.015522   \n",
      "4  0.006779  0.005386 -0.012485 -0.005944 -0.001021  0.036534  0.016744   \n",
      "\n",
      "   PC10_gen  \n",
      "0 -0.001755  \n",
      "1  0.003789  \n",
      "2 -0.000250  \n",
      "3 -0.000490  \n",
      "4  0.002072  \n"
     ]
    }
   ],
   "source": [
    "#drop rows that has null values\n",
    "covar_df_case = covar_df_case[covar_df_case['PC1_gen'].notna()]\n",
    "covar_df_cont = covar_df_cont[covar_df_cont['PC1_gen'].notna()]\n",
    "\n",
    "#reset index\n",
    "covar_df_case = covar_df_case.reset_index(drop=True)\n",
    "covar_df_cont = covar_df_cont.reset_index(drop=True)\n",
    "\n",
    "#rename ID to FID\n",
    "covar_df_case = covar_df_case.rename(columns={'ID':'FID'})\n",
    "covar_df_cont = covar_df_cont.rename(columns={'ID':'FID'})\n",
    "\n",
    "#grab ID column and duplicate into two columns to compare with .fam file FID IID\n",
    "covar_df_case.insert(1, 'IID', covar_df_case['FID'])\n",
    "covar_df_cont.insert(1, 'IID', covar_df_cont['FID'])\n",
    "\n",
    "#replace NaN withmean for nonbinary and median for binary column\n",
    "covar_df_case['FAMILY_HISTORY'] = covar_df_case['FAMILY_HISTORY'].fillna(covar_df_case['FAMILY_HISTORY'].median())\n",
    "covar_df_cont['FAMILY_HISTORY'] = covar_df_cont['FAMILY_HISTORY'].fillna(covar_df_cont['FAMILY_HISTORY'].median())\n",
    "\n",
    "#make ID as list to trim expression data later\n",
    "covar_df_case_list = covar_df_case['FID']\n",
    "covar_df_cont_list = covar_df_cont['FID']\n",
    "\n",
    "print('shape of covar_df_case', covar_df_case.shape)\n",
    "print(covar_df_case.head())\n",
    "print('shape of covar_df_cont', covar_df_cont.shape)\n",
    "print(covar_df_cont.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e92181c7-d9c0-4ed2-a6e1-0b14d48d8414",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique value of EDUCATION: [18 16 13 14 12 23 17 21 24 19 26 20 15 10  8 11 22  9  5 25  3  7  6  0\n",
      "  2  1]\n",
      "Data types of EDUCATION: int64\n",
      "Unique value of FAMILY_HISTORY: [0 1]\n",
      "Data types of FAMILY_HISTORY: int64\n",
      "Unique value of MALE: [0 1]\n",
      "Data types of MALE: int64\n",
      "Unique value of AGE: [39 46 69 35 66 58 52 64 50 72 70 68 55 74 67 75 60 73 57 48 62 56 80 83\n",
      " 63 59 65 54 41 76 51 45 40 71 47 78 38 61 53 49 77 85 79 34 43 42 44 37\n",
      " 82 88 81 94 87 33 92 84 90]\n",
      "Data types of AGE: int64\n",
      "Unique value of EDUCATION: [19 14 18 16 13 20 12 17 15  9 22 21 11 23  8 24  7  2 30 25 10 26 27  6]\n",
      "Data types of EDUCATION: int64\n",
      "Unique value of FAMILY_HISTORY: [0 1]\n",
      "Data types of FAMILY_HISTORY: int64\n",
      "Unique value of MALE: [0 1]\n",
      "Data types of MALE: int64\n",
      "Unique value of AGE: [56 63 69 58 52 75 65 32 41 60 70 72 76 53 38 80 79 77 45 68 54 49 50 61\n",
      " 71 82 42 47 31 59 64 62 74 46 66 73 55 43 44 48 33 57 78 67 83 81 51 36\n",
      " 27 35 29 30 25 37 39 34 87 40 84 22 19 26]\n",
      "Data types of AGE: int64\n",
      "shape of covariate data for case: (733, 16)\n",
      "       FID      IID  EDUCATION  FAMILY_HISTORY  MALE       AGE   PC1_gen  \\\n",
      "0  PP-3400  PP-3400   0.692308               0     0  0.098361 -0.004662   \n",
      "1  PP-3402  PP-3402   0.615385               0     1  0.213115  0.011903   \n",
      "2  PP-3403  PP-3403   0.692308               0     1  0.590164 -0.014213   \n",
      "3  PP-3406  PP-3406   0.692308               0     1  0.032787 -0.008313   \n",
      "4  PP-3407  PP-3407   0.615385               0     1  0.540984  0.033813   \n",
      "\n",
      "    PC2_gen   PC3_gen   PC4_gen   PC5_gen   PC6_gen   PC7_gen   PC8_gen  \\\n",
      "0 -0.023299  0.006330 -0.001153  0.003980 -0.002586 -0.000425 -0.007179   \n",
      "1 -0.023524 -0.001647  0.005045 -0.003863 -0.003192 -0.003624  0.000327   \n",
      "2 -0.025636 -0.002261 -0.003854 -0.017373  0.002293 -0.002750  0.023918   \n",
      "3 -0.010611  0.017661 -0.010679 -0.007702 -0.004243 -0.004153  0.002486   \n",
      "4 -0.012696  0.020515 -0.000225 -0.008233 -0.003591  0.002511 -0.004819   \n",
      "\n",
      "    PC9_gen  PC10_gen  \n",
      "0 -0.020155 -0.000930  \n",
      "1  0.012216 -0.006821  \n",
      "2  0.020800 -0.003866  \n",
      "3  0.002382  0.008385  \n",
      "4  0.002946  0.009602  \n",
      "shape of covariate data for control: (535, 16)\n",
      "       FID      IID  EDUCATION  FAMILY_HISTORY  MALE       AGE   PC1_gen  \\\n",
      "0  PP-3401  PP-3401   0.607143               0     0  0.544118  0.034271   \n",
      "1  PP-3404  PP-3404   0.428571               0     0  0.544118 -0.011506   \n",
      "2  PP-3405  PP-3405   0.571429               0     0  0.647059 -0.011204   \n",
      "3  PP-3100  PP-3100   0.500000               0     0  0.735294 -0.012570   \n",
      "4  PP-3151  PP-3151   0.392857               0     1  0.573529 -0.011135   \n",
      "\n",
      "    PC2_gen   PC3_gen   PC4_gen   PC5_gen   PC6_gen   PC7_gen   PC8_gen  \\\n",
      "0 -0.023656  0.016260  0.000452 -0.005504 -0.000337 -0.000150  0.011953   \n",
      "1 -0.020343  0.016984  0.003306 -0.011024 -0.004369 -0.006111  0.024201   \n",
      "2 -0.014210 -0.000746 -0.004401 -0.019775 -0.003406 -0.012721  0.017960   \n",
      "3 -0.010427  0.008777  0.000524 -0.002120 -0.001913 -0.010488  0.003645   \n",
      "4 -0.016579  0.006779  0.005386 -0.012485 -0.005944 -0.001021  0.036534   \n",
      "\n",
      "    PC9_gen  PC10_gen  \n",
      "0  0.014463 -0.001755  \n",
      "1  0.000526  0.003789  \n",
      "2 -0.013703 -0.000250  \n",
      "3 -0.015522 -0.000490  \n",
      "4  0.016744  0.002072  \n"
     ]
    }
   ],
   "source": [
    "#check unique values and data type in each column\n",
    "for i in covar_df_case.columns[2:6]:\n",
    "    print('Unique value of ' + i + ':', covar_df_case[i].unique())\n",
    "    print('Data types of '  + i + ':',covar_df_case[i].dtypes)\n",
    "\n",
    "for i in covar_df_cont.columns[2:6]:\n",
    "    print('Unique value of ' + i + ':', covar_df_cont[i].unique())\n",
    "    print('Data types of '  + i + ':',covar_df_cont[i].dtypes)\n",
    "    \n",
    "#change float to int in column EDUCATION, FAMILY_HISTORY, and UPSIT\n",
    "\n",
    "##col=(covariate.dtypes=='float64')\n",
    "col_cont=covar_df_case.columns[2:6]\n",
    "covar_df_case.loc[:,col_cont]=covar_df_case.loc[:,col_cont].astype(np.int64)\n",
    "covar_df_cont.loc[:,col_cont]=covar_df_cont.loc[:,col_cont].astype(np.int64)\n",
    "\n",
    "#normalize covariate values for AGE, EDUCATION columns with min_max_scaling\n",
    "\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "scaler = MinMaxScaler()\n",
    "covar_df_case_scale = covar_df_case.copy()\n",
    "covar_df_case_scale[['AGE', 'EDUCATION']] = scaler.fit_transform(covar_df_case_scale[['AGE', 'EDUCATION']])\n",
    "\n",
    "covar_df_cont_scale = covar_df_cont.copy()\n",
    "covar_df_cont_scale[['AGE', 'EDUCATION']] = scaler.fit_transform(covar_df_cont_scale[['AGE', 'EDUCATION']])\n",
    "\n",
    "# apply the min-max scaling in Pandas using the .min() and .max() methods\n",
    "#def min_max_scaling(df):\n",
    "    # copy the dataframe\n",
    " #   df_norm = df.copy()\n",
    "    # apply min-max scaling\n",
    "  #  for column in df_norm.iloc[:, [2, 5]]:\n",
    "   #     df_norm[column] = (df_norm[column] - df_norm[column].min()) / (df_norm[column].max() - df_norm[column].min()) \n",
    "        \n",
    "    #return df_norm\n",
    "    \n",
    "# call the min_max_scaling function\n",
    "#covar_df_case_scale = min_max_scaling(covar_df_case)\n",
    "print('shape of covariate data for case:', covar_df_case_scale.shape)\n",
    "print(covar_df_case_scale.head())\n",
    "\n",
    "#covar_df_cont_scale = min_max_scaling(covar_df_cont)\n",
    "print('shape of covariate data for control:', covar_df_cont_scale.shape)\n",
    "print(covar_df_cont_scale.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1687ab4e-e0ba-4919-920f-c8ec8a468af7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save covariate as txt --> this is used to trim expression data\n",
    "covar_df_case_list.to_csv(r'./data_folder/patient_id_case.txt', sep='\\t' ,index=False)\n",
    "covar_df_cont_list.to_csv(r'./data_folder/patient_id_control.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3840646a-077e-410b-9ffa-3a4df92168d5",
   "metadata": {},
   "source": [
    "Coordinate data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c5652018-b4e7-410f-bced-3d12a5898ab1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of gene location data: (2535323, 4)\n",
      "null value in each column: \n",
      " X.Chr    0\n",
      "start    0\n",
      "end      0\n",
      "ID       0\n",
      "dtype: int64\n",
      "unique chromosom: ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chrM', 'chrX', 'chrY', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrUn_GL000195v1', 'chrUn_GL000213v1', 'chrUn_GL000216v2', 'chrUn_GL000218v1', 'chrUn_GL000219v1', 'chrUn_GL000220v1', 'chrUn_KI270442v1', 'chrUn_KI270744v1', 'chrUn_KI270750v1', 'chr1_GL383518v1_alt', 'chr1_GL383519v1_alt', 'chr1_GL383520v2_alt', 'chr1_KI270759v1_alt', 'chr1_KI270760v1_alt', 'chr1_KI270761v1_alt', 'chr1_KI270762v1_alt', 'chr1_KI270763v1_alt', 'chr1_KI270765v1_alt', 'chr1_KI270766v1_alt', 'chr1_KI270892v1_alt', 'chr1_KN196472v1_fix', 'chr1_KN196473v1_fix', 'chr1_KN196474v1_fix', 'chr1_KN538360v1_fix', 'chr1_KN538361v1_fix', 'chr1_KQ031383v1_fix', 'chr1_KQ458382v1_alt', 'chr1_KQ458383v1_alt', 'chr1_KQ458384v1_alt', 'chr1_KQ983255v1_alt', 'chr1_KV880763v1_alt', 'chr1_KZ208904v1_alt', 'chr1_KZ208905v1_alt', 'chr1_KZ208906v1_fix', 'chr2_GL383521v1_alt', 'chr2_GL383522v1_alt', 'chr2_GL582966v2_alt', 'chr2_KI270767v1_alt', 'chr2_KI270768v1_alt', 'chr2_KI270769v1_alt', 'chr2_KI270770v1_alt', 'chr2_KI270772v1_alt', 'chr2_KI270773v1_alt', 'chr2_KI270774v1_alt', 'chr2_KI270776v1_alt', 'chr2_KI270893v1_alt', 'chr2_KI270894v1_alt', 'chr2_KN538362v1_fix', 'chr2_KN538363v1_fix', 'chr2_KQ031384v1_fix', 'chr2_KQ983256v1_alt', 'chr2_KZ208907v1_alt', 'chr2_KZ208908v1_alt', 'chr3_GL383526v1_alt', 'chr3_JH636055v2_alt', 'chr3_KI270777v1_alt', 'chr3_KI270779v1_alt', 'chr3_KI270780v1_alt', 'chr3_KI270781v1_alt', 'chr3_KI270782v1_alt', 'chr3_KI270783v1_alt', 'chr3_KI270784v1_alt', 'chr3_KI270895v1_alt', 'chr3_KI270924v1_alt', 'chr3_KI270934v1_alt', 'chr3_KI270935v1_alt', 'chr3_KI270936v1_alt', 'chr3_KI270937v1_alt', 'chr3_KN196475v1_fix', 'chr3_KN196476v1_fix', 'chr3_KN538364v1_fix', 'chr3_KQ031385v1_fix', 'chr3_KV766192v1_fix', 'chr3_KZ208909v1_alt', 'chr3_KZ559102v1_alt', 'chr3_KZ559103v1_alt', 'chr3_KZ559104v1_fix', 'chr3_KZ559105v1_alt', 'chr4_GL000257v2_alt', 'chr4_GL383527v1_alt', 'chr4_GL383528v1_alt', 'chr4_KI270785v1_alt', 'chr4_KI270786v1_alt', 'chr4_KI270788v1_alt', 'chr4_KI270789v1_alt', 'chr4_KI270790v1_alt', 'chr4_KI270896v1_alt', 'chr4_KI270925v1_alt', 'chr4_KQ090014v1_alt', 'chr4_KQ090015v1_alt', 'chr4_KQ983257v1_fix', 'chr4_KQ983258v1_alt', 'chr4_KV766193v1_alt', 'chr5_GL339449v2_alt', 'chr5_GL383531v1_alt', 'chr5_GL383532v1_alt', 'chr5_GL949742v1_alt', 'chr5_KI270791v1_alt', 'chr5_KI270792v1_alt', 'chr5_KI270793v1_alt', 'chr5_KI270794v1_alt', 'chr5_KI270795v1_alt', 'chr5_KI270796v1_alt', 'chr5_KI270897v1_alt', 'chr5_KI270898v1_alt', 'chr5_KN196477v1_alt', 'chr5_KV575243v1_alt', 'chr5_KV575244v1_fix', 'chr6_GL000250v2_alt', 'chr6_GL000251v2_alt', 'chr6_GL000252v2_alt', 'chr6_GL000253v2_alt', 'chr6_GL000254v2_alt', 'chr6_GL000255v2_alt', 'chr6_GL000256v2_alt', 'chr6_GL383533v1_alt', 'chr6_KB021644v2_alt', 'chr6_KI270758v1_alt', 'chr6_KI270797v1_alt', 'chr6_KI270798v1_alt', 'chr6_KI270799v1_alt', 'chr6_KI270800v1_alt', 'chr6_KI270801v1_alt', 'chr6_KI270802v1_alt', 'chr6_KN196478v1_fix', 'chr6_KQ031387v1_fix', 'chr6_KQ090016v1_fix', 'chr6_KZ208911v1_fix', 'chr7_GL383534v2_alt', 'chr7_KI270803v1_alt', 'chr7_KI270804v1_alt', 'chr7_KI270805v1_alt', 'chr7_KI270806v1_alt', 'chr7_KI270807v1_alt', 'chr7_KI270808v1_alt', 'chr7_KI270809v1_alt', 'chr7_KI270899v1_alt', 'chr7_KQ031388v1_fix', 'chr7_KV880764v1_fix', 'chr7_KV880765v1_fix', 'chr7_KZ208912v1_fix', 'chr7_KZ208913v1_alt', 'chr7_KZ559106v1_alt', 'chr8_KI270810v1_alt', 'chr8_KI270811v1_alt', 'chr8_KI270812v1_alt', 'chr8_KI270813v1_alt', 'chr8_KI270814v1_alt', 'chr8_KI270815v1_alt', 'chr8_KI270816v1_alt', 'chr8_KI270817v1_alt', 'chr8_KI270818v1_alt', 'chr8_KI270819v1_alt', 'chr8_KI270821v1_alt', 'chr8_KI270822v1_alt', 'chr8_KI270900v1_alt', 'chr8_KI270901v1_alt', 'chr8_KI270926v1_alt', 'chr8_KV880766v1_fix', 'chr8_KZ208914v1_fix', 'chr8_KZ208915v1_fix', 'chr8_KZ559107v1_alt', 'chr9_GL383539v1_alt', 'chr9_GL383540v1_alt', 'chr9_GL383541v1_alt', 'chr9_GL383542v1_alt', 'chr9_KI270823v1_alt', 'chr9_KN196479v1_fix', 'chr9_KQ090018v1_alt', 'chrX_KI270880v1_alt', 'chrX_KI270881v1_alt', 'chrX_KI270913v1_alt', 'chrY_KN196487v1_fix', 'chrY_KZ208924v1_fix', 'chr10_GL383545v1_alt', 'chr10_GL383546v1_alt', 'chr10_KI270824v1_alt', 'chr10_KI270825v1_alt', 'chr10_KN196480v1_fix', 'chr10_KQ090021v1_fix', 'chr11_JH159136v1_alt', 'chr11_JH159137v1_alt', 'chr11_KI270829v1_alt', 'chr11_KI270830v1_alt', 'chr11_KI270831v1_alt', 'chr11_KI270832v1_alt', 'chr11_KI270902v1_alt', 'chr11_KI270903v1_alt', 'chr11_KI270927v1_alt', 'chr11_KN196481v1_fix', 'chr11_KN538368v1_alt', 'chr11_KQ090022v1_fix', 'chr11_KQ759759v1_fix', 'chr11_KV766195v1_fix', 'chr11_KZ559108v1_fix', 'chr11_KZ559109v1_fix', 'chr11_KZ559110v1_alt', 'chr11_KZ559111v1_alt', 'chr12_GL383550v2_alt', 'chr12_GL383551v1_alt', 'chr12_GL383552v1_alt', 'chr12_GL383553v2_alt', 'chr12_GL877875v1_alt', 'chr12_GL877876v1_alt', 'chr12_KI270833v1_alt', 'chr12_KI270834v1_alt', 'chr12_KI270835v1_alt', 'chr12_KI270837v1_alt', 'chr12_KI270904v1_alt', 'chr12_KN538369v1_fix', 'chr12_KN538370v1_fix', 'chr12_KQ759760v1_fix', 'chr12_KZ208916v1_fix', 'chr12_KZ208917v1_fix', 'chr12_KZ559112v1_alt', 'chr13_KI270838v1_alt', 'chr13_KI270840v1_alt', 'chr13_KI270842v1_alt', 'chr13_KN538371v1_fix', 'chr13_KN538372v1_fix', 'chr13_KN538373v1_fix', 'chr14_KI270844v1_alt', 'chr14_KI270845v1_alt', 'chr14_KI270846v1_alt', 'chr14_KI270847v1_alt', 'chr14_KZ208919v1_alt', 'chr14_KZ208920v1_fix', 'chr15_GL383554v1_alt', 'chr15_GL383555v2_alt', 'chr15_KI270848v1_alt', 'chr15_KI270849v1_alt', 'chr15_KI270850v1_alt', 'chr15_KI270851v1_alt', 'chr15_KI270852v1_alt', 'chr15_KI270905v1_alt', 'chr15_KI270906v1_alt', 'chr15_KQ031389v1_alt', 'chr16_GL383556v1_alt', 'chr16_GL383557v1_alt', 'chr16_KI270853v1_alt', 'chr16_KI270854v1_alt', 'chr16_KI270855v1_alt', 'chr16_KI270856v1_alt', 'chr16_KQ090026v1_alt', 'chr16_KQ090027v1_alt', 'chr16_KV880768v1_fix', 'chr16_KZ208921v1_alt', 'chr16_KZ559113v1_fix', 'chr17_GL000258v2_alt', 'chr17_GL383563v3_alt', 'chr17_GL383564v2_alt', 'chr17_GL383565v1_alt', 'chr17_GL383566v1_alt', 'chr17_JH159146v1_alt', 'chr17_JH159147v1_alt', 'chr17_JH159148v1_alt', 'chr17_KI270857v1_alt', 'chr17_KI270858v1_alt', 'chr17_KI270859v1_alt', 'chr17_KI270860v1_alt', 'chr17_KI270861v1_alt', 'chr17_KI270862v1_alt', 'chr17_KI270907v1_alt', 'chr17_KI270908v1_alt', 'chr17_KI270909v1_alt', 'chr17_KI270910v1_alt', 'chr17_KV575245v1_fix', 'chr17_KV766196v1_fix', 'chr17_KV766198v1_alt', 'chr17_KZ559114v1_alt', 'chr18_GL383567v1_alt', 'chr18_GL383569v1_alt', 'chr18_GL383570v1_alt', 'chr18_GL383571v1_alt', 'chr18_GL383572v1_alt', 'chr18_KI270863v1_alt', 'chr18_KI270911v1_alt', 'chr18_KI270912v1_alt', 'chr18_KQ090028v1_fix', 'chr18_KQ458385v1_alt', 'chr18_KZ208922v1_fix', 'chr18_KZ559115v1_fix', 'chr18_KZ559116v1_alt', 'chr19_GL000209v2_alt', 'chr19_GL383573v1_alt', 'chr19_GL383574v1_alt', 'chr19_GL383575v2_alt', 'chr19_GL383576v1_alt', 'chr19_GL949746v1_alt', 'chr19_GL949747v2_alt', 'chr19_GL949748v2_alt', 'chr19_GL949749v2_alt', 'chr19_GL949750v2_alt', 'chr19_GL949751v2_alt', 'chr19_GL949752v1_alt', 'chr19_GL949753v2_alt', 'chr19_KI270865v1_alt', 'chr19_KI270866v1_alt', 'chr19_KI270867v1_alt', 'chr19_KI270868v1_alt', 'chr19_KI270882v1_alt', 'chr19_KI270883v1_alt', 'chr19_KI270884v1_alt', 'chr19_KI270885v1_alt', 'chr19_KI270886v1_alt', 'chr19_KI270887v1_alt', 'chr19_KI270888v1_alt', 'chr19_KI270889v1_alt', 'chr19_KI270890v1_alt', 'chr19_KI270891v1_alt', 'chr19_KI270914v1_alt', 'chr19_KI270915v1_alt', 'chr19_KI270916v1_alt', 'chr19_KI270917v1_alt', 'chr19_KI270918v1_alt', 'chr19_KI270919v1_alt', 'chr19_KI270920v1_alt', 'chr19_KI270921v1_alt', 'chr19_KI270922v1_alt', 'chr19_KI270923v1_alt', 'chr19_KI270929v1_alt', 'chr19_KI270930v1_alt', 'chr19_KI270931v1_alt', 'chr19_KI270932v1_alt', 'chr19_KI270933v1_alt', 'chr19_KI270938v1_alt', 'chr19_KN196484v1_fix', 'chr19_KQ458386v1_fix', 'chr19_KV575246v1_alt', 'chr19_KV575247v1_alt', 'chr19_KV575248v1_alt', 'chr19_KV575249v1_alt', 'chr19_KV575250v1_alt', 'chr19_KV575251v1_alt', 'chr19_KV575252v1_alt', 'chr19_KV575253v1_alt', 'chr19_KV575254v1_alt', 'chr19_KV575255v1_alt', 'chr19_KV575256v1_alt', 'chr19_KV575257v1_alt', 'chr19_KV575258v1_alt', 'chr19_KV575259v1_alt', 'chr19_KV575260v1_alt', 'chr20_GL383577v2_alt', 'chr20_KI270869v1_alt', 'chr20_KI270870v1_alt', 'chr20_KI270871v1_alt', 'chr21_GL383579v2_alt', 'chr21_GL383580v2_alt', 'chr21_GL383581v2_alt', 'chr21_KI270872v1_alt', 'chr21_KI270873v1_alt', 'chr21_KI270874v1_alt', 'chr22_GL383582v2_alt', 'chr22_GL383583v2_alt', 'chr22_KB663609v1_alt', 'chr22_KI270875v1_alt', 'chr22_KI270876v1_alt', 'chr22_KI270877v1_alt', 'chr22_KI270878v1_alt', 'chr22_KI270879v1_alt', 'chr22_KI270928v1_alt', 'chr22_KN196485v1_alt', 'chr22_KN196486v1_alt', 'chr22_KQ458387v1_alt', 'chr22_KQ458388v1_alt', 'chr22_KQ759761v1_alt', 'chr22_KQ759762v1_fix', 'chr1_KI270711v1_random', 'chr1_KI270713v1_random', 'chr11_KI270721v1_random', 'chr14_GL000009v2_random', 'chr14_GL000194v1_random', 'chr14_GL000225v1_random', 'chr14_KI270726v1_random', 'chr15_KI270727v1_random', 'chr16_KI270728v1_random', 'chr17_GL000205v2_random', 'chr22_KI270731v1_random', 'chr22_KI270733v1_random', 'chr22_KI270734v1_random']\n"
     ]
    }
   ],
   "source": [
    "#open te location file hg38_rmsk_TE.gtf.locInd.locations\n",
    "chr_pos_df = pd.read_csv(f\"{datadir}/hg38_gene_annotation.gtf.locations\", sep='\\t', header=None, names=('X.Chr','start','end','ID'))\n",
    "\n",
    "print('shape of gene location data:', chr_pos_df.shape)\n",
    "\n",
    "#remove gene_id, duplicated ENST number, and \"\",;\n",
    "chr_pos_df['ID'] = chr_pos_df.ID.str.extract(r'gene_id .*?\"([^\"]+)\";.+')\n",
    "\n",
    "#check nan numbers in each column \n",
    "print('null value in each column: \\n', chr_pos_df.isna().sum())\n",
    "\n",
    "#checke chromosome inputs in chr_pos_df\n",
    "unique_chromosom = chr_pos_df['X.Chr'].unique().tolist()\n",
    "print('unique chromosom:', unique_chromosom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fe1f91f5-7f53-4fa3-abb3-dc1952da06af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of M chromosom in coordinate data: 76\n",
      "Number of Y chromosom in coordinate data: 7634\n",
      "Number of X chromosom in coordinate data: 74651\n",
      "Number of Unknown chromosom in coordinate data: 82\n",
      "shape of coord_df: (2452880, 4)\n",
      "  X.Chr  start    end                 ID\n",
      "0  chr1  11869  12227  ENST00000456328.2\n",
      "1  chr1  12613  12721  ENST00000456328.2\n",
      "2  chr1  13221  14409  ENST00000456328.2\n",
      "3  chr1  12010  12057  ENST00000450305.2\n",
      "4  chr1  12179  12227  ENST00000450305.2\n",
      "Number of unique ID in coord_df: 239497\n",
      "Number count of ID in coord_df:                      index  size\n",
      "0       ENST00000000233.10    14\n",
      "1        ENST00000000412.8    15\n",
      "2       ENST00000000442.11    15\n",
      "3        ENST00000001008.6    22\n",
      "4        ENST00000001146.6    14\n",
      "...                    ...   ...\n",
      "239492   ENST00000673604.1    20\n",
      "239493   ENST00000673615.1    37\n",
      "239494   ENST00000673616.1    16\n",
      "239495   ENST00000673620.1     9\n",
      "239496   ENST00000673621.1    35\n",
      "\n",
      "[239497 rows x 2 columns]\n",
      "number of transcript ID that has no duplicate:  23860\n"
     ]
    }
   ],
   "source": [
    "#count how many M are in coord\n",
    "m_coords = chr_pos_df[chr_pos_df['X.Chr'].str.contains('M')]\n",
    "print(\"Number of M chromosom in coordinate data:\" , m_coords['X.Chr'].count())\n",
    "\n",
    "#count how many y are in coord\n",
    "y_coords = chr_pos_df[chr_pos_df['X.Chr'].str.contains('Y')]\n",
    "print(\"Number of Y chromosom in coordinate data:\" , y_coords['X.Chr'].count())\n",
    "\n",
    "#count how many X are in coord\n",
    "x_coords = chr_pos_df[chr_pos_df['X.Chr'].str.contains('X')]\n",
    "print(\"Number of X chromosom in coordinate data:\" , x_coords['X.Chr'].count())\n",
    "\n",
    "#count how many Un are in coord\n",
    "un_coords = chr_pos_df[chr_pos_df['X.Chr'].str.contains('Un')]\n",
    "print(\"Number of Unknown chromosom in coordinate data:\" , un_coords['X.Chr'].count())\n",
    "\n",
    "#drop all the rows contain M, X, Y in X.Chr \n",
    "word_list = [\"M\", \"X\", \"Y\", \"Un\"]\n",
    "coord_df = chr_pos_df[~chr_pos_df['X.Chr'].str.contains('|'.join(word_list))]\n",
    "coord_df = coord_df.reset_index(drop=True)\n",
    "print(\"shape of coord_df:\", coord_df.shape)\n",
    "print(coord_df.head())\n",
    "\n",
    "#check how many ID are unique\n",
    "print(\"Number of unique ID in coord_df:\", coord_df['ID'].nunique())\n",
    "\n",
    "#count how many times patient number is duplicated\n",
    "duplicate = coord_df.groupby(coord_df['ID'].tolist(),as_index=False).size()\n",
    "print(\"Number count of ID in coord_df:\", duplicate)\n",
    "\n",
    "#check how many IDs are only count as 1\n",
    "duplicate_1 = duplicate.query(\"size==1\")\n",
    "print('number of transcript ID that has no duplicate: ', len(duplicate_1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "40eede1d-e704-4b6f-b669-408ebfaad5c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of ENST00000673620.1 in coord data:  9\n",
      "number of ENST00000000233.10 in coord data:  14\n",
      "                        ID  X.Chr\n",
      "0       ENST00000000233.10      1\n",
      "1        ENST00000000412.8      1\n",
      "2       ENST00000000442.11      1\n",
      "3        ENST00000001008.6      1\n",
      "4        ENST00000001146.6      1\n",
      "...                    ...    ...\n",
      "239492   ENST00000673604.1      1\n",
      "239493   ENST00000673615.1      1\n",
      "239494   ENST00000673616.1      1\n",
      "239495   ENST00000673620.1      1\n",
      "239496   ENST00000673621.1      1\n",
      "\n",
      "[239497 rows x 2 columns]\n",
      "number of chromosomes that only have 1 kind in each ID:  239497\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>127588411</td>\n",
       "      <td>127591700</td>\n",
       "      <td>ENST00000000233.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>8940361</td>\n",
       "      <td>8949645</td>\n",
       "      <td>ENST00000000412.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>64305524</td>\n",
       "      <td>64316743</td>\n",
       "      <td>ENST00000000442.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>2794970</td>\n",
       "      <td>2805423</td>\n",
       "      <td>ENST00000001008.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>72129238</td>\n",
       "      <td>72148038</td>\n",
       "      <td>ENST00000001146.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239492</th>\n",
       "      <td>chr11</td>\n",
       "      <td>47242846</td>\n",
       "      <td>47248840</td>\n",
       "      <td>ENST00000673604.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239493</th>\n",
       "      <td>chr8</td>\n",
       "      <td>132063785</td>\n",
       "      <td>132111159</td>\n",
       "      <td>ENST00000673615.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239494</th>\n",
       "      <td>chr19_KV575251v1_alt</td>\n",
       "      <td>105193</td>\n",
       "      <td>116143</td>\n",
       "      <td>ENST00000673616.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239495</th>\n",
       "      <td>chr3</td>\n",
       "      <td>15601726</td>\n",
       "      <td>15672955</td>\n",
       "      <td>ENST00000673620.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239496</th>\n",
       "      <td>chr3</td>\n",
       "      <td>42013802</td>\n",
       "      <td>42223737</td>\n",
       "      <td>ENST00000673621.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>239497 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       X.Chr      start        end                  ID\n",
       "0                       chr7  127588411  127591700  ENST00000000233.10\n",
       "1                      chr12    8940361    8949645   ENST00000000412.8\n",
       "2                      chr11   64305524   64316743  ENST00000000442.11\n",
       "3                      chr12    2794970    2805423   ENST00000001008.6\n",
       "4                       chr2   72129238   72148038   ENST00000001146.6\n",
       "...                      ...        ...        ...                 ...\n",
       "239492                 chr11   47242846   47248840   ENST00000673604.1\n",
       "239493                  chr8  132063785  132111159   ENST00000673615.1\n",
       "239494  chr19_KV575251v1_alt     105193     116143   ENST00000673616.1\n",
       "239495                  chr3   15601726   15672955   ENST00000673620.1\n",
       "239496                  chr3   42013802   42223737   ENST00000673621.1\n",
       "\n",
       "[239497 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print a couple IDs that have duplicated to see what is going on\n",
    "exp_coord_1 = coord_df[coord_df['ID'].str.contains('ENST00000673620.1')]\n",
    "print('number of ENST00000673620.1 in coord data: ', exp_coord_1.shape[0])\n",
    "\n",
    "#print a couple IDs that have duplicated to see what is going on\n",
    "exp_coord_2 = coord_df[coord_df['ID'].str.contains('ENST00000000233.10')]\n",
    "print('number of ENST00000000233.10 in coord data: ', exp_coord_2.shape[0])\n",
    "\n",
    "#count unique values in X.Chr in same ID\n",
    "grouped_df = coord_df.groupby('ID')\n",
    "grouped_df = grouped_df.agg({\"X.Chr\": \"nunique\"})\n",
    "grouped_df = grouped_df.reset_index()\n",
    "print(grouped_df)\n",
    "#check how many have 1 kind of X.Chr\n",
    "print('number of chromosomes that only have 1 kind in each ID: ', len(grouped_df[grouped_df['X.Chr'] == 1]))\n",
    "##safe to assume duplicated IDs are in same Chromosome\n",
    "\n",
    "#merge duplicated ID into 1 and take smallest start and biggest end\n",
    "coord_comb = coord_df.groupby(['ID']).agg({'start': 'min', 'end': 'max', 'X.Chr': 'first'}).reset_index()\n",
    "\n",
    "#function to move columns\n",
    "def movecol(df, cols_to_move=[], ref_col='', place='After'):\n",
    "    \n",
    "    cols = df.columns.tolist()\n",
    "    if place == 'After':\n",
    "        seg1 = cols[:list(cols).index(ref_col) + 1]\n",
    "        seg2 = cols_to_move\n",
    "    if place == 'Before':\n",
    "        seg1 = cols[:list(cols).index(ref_col)]\n",
    "        seg2 = cols_to_move + [ref_col]\n",
    "    \n",
    "    seg1 = [i for i in seg1 if i not in seg2]\n",
    "    seg3 = [i for i in cols if i not in seg1 + seg2]\n",
    "    \n",
    "    return(df[seg1 + seg2 + seg3])\n",
    "\n",
    "#rearrange Gene_Symbol, Chr, Start, Stop columns right nextg to index column \n",
    "coord_comb = movecol(coord_comb, cols_to_move=['X.Chr'], ref_col='ID', place='Before')\n",
    "coord_comb = movecol(coord_comb, cols_to_move=['ID'], ref_col='end', place='After')\n",
    "coord_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b9f5d11b-6350-4c7e-9dfe-76e6327d805c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of coordinate data greater than 1 mega: (170, 4)\n",
      "        X.Chr      start        end                  ID\n",
      "630     chr20   13995369   16053197   ENST00000217246.8\n",
      "4519    chr16   82626965   83795376  ENST00000268613.14\n",
      "5436     chr4   92303966   93774566   ENST00000282020.9\n",
      "6839     chr8  112222928  113436939  ENST00000297405.10\n",
      "11717    chr2  213284464  214410501  ENST00000331683.10\n",
      "...       ...        ...        ...                 ...\n",
      "218934   chr8   31639304   32728077   ENST00000651149.1\n",
      "219107   chr8   31640659   32855666   ENST00000651335.1\n",
      "220377   chr8   31639245   32774046   ENST00000652698.1\n",
      "223686   chr6   97305577   98398863   ENST00000656098.1\n",
      "236274   chr6   97305553   98398783   ENST00000669069.1\n",
      "\n",
      "[170 rows x 4 columns]\n",
      "shape of coordinate data greater than 2 mega: (12, 4)\n",
      "        X.Chr      start        end                 ID\n",
      "630     chr20   13995369   16053197  ENST00000217246.8\n",
      "17695    chr7  146116801  148420998  ENST00000361727.8\n",
      "24177   chr11   83457919   85627270  ENST00000376104.6\n",
      "33377    chr8    2938099    4994701  ENST00000400186.7\n",
      "72357    chr9    8528152   10612723  ENST00000463477.5\n",
      "117975   chr8    2938099    4994972  ENST00000520002.5\n",
      "179947   chr8    2938099    4994972  ENST00000602557.5\n",
      "180071   chr8    2938099    4994701  ENST00000602723.5\n",
      "204624   chr8    2935361    4994914  ENST00000635120.2\n",
      "209938  chr16    5239802    7711458  ENST00000641259.1\n",
      "211308  chr20   13995659   16050041  ENST00000642719.1\n",
      "218452  chr11   83457919   85627270  ENST00000650630.1\n"
     ]
    }
   ],
   "source": [
    "#check if any ID have end-start > 1000000\n",
    "\n",
    "coord_comb_1 = coord_comb[((coord_comb['end'] - coord_comb['start']) > 1000000)]\n",
    "print('shape of coordinate data greater than 1 mega:', coord_comb_1.shape)\n",
    "print(coord_comb_1)\n",
    "\n",
    "coord_comb_2 = coord_comb[((coord_comb['end'] - coord_comb['start']) > 2000000)]\n",
    "print('shape of coordinate data greater than 2 mega:', coord_comb_2.shape)\n",
    "print(coord_comb_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ee832900-d059-44df-8469-d1599d9220c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>90916390</td>\n",
       "      <td>91022014</td>\n",
       "      <td>ENST00000467231.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>15456732</td>\n",
       "      <td>15472091</td>\n",
       "      <td>ENST00000359621.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>21233570</td>\n",
       "      <td>21238350</td>\n",
       "      <td>ENST00000470394.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>54032019</td>\n",
       "      <td>54036802</td>\n",
       "      <td>ENST00000470395.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>151266308</td>\n",
       "      <td>151266725</td>\n",
       "      <td>ENST00000470396.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239492</th>\n",
       "      <td>22</td>\n",
       "      <td>30555938</td>\n",
       "      <td>30560759</td>\n",
       "      <td>ENST00000447224.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239493</th>\n",
       "      <td>22</td>\n",
       "      <td>44196098</td>\n",
       "      <td>44219533</td>\n",
       "      <td>ENST00000472551.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239494</th>\n",
       "      <td>22</td>\n",
       "      <td>24632915</td>\n",
       "      <td>24647760</td>\n",
       "      <td>ENST00000447261.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239495</th>\n",
       "      <td>22</td>\n",
       "      <td>20967248</td>\n",
       "      <td>20974096</td>\n",
       "      <td>ENST00000426113.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239496</th>\n",
       "      <td>22</td>\n",
       "      <td>20995089</td>\n",
       "      <td>20999032</td>\n",
       "      <td>ENST00000463909.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>239497 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        X.Chr      start        end                 ID\n",
       "0           1   90916390   91022014  ENST00000467231.5\n",
       "1           1   15456732   15472091  ENST00000359621.5\n",
       "2           1   21233570   21238350  ENST00000470394.1\n",
       "3           1   54032019   54036802  ENST00000470395.1\n",
       "4           1  151266308  151266725  ENST00000470396.1\n",
       "...       ...        ...        ...                ...\n",
       "239492     22   30555938   30560759  ENST00000447224.5\n",
       "239493     22   44196098   44219533  ENST00000472551.1\n",
       "239494     22   24632915   24647760  ENST00000447261.6\n",
       "239495     22   20967248   20974096  ENST00000426113.5\n",
       "239496     22   20995089   20999032  ENST00000463909.1\n",
       "\n",
       "[239497 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#seperate any character after chr number in X.Chr and put it in a new column\n",
    "coord_comb['Chr_info'] = coord_comb['X.Chr']\n",
    "coord_comb['X.Chr'] = coord_comb['X.Chr'].str.split('_').str[0]\n",
    "\n",
    "#seperate information in Chr_info into info_1 and info_2\n",
    "coord_comb['info_2'] = coord_comb['Chr_info']\n",
    "coord_comb['info_2'] = coord_comb['Chr_info'].str.split('_').str[2]\n",
    "coord_comb['Chr_info'] = coord_comb['Chr_info'].str.split('_').str[1]\n",
    "coord_comb = coord_comb.rename(columns={'Chr_info': 'info_1'})\n",
    "\n",
    "#delete chr from X.Chr\n",
    "coord_comb['X.Chr'] = coord_comb['X.Chr'].str.replace('chr', '')\n",
    "\n",
    "#convert X.Chr to int\n",
    "coord_comb['X.Chr'] = coord_comb['X.Chr'].astype(np.int64)\n",
    "\n",
    "#arrange by assending order of X.Chr\n",
    "coord_comb = coord_comb.sort_values(by = ['X.Chr'], ascending=True).reset_index(drop=True)\n",
    "\n",
    "#delete info_1, info_2\n",
    "coord_comb_3 = coord_comb.drop(['info_1', 'info_2'], axis=1)\n",
    "coord_comb_3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8129c042-5124-47a7-b7bf-aad9b50fd92b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save coord_df as txt -- run only once for download\n",
    "coord_comb.to_csv(r'./data_folder/transcript_coordinate_long.txt', sep='\\t' ,index=False)\n",
    "\n",
    "#save coord_df as txt -- run only once for download\n",
    "coord_comb_3.to_csv(r'./data_folder/transcript_coordinate_short.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "23c5b061-8f34-47ba-9025-ee6b6608fb81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes in gene_list: 239497\n"
     ]
    }
   ],
   "source": [
    "#get ID as a list and save to compare with dataset\n",
    "gene_list = coord_comb_3['ID']\n",
    "print(\"number of genes in gene_list:\", len(gene_list))\n",
    "\n",
    "#create gene_list as list to drop columns in exp_case \n",
    "gene_list_1 = coord_comb_3['ID'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05ad369e-4e51-469e-a327-63b7acf7cc8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save gene_list as txt -- run only once for download\n",
    "gene_list.to_csv(r'./data_folder/transcript_id_list.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4c4f128-ac9f-482a-8c7c-acf3ed74bcd9",
   "metadata": {},
   "source": [
    "Expression data (baseline & transcript ID only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50568f2b-6dc2-41f0-bc4b-ea6a810d511f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#split data into 4 in equal size to open easily\n",
    "!split -l $(( $( wc -l < ./data_folder/baseline_GENE_TE_norm_adjusted.cntTable ) / 4 + 1 )) ./data_folder/baseline_GENE_TE_norm_adjusted.cntTable ./data_folder/baseline_GENE_TE_norm_adjusted.cntTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f628b009-0a6e-47b2-8f84-8af4b28c0d30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1042229\n",
      "1042229\n",
      "1042229\n",
      "1042225\n"
     ]
    }
   ],
   "source": [
    "#check the number of rows \n",
    "!cat ./data_folder/baseline_GENE_TE_norm_adjusted.cntTableaa | wc -l\n",
    "!cat ./data_folder/baseline_GENE_TE_norm_adjusted.cntTableab | wc -l\n",
    "!cat ./data_folder/baseline_GENE_TE_norm_adjusted.cntTableac | wc -l\n",
    "!cat ./data_folder/baseline_GENE_TE_norm_adjusted.cntTablead | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58a71eed-64e2-458a-aa56-07c29794951d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read column names (header) (patient number, which is our sample ID) and output as text file\n",
    "!head -1 ./data_folder/baseline_GENE_TE_norm_adjusted.cntTable > ./data_folder/baseline_id.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d7bd9729-ffe4-413c-9179-735a9e98eb67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1565"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read baseline dataset header\n",
    "base = pd.read_csv(f\"{datadir}/baseline_id.txt\", engine='c', sep= '\\t')\n",
    "base_header = base.columns.tolist()\n",
    "len(base_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "92f14116-8a20-4a93-992d-6eabc0b1c906",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read index and output as text file\n",
    "!awk '{print $1}' ./data_folder/baseline_GENE_TE_norm_adjusted.cntTableaa > ./data_folder/aa_trans.txt\n",
    "!awk '{print $1}' ./data_folder/baseline_GENE_TE_norm_adjusted.cntTableab > ./data_folder/ab_trans.txt\n",
    "!awk '{print $1}' ./data_folder/baseline_GENE_TE_norm_adjusted.cntTableac > ./data_folder/ac_trans.txt\n",
    "!awk '{print $1}' ./data_folder/baseline_GENE_TE_norm_adjusted.cntTablead > ./data_folder/ad_trans.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bc06115b-1e77-43c2-9122-b81ff89f1be6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of ENST in Tableaa expression data:  227021\n",
      "number of ENST in Tableab expression data:  0\n",
      "number of ENST in Tableac expression data:  0\n",
      "number of ENST in Tablead expression data:  0\n"
     ]
    }
   ],
   "source": [
    "#read baseline_2 data\n",
    "aa_trans = pd.read_csv(f\"{datadir}/aa_trans.txt\", header=None, names=['ID'], engine='c',sep='\\t')\n",
    "ab_trans = pd.read_csv(f\"{datadir}/ab_trans.txt\", header=None, names=['ID'], engine='c',sep='\\t')\n",
    "ac_trans = pd.read_csv(f\"{datadir}/ac_trans.txt\", header=None, names=['ID'], engine='c',sep='\\t')\n",
    "ad_trans = pd.read_csv(f\"{datadir}/ad_trans.txt\", header=None, names=['ID'], engine='c',sep='\\t')\n",
    "\n",
    "#print a couple IDs that have duplicated to see what is going on\n",
    "aa_trans_enst = aa_trans[aa_trans['ID'].str.contains('ENST')]\n",
    "print('number of ENST in Tableaa expression data: ', aa_trans_enst.shape[0])\n",
    "\n",
    "#print a couple IDs that have duplicated to see what is going on\n",
    "ab_trans_enst = ab_trans[ab_trans['ID'].str.contains('ENST')]\n",
    "print('number of ENST in Tableab expression data: ', ab_trans_enst.shape[0])\n",
    "\n",
    "#print a couple IDs that have duplicated to see what is going on\n",
    "ac_trans_enst = ac_trans[ac_trans['ID'].str.contains('ENST')]\n",
    "print('number of ENST in Tableac expression data: ', ac_trans_enst.shape[0])\n",
    "\n",
    "#print a couple IDs that have duplicated to see what is going on\n",
    "ad_trans_enst = ad_trans[ad_trans['ID'].str.contains('ENST')]\n",
    "print('number of ENST in Tablead expression data: ', ad_trans_enst.shape[0])\n",
    "\n",
    "##will process only Tableaa since all ENST values are in there"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ace44174-cc67-4fb9-b0e3-cae3ec1858a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of case ID:  733\n",
      "Number of control ID:  535\n",
      "Number of transcript ID:  227021\n",
      "Number of transcript ID that has coordinate data:  239497\n"
     ]
    }
   ],
   "source": [
    "#read patient_id\n",
    "case_id = pd.read_csv(f\"{datadir}/patient_id_case.txt\",  engine='c',sep='\\t')\n",
    "case_id_list = case_id['FID'].tolist()\n",
    "print('Number of case ID: ', len(case_id))\n",
    "\n",
    "#read patient_id\n",
    "cont_id = pd.read_csv(f\"{datadir}/patient_id_control.txt\",  engine='c',sep='\\t')\n",
    "cont_id_list = cont_id['FID'].tolist()\n",
    "print('Number of control ID: ', len(cont_id))\n",
    "\n",
    "#read transcript_id_list\n",
    "aa_trans_enst_list = aa_trans_enst['ID'].tolist()\n",
    "print('Number of transcript ID: ', len(aa_trans_enst_list))\n",
    "\n",
    "#read transcript_id_list \n",
    "tr = pd.read_csv(f\"{datadir}/transcript_id_list.txt\", sep='\\t')\n",
    "print('Number of transcript ID that has coordinate data: ', len(tr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "27a7610d-5134-4f66-9d0c-014eac86325d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-27-f656285d6144>:8: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  base_aa.columns = base_aa.columns.str.replace('pp.', 'PP-')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PP-10874</th>\n",
       "      <th>PP-12224</th>\n",
       "      <th>PP-12499</th>\n",
       "      <th>PP-12593</th>\n",
       "      <th>PP-13039</th>\n",
       "      <th>PP-13424</th>\n",
       "      <th>PP-14281</th>\n",
       "      <th>PP-14331</th>\n",
       "      <th>PP-14426</th>\n",
       "      <th>PP-15761</th>\n",
       "      <th>...</th>\n",
       "      <th>PP-65008</th>\n",
       "      <th>PP-70239</th>\n",
       "      <th>PP-85062</th>\n",
       "      <th>PP-85236</th>\n",
       "      <th>PP-85242</th>\n",
       "      <th>PP-90456</th>\n",
       "      <th>PP-91097</th>\n",
       "      <th>PP-91837</th>\n",
       "      <th>PP-92490</th>\n",
       "      <th>PP-92834</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000000233.10</th>\n",
       "      <td>1.055457</td>\n",
       "      <td>-0.090322</td>\n",
       "      <td>0.489696</td>\n",
       "      <td>0.108862</td>\n",
       "      <td>1.284662</td>\n",
       "      <td>0.916743</td>\n",
       "      <td>0.528411</td>\n",
       "      <td>0.982363</td>\n",
       "      <td>1.079028</td>\n",
       "      <td>0.541646</td>\n",
       "      <td>...</td>\n",
       "      <td>0.493132</td>\n",
       "      <td>0.888485</td>\n",
       "      <td>0.828482</td>\n",
       "      <td>1.165689</td>\n",
       "      <td>1.536568</td>\n",
       "      <td>0.840764</td>\n",
       "      <td>1.066142</td>\n",
       "      <td>-0.793131</td>\n",
       "      <td>1.626177</td>\n",
       "      <td>1.623303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000000412.8</th>\n",
       "      <td>3.701636</td>\n",
       "      <td>2.136299</td>\n",
       "      <td>2.517864</td>\n",
       "      <td>2.635514</td>\n",
       "      <td>3.526311</td>\n",
       "      <td>2.515658</td>\n",
       "      <td>2.373977</td>\n",
       "      <td>3.547481</td>\n",
       "      <td>3.384911</td>\n",
       "      <td>2.996052</td>\n",
       "      <td>...</td>\n",
       "      <td>1.673927</td>\n",
       "      <td>2.566919</td>\n",
       "      <td>3.439492</td>\n",
       "      <td>3.236078</td>\n",
       "      <td>2.080992</td>\n",
       "      <td>2.773280</td>\n",
       "      <td>2.492843</td>\n",
       "      <td>0.083720</td>\n",
       "      <td>3.180079</td>\n",
       "      <td>3.450967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000000442.11</th>\n",
       "      <td>3.143852</td>\n",
       "      <td>1.563251</td>\n",
       "      <td>1.977397</td>\n",
       "      <td>1.849058</td>\n",
       "      <td>2.940845</td>\n",
       "      <td>2.039643</td>\n",
       "      <td>2.220080</td>\n",
       "      <td>2.683409</td>\n",
       "      <td>2.783043</td>\n",
       "      <td>2.126609</td>\n",
       "      <td>...</td>\n",
       "      <td>1.379228</td>\n",
       "      <td>2.192926</td>\n",
       "      <td>2.389890</td>\n",
       "      <td>2.277158</td>\n",
       "      <td>2.153871</td>\n",
       "      <td>2.351454</td>\n",
       "      <td>2.361598</td>\n",
       "      <td>0.228564</td>\n",
       "      <td>2.785307</td>\n",
       "      <td>3.587059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000001008.6</th>\n",
       "      <td>1.884540</td>\n",
       "      <td>1.494641</td>\n",
       "      <td>1.580101</td>\n",
       "      <td>1.475380</td>\n",
       "      <td>1.637678</td>\n",
       "      <td>1.773128</td>\n",
       "      <td>1.673841</td>\n",
       "      <td>2.092372</td>\n",
       "      <td>2.035153</td>\n",
       "      <td>1.568978</td>\n",
       "      <td>...</td>\n",
       "      <td>1.070522</td>\n",
       "      <td>1.783352</td>\n",
       "      <td>1.980777</td>\n",
       "      <td>1.742746</td>\n",
       "      <td>2.263322</td>\n",
       "      <td>1.626536</td>\n",
       "      <td>1.694174</td>\n",
       "      <td>0.436736</td>\n",
       "      <td>1.396854</td>\n",
       "      <td>2.136774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000001146.6</th>\n",
       "      <td>-1.545447</td>\n",
       "      <td>-0.430481</td>\n",
       "      <td>-6.233680</td>\n",
       "      <td>-1.583514</td>\n",
       "      <td>-1.749059</td>\n",
       "      <td>-4.335397</td>\n",
       "      <td>-1.708629</td>\n",
       "      <td>-1.778449</td>\n",
       "      <td>-6.682523</td>\n",
       "      <td>-0.996774</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.529708</td>\n",
       "      <td>-4.749588</td>\n",
       "      <td>-3.660453</td>\n",
       "      <td>-7.086976</td>\n",
       "      <td>-7.096427</td>\n",
       "      <td>-1.501998</td>\n",
       "      <td>-2.698144</td>\n",
       "      <td>-0.199452</td>\n",
       "      <td>-6.663842</td>\n",
       "      <td>-3.825158</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1565 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    PP-10874  PP-12224  PP-12499  PP-12593  PP-13039  \\\n",
       "ENST00000000233.10  1.055457 -0.090322  0.489696  0.108862  1.284662   \n",
       "ENST00000000412.8   3.701636  2.136299  2.517864  2.635514  3.526311   \n",
       "ENST00000000442.11  3.143852  1.563251  1.977397  1.849058  2.940845   \n",
       "ENST00000001008.6   1.884540  1.494641  1.580101  1.475380  1.637678   \n",
       "ENST00000001146.6  -1.545447 -0.430481 -6.233680 -1.583514 -1.749059   \n",
       "\n",
       "                    PP-13424  PP-14281  PP-14331  PP-14426  PP-15761  ...  \\\n",
       "ENST00000000233.10  0.916743  0.528411  0.982363  1.079028  0.541646  ...   \n",
       "ENST00000000412.8   2.515658  2.373977  3.547481  3.384911  2.996052  ...   \n",
       "ENST00000000442.11  2.039643  2.220080  2.683409  2.783043  2.126609  ...   \n",
       "ENST00000001008.6   1.773128  1.673841  2.092372  2.035153  1.568978  ...   \n",
       "ENST00000001146.6  -4.335397 -1.708629 -1.778449 -6.682523 -0.996774  ...   \n",
       "\n",
       "                    PP-65008  PP-70239  PP-85062  PP-85236  PP-85242  \\\n",
       "ENST00000000233.10  0.493132  0.888485  0.828482  1.165689  1.536568   \n",
       "ENST00000000412.8   1.673927  2.566919  3.439492  3.236078  2.080992   \n",
       "ENST00000000442.11  1.379228  2.192926  2.389890  2.277158  2.153871   \n",
       "ENST00000001008.6   1.070522  1.783352  1.980777  1.742746  2.263322   \n",
       "ENST00000001146.6  -1.529708 -4.749588 -3.660453 -7.086976 -7.096427   \n",
       "\n",
       "                    PP-90456  PP-91097  PP-91837  PP-92490  PP-92834  \n",
       "ENST00000000233.10  0.840764  1.066142 -0.793131  1.626177  1.623303  \n",
       "ENST00000000412.8   2.773280  2.492843  0.083720  3.180079  3.450967  \n",
       "ENST00000000442.11  2.351454  2.361598  0.228564  2.785307  3.587059  \n",
       "ENST00000001008.6   1.626536  1.694174  0.436736  1.396854  2.136774  \n",
       "ENST00000001146.6  -1.501998 -2.698144 -0.199452 -6.663842 -3.825158  \n",
       "\n",
       "[5 rows x 1565 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read baseline_1 data\n",
    "base_aa = pd.read_csv(f\"{datadir}/baseline_GENE_TE_norm_adjusted.cntTableaa\", engine='c',sep='\\t')\n",
    "\n",
    "#round the decimal numbers to 6 digits\n",
    "base_aa = base_aa.round(decimals=6)\n",
    "\n",
    "#change ID format as PP-\n",
    "base_aa.columns = base_aa.columns.str.replace('pp.', 'PP-')\n",
    "\n",
    "base_aa.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "12bf20fe-89e0-486d-8476-4e4a8a19a425",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>ENST00000000233.10</th>\n",
       "      <th>ENST00000000412.8</th>\n",
       "      <th>ENST00000000442.11</th>\n",
       "      <th>ENST00000001008.6</th>\n",
       "      <th>ENST00000001146.6</th>\n",
       "      <th>ENST00000002125.9</th>\n",
       "      <th>ENST00000002165.11</th>\n",
       "      <th>ENST00000002501.10</th>\n",
       "      <th>...</th>\n",
       "      <th>ENST00000673594.1</th>\n",
       "      <th>ENST00000673597.1</th>\n",
       "      <th>ENST00000673598.1</th>\n",
       "      <th>ENST00000673599.1</th>\n",
       "      <th>ENST00000673600.1</th>\n",
       "      <th>ENST00000673603.1</th>\n",
       "      <th>ENST00000673604.1</th>\n",
       "      <th>ENST00000673615.1</th>\n",
       "      <th>ENST00000673620.1</th>\n",
       "      <th>ENST00000673621.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP-3001</td>\n",
       "      <td>PP-3001</td>\n",
       "      <td>1.231180</td>\n",
       "      <td>3.278386</td>\n",
       "      <td>2.645443</td>\n",
       "      <td>2.145071</td>\n",
       "      <td>-6.633006</td>\n",
       "      <td>0.959451</td>\n",
       "      <td>2.498851</td>\n",
       "      <td>-4.311078</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.385079</td>\n",
       "      <td>-6.633006</td>\n",
       "      <td>-2.726115</td>\n",
       "      <td>-0.800116</td>\n",
       "      <td>3.760384</td>\n",
       "      <td>-3.463081</td>\n",
       "      <td>-2.932566</td>\n",
       "      <td>-6.633006</td>\n",
       "      <td>-0.190063</td>\n",
       "      <td>0.135178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP-3002</td>\n",
       "      <td>PP-3002</td>\n",
       "      <td>0.112015</td>\n",
       "      <td>2.722850</td>\n",
       "      <td>1.680395</td>\n",
       "      <td>1.477788</td>\n",
       "      <td>-0.373412</td>\n",
       "      <td>0.282856</td>\n",
       "      <td>1.920547</td>\n",
       "      <td>-1.912931</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.695340</td>\n",
       "      <td>-3.879764</td>\n",
       "      <td>-3.638756</td>\n",
       "      <td>-1.557836</td>\n",
       "      <td>3.278272</td>\n",
       "      <td>-3.251733</td>\n",
       "      <td>-3.879764</td>\n",
       "      <td>-1.981644</td>\n",
       "      <td>-0.544780</td>\n",
       "      <td>-0.305773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP-3003</td>\n",
       "      <td>PP-3003</td>\n",
       "      <td>0.996263</td>\n",
       "      <td>3.426735</td>\n",
       "      <td>2.620935</td>\n",
       "      <td>1.814124</td>\n",
       "      <td>-3.051490</td>\n",
       "      <td>0.715676</td>\n",
       "      <td>2.425490</td>\n",
       "      <td>-3.582005</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.228368</td>\n",
       "      <td>-6.751930</td>\n",
       "      <td>-3.051490</td>\n",
       "      <td>-0.869287</td>\n",
       "      <td>3.744924</td>\n",
       "      <td>-2.228368</td>\n",
       "      <td>-3.582005</td>\n",
       "      <td>-5.166967</td>\n",
       "      <td>-0.122573</td>\n",
       "      <td>-0.122573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PP-3006</td>\n",
       "      <td>PP-3006</td>\n",
       "      <td>0.570453</td>\n",
       "      <td>3.279767</td>\n",
       "      <td>2.629038</td>\n",
       "      <td>1.801636</td>\n",
       "      <td>-1.932047</td>\n",
       "      <td>0.744658</td>\n",
       "      <td>2.283302</td>\n",
       "      <td>-3.806516</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.452880</td>\n",
       "      <td>-6.976441</td>\n",
       "      <td>-2.888979</td>\n",
       "      <td>-0.999162</td>\n",
       "      <td>3.769233</td>\n",
       "      <td>-3.806516</td>\n",
       "      <td>-3.276002</td>\n",
       "      <td>-6.976441</td>\n",
       "      <td>-0.406586</td>\n",
       "      <td>-0.318230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PP-3007</td>\n",
       "      <td>PP-3007</td>\n",
       "      <td>0.560225</td>\n",
       "      <td>3.055598</td>\n",
       "      <td>2.140143</td>\n",
       "      <td>1.910574</td>\n",
       "      <td>-5.432242</td>\n",
       "      <td>0.744347</td>\n",
       "      <td>2.293409</td>\n",
       "      <td>-3.557772</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.373348</td>\n",
       "      <td>-7.017204</td>\n",
       "      <td>-2.929741</td>\n",
       "      <td>-0.908680</td>\n",
       "      <td>3.428845</td>\n",
       "      <td>-2.373348</td>\n",
       "      <td>-3.110313</td>\n",
       "      <td>-5.432242</td>\n",
       "      <td>-0.275737</td>\n",
       "      <td>0.201964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>PP-58510</td>\n",
       "      <td>PP-58510</td>\n",
       "      <td>1.036279</td>\n",
       "      <td>3.110402</td>\n",
       "      <td>2.481665</td>\n",
       "      <td>1.684186</td>\n",
       "      <td>-7.748356</td>\n",
       "      <td>0.715169</td>\n",
       "      <td>2.050926</td>\n",
       "      <td>-3.356038</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.020435</td>\n",
       "      <td>-7.748356</td>\n",
       "      <td>-2.890375</td>\n",
       "      <td>-0.805841</td>\n",
       "      <td>3.717721</td>\n",
       "      <td>-4.047916</td>\n",
       "      <td>-3.104500</td>\n",
       "      <td>-7.748356</td>\n",
       "      <td>0.152511</td>\n",
       "      <td>0.290563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>PP-58858</td>\n",
       "      <td>PP-58858</td>\n",
       "      <td>0.842236</td>\n",
       "      <td>3.181759</td>\n",
       "      <td>2.535167</td>\n",
       "      <td>1.638167</td>\n",
       "      <td>-3.047060</td>\n",
       "      <td>0.877213</td>\n",
       "      <td>2.379205</td>\n",
       "      <td>-3.336567</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.114174</td>\n",
       "      <td>-6.506492</td>\n",
       "      <td>-3.047060</td>\n",
       "      <td>-0.891782</td>\n",
       "      <td>4.249230</td>\n",
       "      <td>-4.921529</td>\n",
       "      <td>-3.047060</td>\n",
       "      <td>-6.506492</td>\n",
       "      <td>-0.030758</td>\n",
       "      <td>0.313687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>PP-58951</td>\n",
       "      <td>PP-58951</td>\n",
       "      <td>0.904115</td>\n",
       "      <td>2.548815</td>\n",
       "      <td>2.325297</td>\n",
       "      <td>1.230259</td>\n",
       "      <td>-6.775365</td>\n",
       "      <td>0.143498</td>\n",
       "      <td>1.551064</td>\n",
       "      <td>-3.315934</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.383048</td>\n",
       "      <td>-6.775365</td>\n",
       "      <td>-3.315934</td>\n",
       "      <td>-0.942475</td>\n",
       "      <td>3.518106</td>\n",
       "      <td>-3.605440</td>\n",
       "      <td>-3.968010</td>\n",
       "      <td>-5.190403</td>\n",
       "      <td>-0.007181</td>\n",
       "      <td>0.537518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>PP-59398</td>\n",
       "      <td>PP-59398</td>\n",
       "      <td>0.956323</td>\n",
       "      <td>2.863001</td>\n",
       "      <td>2.282723</td>\n",
       "      <td>1.468014</td>\n",
       "      <td>-6.275698</td>\n",
       "      <td>0.363341</td>\n",
       "      <td>1.755888</td>\n",
       "      <td>-4.401229</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.978017</td>\n",
       "      <td>-7.860660</td>\n",
       "      <td>-3.105773</td>\n",
       "      <td>-0.965843</td>\n",
       "      <td>3.265399</td>\n",
       "      <td>-2.368807</td>\n",
       "      <td>-3.216804</td>\n",
       "      <td>-7.860660</td>\n",
       "      <td>0.232097</td>\n",
       "      <td>0.569792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>PP-70239</td>\n",
       "      <td>PP-70239</td>\n",
       "      <td>0.888485</td>\n",
       "      <td>2.566919</td>\n",
       "      <td>2.192926</td>\n",
       "      <td>1.783352</td>\n",
       "      <td>-4.749588</td>\n",
       "      <td>0.147652</td>\n",
       "      <td>1.350548</td>\n",
       "      <td>-2.679199</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.547955</td>\n",
       "      <td>-7.071516</td>\n",
       "      <td>-3.371077</td>\n",
       "      <td>-1.188873</td>\n",
       "      <td>3.205771</td>\n",
       "      <td>-3.901591</td>\n",
       "      <td>-3.612085</td>\n",
       "      <td>-7.071516</td>\n",
       "      <td>0.329363</td>\n",
       "      <td>0.635843</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>705 rows × 219245 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          FID       IID  ENST00000000233.10  ENST00000000412.8  \\\n",
       "0     PP-3001   PP-3001            1.231180           3.278386   \n",
       "1     PP-3002   PP-3002            0.112015           2.722850   \n",
       "2     PP-3003   PP-3003            0.996263           3.426735   \n",
       "3     PP-3006   PP-3006            0.570453           3.279767   \n",
       "4     PP-3007   PP-3007            0.560225           3.055598   \n",
       "..        ...       ...                 ...                ...   \n",
       "700  PP-58510  PP-58510            1.036279           3.110402   \n",
       "701  PP-58858  PP-58858            0.842236           3.181759   \n",
       "702  PP-58951  PP-58951            0.904115           2.548815   \n",
       "703  PP-59398  PP-59398            0.956323           2.863001   \n",
       "704  PP-70239  PP-70239            0.888485           2.566919   \n",
       "\n",
       "     ENST00000000442.11  ENST00000001008.6  ENST00000001146.6  \\\n",
       "0              2.645443           2.145071          -6.633006   \n",
       "1              1.680395           1.477788          -0.373412   \n",
       "2              2.620935           1.814124          -3.051490   \n",
       "3              2.629038           1.801636          -1.932047   \n",
       "4              2.140143           1.910574          -5.432242   \n",
       "..                  ...                ...                ...   \n",
       "700            2.481665           1.684186          -7.748356   \n",
       "701            2.535167           1.638167          -3.047060   \n",
       "702            2.325297           1.230259          -6.775365   \n",
       "703            2.282723           1.468014          -6.275698   \n",
       "704            2.192926           1.783352          -4.749588   \n",
       "\n",
       "     ENST00000002125.9  ENST00000002165.11  ENST00000002501.10  ...  \\\n",
       "0             0.959451            2.498851           -4.311078  ...   \n",
       "1             0.282856            1.920547           -1.912931  ...   \n",
       "2             0.715676            2.425490           -3.582005  ...   \n",
       "3             0.744658            2.283302           -3.806516  ...   \n",
       "4             0.744347            2.293409           -3.557772  ...   \n",
       "..                 ...                 ...                 ...  ...   \n",
       "700           0.715169            2.050926           -3.356038  ...   \n",
       "701           0.877213            2.379205           -3.336567  ...   \n",
       "702           0.143498            1.551064           -3.315934  ...   \n",
       "703           0.363341            1.755888           -4.401229  ...   \n",
       "704           0.147652            1.350548           -2.679199  ...   \n",
       "\n",
       "     ENST00000673594.1  ENST00000673597.1  ENST00000673598.1  \\\n",
       "0            -2.385079          -6.633006          -2.726115   \n",
       "1            -2.695340          -3.879764          -3.638756   \n",
       "2            -2.228368          -6.751930          -3.051490   \n",
       "3            -2.452880          -6.976441          -2.888979   \n",
       "4            -2.373348          -7.017204          -2.929741   \n",
       "..                 ...                ...                ...   \n",
       "700          -2.020435          -7.748356          -2.890375   \n",
       "701          -2.114174          -6.506492          -3.047060   \n",
       "702          -2.383048          -6.775365          -3.315934   \n",
       "703          -1.978017          -7.860660          -3.105773   \n",
       "704          -2.547955          -7.071516          -3.371077   \n",
       "\n",
       "     ENST00000673599.1  ENST00000673600.1  ENST00000673603.1  \\\n",
       "0            -0.800116           3.760384          -3.463081   \n",
       "1            -1.557836           3.278272          -3.251733   \n",
       "2            -0.869287           3.744924          -2.228368   \n",
       "3            -0.999162           3.769233          -3.806516   \n",
       "4            -0.908680           3.428845          -2.373348   \n",
       "..                 ...                ...                ...   \n",
       "700          -0.805841           3.717721          -4.047916   \n",
       "701          -0.891782           4.249230          -4.921529   \n",
       "702          -0.942475           3.518106          -3.605440   \n",
       "703          -0.965843           3.265399          -2.368807   \n",
       "704          -1.188873           3.205771          -3.901591   \n",
       "\n",
       "     ENST00000673604.1  ENST00000673615.1  ENST00000673620.1  \\\n",
       "0            -2.932566          -6.633006          -0.190063   \n",
       "1            -3.879764          -1.981644          -0.544780   \n",
       "2            -3.582005          -5.166967          -0.122573   \n",
       "3            -3.276002          -6.976441          -0.406586   \n",
       "4            -3.110313          -5.432242          -0.275737   \n",
       "..                 ...                ...                ...   \n",
       "700          -3.104500          -7.748356           0.152511   \n",
       "701          -3.047060          -6.506492          -0.030758   \n",
       "702          -3.968010          -5.190403          -0.007181   \n",
       "703          -3.216804          -7.860660           0.232097   \n",
       "704          -3.612085          -7.071516           0.329363   \n",
       "\n",
       "     ENST00000673621.1  \n",
       "0             0.135178  \n",
       "1            -0.305773  \n",
       "2            -0.122573  \n",
       "3            -0.318230  \n",
       "4             0.201964  \n",
       "..                 ...  \n",
       "700           0.290563  \n",
       "701           0.313687  \n",
       "702           0.537518  \n",
       "703           0.569792  \n",
       "704           0.635843  \n",
       "\n",
       "[705 rows x 219245 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop columns if not case_id_list\n",
    "base_aa_case = base_aa.drop(columns=[col for col in base_aa[base_aa.columns] if col not in case_id_list])\n",
    "\n",
    "#reset index\n",
    "base_aa_case = base_aa_case.reset_index()\n",
    "\n",
    "#rename index column as ID\n",
    "base_aa_case = base_aa_case.rename(columns={'index': 'ID'})\n",
    "\n",
    "#split ID \n",
    "base_aa_case['ID'] = base_aa_case['ID'].str.split(':').str[0]\n",
    "\n",
    "#merge tr_list with base_aa_case\n",
    "base_tr_case = pd.merge(base_aa_case, tr, on=['ID'], how='inner')\n",
    "\n",
    "#set index\n",
    "base_tr_case.set_index('ID', inplace=True)\n",
    "\n",
    "#transpose the dataframe\n",
    "base_tr_case = base_tr_case.T\n",
    "\n",
    "#add ID as index name and reset the index\n",
    "base_tr_case = base_tr_case.rename_axis('FID').reset_index()\n",
    "\n",
    "#grab ID column and duplicate into two columns to compare with .fam file FID IID\n",
    "base_tr_case.insert(1, 'IID', base_tr_case['FID'])\n",
    "\n",
    "#remove index name\n",
    "base_tr_case = base_tr_case.rename_axis(None, axis=1)\n",
    "\n",
    "base_tr_case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6b40f6c6-0489-4adc-9f89-ad56c762fb42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>ENST00000000233.10</th>\n",
       "      <th>ENST00000000412.8</th>\n",
       "      <th>ENST00000000442.11</th>\n",
       "      <th>ENST00000001008.6</th>\n",
       "      <th>ENST00000001146.6</th>\n",
       "      <th>ENST00000002125.9</th>\n",
       "      <th>ENST00000002165.11</th>\n",
       "      <th>ENST00000002501.10</th>\n",
       "      <th>...</th>\n",
       "      <th>ENST00000673594.1</th>\n",
       "      <th>ENST00000673597.1</th>\n",
       "      <th>ENST00000673598.1</th>\n",
       "      <th>ENST00000673599.1</th>\n",
       "      <th>ENST00000673600.1</th>\n",
       "      <th>ENST00000673603.1</th>\n",
       "      <th>ENST00000673604.1</th>\n",
       "      <th>ENST00000673615.1</th>\n",
       "      <th>ENST00000673620.1</th>\n",
       "      <th>ENST00000673621.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP-14426</td>\n",
       "      <td>PP-14426</td>\n",
       "      <td>1.079028</td>\n",
       "      <td>3.384911</td>\n",
       "      <td>2.783043</td>\n",
       "      <td>2.035153</td>\n",
       "      <td>-6.682523</td>\n",
       "      <td>1.206220</td>\n",
       "      <td>2.915529</td>\n",
       "      <td>-2.775633</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.927636</td>\n",
       "      <td>-6.682523</td>\n",
       "      <td>-2.775633</td>\n",
       "      <td>-0.453705</td>\n",
       "      <td>4.289739</td>\n",
       "      <td>-3.875168</td>\n",
       "      <td>-3.223092</td>\n",
       "      <td>-6.682523</td>\n",
       "      <td>-0.143364</td>\n",
       "      <td>0.137656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP-15761</td>\n",
       "      <td>PP-15761</td>\n",
       "      <td>0.541646</td>\n",
       "      <td>2.996052</td>\n",
       "      <td>2.126609</td>\n",
       "      <td>1.568978</td>\n",
       "      <td>-0.996774</td>\n",
       "      <td>0.338829</td>\n",
       "      <td>2.110755</td>\n",
       "      <td>-2.330198</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.450492</td>\n",
       "      <td>-4.652126</td>\n",
       "      <td>-3.273614</td>\n",
       "      <td>-1.419465</td>\n",
       "      <td>3.289215</td>\n",
       "      <td>-2.450492</td>\n",
       "      <td>-3.514622</td>\n",
       "      <td>-2.886591</td>\n",
       "      <td>-0.344697</td>\n",
       "      <td>-0.564663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP-17608</td>\n",
       "      <td>PP-17608</td>\n",
       "      <td>0.386916</td>\n",
       "      <td>3.019584</td>\n",
       "      <td>2.272599</td>\n",
       "      <td>2.000488</td>\n",
       "      <td>-2.722383</td>\n",
       "      <td>0.572578</td>\n",
       "      <td>1.976001</td>\n",
       "      <td>-4.669915</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.432876</td>\n",
       "      <td>-7.477270</td>\n",
       "      <td>-2.953708</td>\n",
       "      <td>-1.137420</td>\n",
       "      <td>3.265039</td>\n",
       "      <td>-4.017838</td>\n",
       "      <td>-3.389807</td>\n",
       "      <td>-7.477270</td>\n",
       "      <td>0.517083</td>\n",
       "      <td>0.494273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PP-3000</td>\n",
       "      <td>PP-3000</td>\n",
       "      <td>0.453616</td>\n",
       "      <td>2.575986</td>\n",
       "      <td>2.197861</td>\n",
       "      <td>1.569525</td>\n",
       "      <td>-2.597812</td>\n",
       "      <td>0.040448</td>\n",
       "      <td>1.679213</td>\n",
       "      <td>-2.797121</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.342555</td>\n",
       "      <td>-5.230080</td>\n",
       "      <td>-3.464546</td>\n",
       "      <td>-1.248228</td>\n",
       "      <td>2.954795</td>\n",
       "      <td>-3.464546</td>\n",
       "      <td>-3.645118</td>\n",
       "      <td>-4.382084</td>\n",
       "      <td>0.010234</td>\n",
       "      <td>-0.005114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PP-3004</td>\n",
       "      <td>PP-3004</td>\n",
       "      <td>0.554084</td>\n",
       "      <td>2.515141</td>\n",
       "      <td>1.921878</td>\n",
       "      <td>1.236488</td>\n",
       "      <td>-1.187848</td>\n",
       "      <td>0.534116</td>\n",
       "      <td>1.882937</td>\n",
       "      <td>-3.382864</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.421338</td>\n",
       "      <td>-4.823436</td>\n",
       "      <td>-3.238474</td>\n",
       "      <td>-1.748148</td>\n",
       "      <td>3.676979</td>\n",
       "      <td>-2.986935</td>\n",
       "      <td>-3.930351</td>\n",
       "      <td>-2.273239</td>\n",
       "      <td>-0.247087</td>\n",
       "      <td>-0.008739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>PP-65000</td>\n",
       "      <td>PP-65000</td>\n",
       "      <td>1.270003</td>\n",
       "      <td>2.563567</td>\n",
       "      <td>2.436615</td>\n",
       "      <td>1.568766</td>\n",
       "      <td>-5.550176</td>\n",
       "      <td>-0.289648</td>\n",
       "      <td>1.519498</td>\n",
       "      <td>-3.965213</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.090744</td>\n",
       "      <td>-7.135138</td>\n",
       "      <td>-3.434698</td>\n",
       "      <td>-0.795288</td>\n",
       "      <td>3.185662</td>\n",
       "      <td>-2.611576</td>\n",
       "      <td>-2.380251</td>\n",
       "      <td>-7.135138</td>\n",
       "      <td>-0.420893</td>\n",
       "      <td>1.126957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>PP-65002</td>\n",
       "      <td>PP-65002</td>\n",
       "      <td>1.196024</td>\n",
       "      <td>3.891377</td>\n",
       "      <td>3.174650</td>\n",
       "      <td>2.052378</td>\n",
       "      <td>-6.061364</td>\n",
       "      <td>1.157804</td>\n",
       "      <td>2.768359</td>\n",
       "      <td>-3.739436</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.537802</td>\n",
       "      <td>-6.061364</td>\n",
       "      <td>-2.360924</td>\n",
       "      <td>-0.775962</td>\n",
       "      <td>4.472939</td>\n",
       "      <td>-3.254009</td>\n",
       "      <td>-2.360924</td>\n",
       "      <td>-6.061364</td>\n",
       "      <td>-0.333444</td>\n",
       "      <td>0.313675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>PP-65003</td>\n",
       "      <td>PP-65003</td>\n",
       "      <td>0.260479</td>\n",
       "      <td>2.779444</td>\n",
       "      <td>1.707703</td>\n",
       "      <td>1.763393</td>\n",
       "      <td>-1.927148</td>\n",
       "      <td>0.552542</td>\n",
       "      <td>2.269249</td>\n",
       "      <td>-2.464805</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.895439</td>\n",
       "      <td>-5.834039</td>\n",
       "      <td>-3.026684</td>\n",
       "      <td>-1.115220</td>\n",
       "      <td>3.123063</td>\n",
       "      <td>-3.026684</td>\n",
       "      <td>-3.512110</td>\n",
       "      <td>-3.512110</td>\n",
       "      <td>0.112380</td>\n",
       "      <td>0.232051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>PP-65004</td>\n",
       "      <td>PP-65004</td>\n",
       "      <td>0.821093</td>\n",
       "      <td>2.343860</td>\n",
       "      <td>2.134473</td>\n",
       "      <td>1.670442</td>\n",
       "      <td>-7.067651</td>\n",
       "      <td>-0.247472</td>\n",
       "      <td>1.601234</td>\n",
       "      <td>-4.260296</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.675333</td>\n",
       "      <td>-7.067651</td>\n",
       "      <td>-3.367211</td>\n",
       "      <td>-1.185008</td>\n",
       "      <td>2.961637</td>\n",
       "      <td>-3.160760</td>\n",
       "      <td>-3.160760</td>\n",
       "      <td>-7.067651</td>\n",
       "      <td>-0.056423</td>\n",
       "      <td>0.625836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>PP-65008</td>\n",
       "      <td>PP-65008</td>\n",
       "      <td>0.493132</td>\n",
       "      <td>1.673927</td>\n",
       "      <td>1.379228</td>\n",
       "      <td>1.070522</td>\n",
       "      <td>-1.529708</td>\n",
       "      <td>-0.264118</td>\n",
       "      <td>1.139705</td>\n",
       "      <td>-2.691979</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.996834</td>\n",
       "      <td>-5.499334</td>\n",
       "      <td>-3.914372</td>\n",
       "      <td>-1.592443</td>\n",
       "      <td>2.425478</td>\n",
       "      <td>-3.177406</td>\n",
       "      <td>-3.177406</td>\n",
       "      <td>-2.226316</td>\n",
       "      <td>-0.264118</td>\n",
       "      <td>0.415549</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>526 rows × 219245 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          FID       IID  ENST00000000233.10  ENST00000000412.8  \\\n",
       "0    PP-14426  PP-14426            1.079028           3.384911   \n",
       "1    PP-15761  PP-15761            0.541646           2.996052   \n",
       "2    PP-17608  PP-17608            0.386916           3.019584   \n",
       "3     PP-3000   PP-3000            0.453616           2.575986   \n",
       "4     PP-3004   PP-3004            0.554084           2.515141   \n",
       "..        ...       ...                 ...                ...   \n",
       "521  PP-65000  PP-65000            1.270003           2.563567   \n",
       "522  PP-65002  PP-65002            1.196024           3.891377   \n",
       "523  PP-65003  PP-65003            0.260479           2.779444   \n",
       "524  PP-65004  PP-65004            0.821093           2.343860   \n",
       "525  PP-65008  PP-65008            0.493132           1.673927   \n",
       "\n",
       "     ENST00000000442.11  ENST00000001008.6  ENST00000001146.6  \\\n",
       "0              2.783043           2.035153          -6.682523   \n",
       "1              2.126609           1.568978          -0.996774   \n",
       "2              2.272599           2.000488          -2.722383   \n",
       "3              2.197861           1.569525          -2.597812   \n",
       "4              1.921878           1.236488          -1.187848   \n",
       "..                  ...                ...                ...   \n",
       "521            2.436615           1.568766          -5.550176   \n",
       "522            3.174650           2.052378          -6.061364   \n",
       "523            1.707703           1.763393          -1.927148   \n",
       "524            2.134473           1.670442          -7.067651   \n",
       "525            1.379228           1.070522          -1.529708   \n",
       "\n",
       "     ENST00000002125.9  ENST00000002165.11  ENST00000002501.10  ...  \\\n",
       "0             1.206220            2.915529           -2.775633  ...   \n",
       "1             0.338829            2.110755           -2.330198  ...   \n",
       "2             0.572578            1.976001           -4.669915  ...   \n",
       "3             0.040448            1.679213           -2.797121  ...   \n",
       "4             0.534116            1.882937           -3.382864  ...   \n",
       "..                 ...                 ...                 ...  ...   \n",
       "521          -0.289648            1.519498           -3.965213  ...   \n",
       "522           1.157804            2.768359           -3.739436  ...   \n",
       "523           0.552542            2.269249           -2.464805  ...   \n",
       "524          -0.247472            1.601234           -4.260296  ...   \n",
       "525          -0.264118            1.139705           -2.691979  ...   \n",
       "\n",
       "     ENST00000673594.1  ENST00000673597.1  ENST00000673598.1  \\\n",
       "0            -1.927636          -6.682523          -2.775633   \n",
       "1            -2.450492          -4.652126          -3.273614   \n",
       "2            -2.432876          -7.477270          -2.953708   \n",
       "3            -2.342555          -5.230080          -3.464546   \n",
       "4            -2.421338          -4.823436          -3.238474   \n",
       "..                 ...                ...                ...   \n",
       "521          -2.090744          -7.135138          -3.434698   \n",
       "522          -1.537802          -6.061364          -2.360924   \n",
       "523          -2.895439          -5.834039          -3.026684   \n",
       "524          -2.675333          -7.067651          -3.367211   \n",
       "525          -2.996834          -5.499334          -3.914372   \n",
       "\n",
       "     ENST00000673599.1  ENST00000673600.1  ENST00000673603.1  \\\n",
       "0            -0.453705           4.289739          -3.875168   \n",
       "1            -1.419465           3.289215          -2.450492   \n",
       "2            -1.137420           3.265039          -4.017838   \n",
       "3            -1.248228           2.954795          -3.464546   \n",
       "4            -1.748148           3.676979          -2.986935   \n",
       "..                 ...                ...                ...   \n",
       "521          -0.795288           3.185662          -2.611576   \n",
       "522          -0.775962           4.472939          -3.254009   \n",
       "523          -1.115220           3.123063          -3.026684   \n",
       "524          -1.185008           2.961637          -3.160760   \n",
       "525          -1.592443           2.425478          -3.177406   \n",
       "\n",
       "     ENST00000673604.1  ENST00000673615.1  ENST00000673620.1  \\\n",
       "0            -3.223092          -6.682523          -0.143364   \n",
       "1            -3.514622          -2.886591          -0.344697   \n",
       "2            -3.389807          -7.477270           0.517083   \n",
       "3            -3.645118          -4.382084           0.010234   \n",
       "4            -3.930351          -2.273239          -0.247087   \n",
       "..                 ...                ...                ...   \n",
       "521          -2.380251          -7.135138          -0.420893   \n",
       "522          -2.360924          -6.061364          -0.333444   \n",
       "523          -3.512110          -3.512110           0.112380   \n",
       "524          -3.160760          -7.067651          -0.056423   \n",
       "525          -3.177406          -2.226316          -0.264118   \n",
       "\n",
       "     ENST00000673621.1  \n",
       "0             0.137656  \n",
       "1            -0.564663  \n",
       "2             0.494273  \n",
       "3            -0.005114  \n",
       "4            -0.008739  \n",
       "..                 ...  \n",
       "521           1.126957  \n",
       "522           0.313675  \n",
       "523           0.232051  \n",
       "524           0.625836  \n",
       "525           0.415549  \n",
       "\n",
       "[526 rows x 219245 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop columns if not case_id_list\n",
    "base_aa_cont = base_aa.drop(columns=[col for col in base_aa[base_aa.columns] if col not in cont_id_list])\n",
    "\n",
    "#reset index\n",
    "base_aa_cont = base_aa_cont.reset_index()\n",
    "\n",
    "#rename index column as ID\n",
    "base_aa_cont = base_aa_cont.rename(columns={'index': 'ID'})\n",
    "\n",
    "#split ID \n",
    "base_aa_cont['ID'] = base_aa_cont['ID'].str.split(':').str[0]\n",
    "\n",
    "#merge tr_list with base_aa_case\n",
    "base_tr_cont = pd.merge(base_aa_cont, tr, on=['ID'], how='inner')\n",
    "\n",
    "#set index\n",
    "base_tr_cont.set_index('ID', inplace=True)\n",
    "\n",
    "#transpose the dataframe\n",
    "base_tr_cont = base_tr_cont.T\n",
    "\n",
    "#add ID as index name and reset the index\n",
    "base_tr_cont = base_tr_cont.rename_axis('FID').reset_index()\n",
    "\n",
    "#grab ID column and duplicate into two columns to compare with .fam file FID IID\n",
    "base_tr_cont.insert(1, 'IID', base_tr_cont['FID'])\n",
    "\n",
    "#remove index name\n",
    "base_tr_cont = base_tr_cont.rename_axis(None, axis=1)\n",
    "\n",
    "base_tr_cont"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b533a43e-98d1-485a-bcb9-31e2b78f1cac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case ID: \n",
      "        FID      IID\n",
      "0  PP-3001  PP-3001\n",
      "1  PP-3002  PP-3002\n",
      "2  PP-3003  PP-3003\n",
      "3  PP-3006  PP-3006\n",
      "4  PP-3007  PP-3007\n",
      "Control ID: \n",
      "         FID       IID\n",
      "0  PP-14426  PP-14426\n",
      "1  PP-15761  PP-15761\n",
      "2  PP-17608  PP-17608\n",
      "3   PP-3000   PP-3000\n",
      "4   PP-3004   PP-3004\n"
     ]
    }
   ],
   "source": [
    "#make case_id_df\n",
    "case_id_df = base_tr_case.copy()\n",
    "case_id_df.drop(case_id_df.iloc[:,2:219245], axis=1, inplace=True)\n",
    "case_id = case_id_df['FID'].tolist()\n",
    "print('Case ID: \\n', case_id_df.head())\n",
    "\n",
    "#make control_id_df\n",
    "cont_id_df = base_tr_cont.copy()\n",
    "cont_id_df.drop(cont_id_df.iloc[:,2:219245], axis=1, inplace=True)\n",
    "cont_id = cont_id_df['FID'].tolist()\n",
    "print('Control ID: \\n', cont_id_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "18d479fc-6b7f-49a1-be01-3c4c35b65e69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of covar_df_case_scale_trim: (705, 16)\n",
      "       FID      IID  EDUCATION  FAMILY_HISTORY  MALE       AGE   PC1_gen  \\\n",
      "0  PP-3400  PP-3400   0.692308               0     0  0.098361 -0.004662   \n",
      "1  PP-3402  PP-3402   0.615385               0     1  0.213115  0.011903   \n",
      "2  PP-3403  PP-3403   0.692308               0     1  0.590164 -0.014213   \n",
      "3  PP-3406  PP-3406   0.692308               0     1  0.032787 -0.008313   \n",
      "4  PP-3407  PP-3407   0.615385               0     1  0.540984  0.033813   \n",
      "\n",
      "    PC2_gen   PC3_gen   PC4_gen   PC5_gen   PC6_gen   PC7_gen   PC8_gen  \\\n",
      "0 -0.023299  0.006330 -0.001153  0.003980 -0.002586 -0.000425 -0.007179   \n",
      "1 -0.023524 -0.001647  0.005045 -0.003863 -0.003192 -0.003624  0.000327   \n",
      "2 -0.025636 -0.002261 -0.003854 -0.017373  0.002293 -0.002750  0.023918   \n",
      "3 -0.010611  0.017661 -0.010679 -0.007702 -0.004243 -0.004153  0.002486   \n",
      "4 -0.012696  0.020515 -0.000225 -0.008233 -0.003591  0.002511 -0.004819   \n",
      "\n",
      "    PC9_gen  PC10_gen  \n",
      "0 -0.020155 -0.000930  \n",
      "1  0.012216 -0.006821  \n",
      "2  0.020800 -0.003866  \n",
      "3  0.002382  0.008385  \n",
      "4  0.002946  0.009602  \n",
      "Shape of covar_df_cont_scale_trim: (526, 16)\n",
      "       FID      IID  EDUCATION  FAMILY_HISTORY  MALE       AGE   PC1_gen  \\\n",
      "0  PP-3401  PP-3401   0.607143               0     0  0.544118  0.034271   \n",
      "1  PP-3404  PP-3404   0.428571               0     0  0.544118 -0.011506   \n",
      "2  PP-3405  PP-3405   0.571429               0     0  0.647059 -0.011204   \n",
      "3  PP-3100  PP-3100   0.500000               0     0  0.735294 -0.012570   \n",
      "4  PP-3151  PP-3151   0.392857               0     1  0.573529 -0.011135   \n",
      "\n",
      "    PC2_gen   PC3_gen   PC4_gen   PC5_gen   PC6_gen   PC7_gen   PC8_gen  \\\n",
      "0 -0.023656  0.016260  0.000452 -0.005504 -0.000337 -0.000150  0.011953   \n",
      "1 -0.020343  0.016984  0.003306 -0.011024 -0.004369 -0.006111  0.024201   \n",
      "2 -0.014210 -0.000746 -0.004401 -0.019775 -0.003406 -0.012721  0.017960   \n",
      "3 -0.010427  0.008777  0.000524 -0.002120 -0.001913 -0.010488  0.003645   \n",
      "4 -0.016579  0.006779  0.005386 -0.012485 -0.005944 -0.001021  0.036534   \n",
      "\n",
      "    PC9_gen  PC10_gen  \n",
      "0  0.014463 -0.001755  \n",
      "1  0.000526  0.003789  \n",
      "2 -0.013703 -0.000250  \n",
      "3 -0.015522 -0.000490  \n",
      "4  0.016744  0.002072  \n"
     ]
    }
   ],
   "source": [
    "#drop rows covariate data if FID is not in case_id\n",
    "covar_df_case_scale_trim = covar_df_case_scale[covar_df_case_scale['FID'].isin(case_id)] \n",
    "print('Shape of covar_df_case_scale_trim:', covar_df_case_scale_trim.shape)\n",
    "print(covar_df_case_scale_trim.head())\n",
    "\n",
    "#drop rows covariate data if FID is not in cont_id\n",
    "covar_df_cont_scale_trim = covar_df_cont_scale[covar_df_cont_scale['FID'].isin(cont_id)]\n",
    "print('Shape of covar_df_cont_scale_trim:', covar_df_cont_scale_trim.shape)\n",
    "print(covar_df_cont_scale_trim.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ca1a5a60-9d4f-48cd-8c2c-ece7fe46dd32",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save base_aa_case as csv --run only once\n",
    "base_tr_case.to_hdf(r'./data_folder/base_transcript_case.h5', key='stage', mode='w')\n",
    "base_tr_cont.to_hdf(r'./data_folder/base_transcript_control.h5', key='stage', mode='w')\n",
    "\n",
    "#base_ab_case.to_csv(r'./data_folder/base_2_case.csv') --> this takes too long\n",
    "\n",
    "#save id_df as txt  -- run only once for download\n",
    "case_id_df.to_csv(r'./data_folder/case_id_df.txt', sep='\\t' ,index=False)\n",
    "cont_id_df.to_csv(r'./data_folder/control_id_df.txt', sep='\\t' ,index=False)\n",
    "\n",
    "#save covariate as txt -- run only once for download\n",
    "covar_df_case_scale.to_csv(r'./data_folder/te_covariates_case.txt', sep='\\t' ,index=False)\n",
    "covar_df_cont_scale.to_csv(r'./data_folder/te_covariates_control.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f554e5ac-4f63-4a93-8095-35ba7ee0aa45",
   "metadata": {},
   "source": [
    "Trim plink files with id_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbd32e0a-d467-4538-b337-639578a3b6a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter sample id using fid_df.txt to reduce rows that are not in fid_df.txt\n",
    "##plink --bfile <genotype file> --keep <id file> --make-bed --out <outfile>\n",
    "!plink --bfile /{datadir}/hits_plus_tags_AMPPD_Euros --keep /{datadir}/case_id_df.txt --maf 0.01 --make-bed --out /{datadir}/qc_genotypes_tes_case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93939979-1d4b-42aa-bf6b-07c5ba511f38",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter sample id using fid_df.txt to reduce rows that are not in fid_df.txt\n",
    "##plink --bfile <genotype file> --keep <id file> --make-bed --out <outfile>\n",
    "!plink --bfile /{datadir}/hits_plus_tags_AMPPD_Euros --keep /{datadir}/control_id_df.txt  --maf 0.01 --make-bed --out /{datadir}/qc_genotypes_tes_control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "cf1c86ac-7e8d-491d-886d-06d3aadb4458",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "705\n",
      "PP-3001 PP-3001 0 0 0 -9\n",
      "PP-3002 PP-3002 0 0 0 -9\n",
      "PP-3003 PP-3003 0 0 0 -9\n",
      "PP-3006 PP-3006 0 0 0 -9\n",
      "PP-3007 PP-3007 0 0 0 -9\n",
      "PP-3010 PP-3010 0 0 0 -9\n",
      "PP-3012 PP-3012 0 0 0 -9\n",
      "PP-3014 PP-3014 0 0 0 -9\n",
      "PP-3018 PP-3018 0 0 0 -9\n",
      "PP-3020 PP-3020 0 0 0 -9\n",
      "305138\n",
      "1\trs200683566\t0\t13838\tT\tC\n",
      "1\trs370886505\t0\t14397\tC\tCTGT\n",
      "1\trs375086259\t0\t14653\tT\tC\n",
      "1\trs79585140\t0\t14907\tG\tA\n",
      "1\trs199856693\t0\t14933\tA\tG\n",
      "1\trs71252250\t0\t15118\tG\tA\n",
      "1\trs201635489\t0\t15447\tG\tA\n",
      "1\trs201026389\t0\t16125\tG\tT\n",
      "1\trs78588380\t0\t16257\tC\tG\n",
      "1\trs200736374\t0\t16288\tG\tC\n"
     ]
    }
   ],
   "source": [
    "#check number of IDs in original .fam file\n",
    "!cat {datadir}/qc_genotypes_tes_case.fam | wc -l\n",
    "#check .fam file\n",
    "!head /{datadir}/qc_genotypes_tes_case.fam\n",
    "\n",
    "#check number of IDs in original .fam file\n",
    "!cat {datadir}/qc_genotypes_tes_case.bim | wc -l\n",
    "#check .fam file\n",
    "!head /{datadir}/qc_genotypes_tes_case.bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f382c479-b27e-4211-8745-0105b3fcd3e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "526\n",
      "PP-3000 PP-3000 0 0 0 -9\n",
      "PP-3004 PP-3004 0 0 0 -9\n",
      "PP-3008 PP-3008 0 0 0 -9\n",
      "PP-3011 PP-3011 0 0 0 -9\n",
      "PP-3013 PP-3013 0 0 0 -9\n",
      "PP-3016 PP-3016 0 0 0 -9\n",
      "PP-3029 PP-3029 0 0 0 -9\n",
      "PP-3053 PP-3053 0 0 0 -9\n",
      "PP-3055 PP-3055 0 0 0 -9\n",
      "PP-3057 PP-3057 0 0 0 -9\n",
      "305136\n",
      "1\trs200683566\t0\t13838\tT\tC\n",
      "1\trs370886505\t0\t14397\tC\tCTGT\n",
      "1\trs375086259\t0\t14653\tT\tC\n",
      "1\trs79585140\t0\t14907\tG\tA\n",
      "1\trs199856693\t0\t14933\tA\tG\n",
      "1\trs71252250\t0\t15118\tG\tA\n",
      "1\trs201635489\t0\t15447\tG\tA\n",
      "1\trs201026389\t0\t16125\tG\tT\n",
      "1\trs78588380\t0\t16257\tC\tG\n",
      "1\trs200736374\t0\t16288\tG\tC\n"
     ]
    }
   ],
   "source": [
    "#check number of IDs in original .fam file\n",
    "!cat {datadir}/qc_genotypes_tes_control.fam | wc -l\n",
    "#check .fam file\n",
    "!head /{datadir}/qc_genotypes_tes_control.fam\n",
    "\n",
    "#check number of IDs in original .fam file\n",
    "!cat {datadir}/qc_genotypes_tes_control.bim | wc -l\n",
    "#check .fam file\n",
    "!head /{datadir}/qc_genotypes_tes_control.bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "762882d8-732e-436c-ba95-d72e55c3b1b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "166c254c-4358-4bf7-9848-402237666fb9",
   "metadata": {},
   "source": [
    "Compare patient ID in twas and tes (this step is not necessary if you are not comparing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c9702e33-bbf4-487c-9537-0a3d3499bb35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of twas control id: (575, 2)\n",
      "number of twas case id: (1138, 2)\n"
     ]
    }
   ],
   "source": [
    "twas_cont_id = pd.read_csv(f\"/data/songy4/twas/data_folder/control_id_df.txt\", engine='c', sep='\\t')\n",
    "twas_case_id = pd.read_csv(f\"/data/songy4/twas/data_folder/case_id_df.txt\", engine='c', sep='\\t')\n",
    "print('number of twas control id:', twas_cont_id.shape)\n",
    "print('number of twas case id:', twas_case_id.shape)\n",
    "\n",
    "#make a list on each id columns\n",
    "twas_case = twas_case_id['FID'].tolist()\n",
    "twas_cont = twas_cont_id['FID'].tolist()\n",
    "te_case = case_id_df['FID'].tolist()\n",
    "te_cont = cont_id_df['FID'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "86bf2661-1190-4a87-a2d2-2fb6fad1a29e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of case only patient ID in TWAS and TE:  401\n",
      "number of control only patient ID in TWAS and TE:  163\n"
     ]
    }
   ],
   "source": [
    "####compare FID from TE data with TWAS data\n",
    "\n",
    "def comprehension(a, b):\n",
    "    return [x for x in a if x in b]\n",
    "\n",
    "case_id = comprehension(twas_case,te_case)\n",
    "cont_id = comprehension(twas_cont,te_cont)\n",
    "\n",
    "print('number of case only patient ID in TWAS and TE: ', len(case_id))\n",
    "print('number of control only patient ID in TWAS and TE: ', len(cont_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0f844939-297c-4f66-a174-96b662415ece",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of patient ID in TWAS and TE:  568\n"
     ]
    }
   ],
   "source": [
    "#check if case and control combined lists has more overlap between twas and te\n",
    "twas_id_list = twas_case + twas_cont\n",
    "te_id_list = te_case + te_cont\n",
    "id_list = comprehension(twas_id_list,te_id_list)\n",
    "print('number of patient ID in TWAS and TE: ', len(id_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0077f029-a2b9-4847-a479-2a01a900befe",
   "metadata": {},
   "source": [
    "data statistics (to write a paper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6c866c2-c522-401d-9282-63a4a0d73bd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>MALE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>PC1_gen</th>\n",
       "      <th>PC2_gen</th>\n",
       "      <th>PC3_gen</th>\n",
       "      <th>PC4_gen</th>\n",
       "      <th>PC5_gen</th>\n",
       "      <th>PC6_gen</th>\n",
       "      <th>PC7_gen</th>\n",
       "      <th>PC8_gen</th>\n",
       "      <th>PC9_gen</th>\n",
       "      <th>PC10_gen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP-3400</td>\n",
       "      <td>PP-3400</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.098361</td>\n",
       "      <td>-0.004662</td>\n",
       "      <td>-0.023299</td>\n",
       "      <td>0.006330</td>\n",
       "      <td>-0.001153</td>\n",
       "      <td>0.003980</td>\n",
       "      <td>-0.002586</td>\n",
       "      <td>-0.000425</td>\n",
       "      <td>-0.007179</td>\n",
       "      <td>-0.020155</td>\n",
       "      <td>-0.000930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP-3402</td>\n",
       "      <td>PP-3402</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.213115</td>\n",
       "      <td>0.011903</td>\n",
       "      <td>-0.023524</td>\n",
       "      <td>-0.001647</td>\n",
       "      <td>0.005045</td>\n",
       "      <td>-0.003863</td>\n",
       "      <td>-0.003192</td>\n",
       "      <td>-0.003624</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.012216</td>\n",
       "      <td>-0.006821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP-3403</td>\n",
       "      <td>PP-3403</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.590164</td>\n",
       "      <td>-0.014213</td>\n",
       "      <td>-0.025636</td>\n",
       "      <td>-0.002261</td>\n",
       "      <td>-0.003854</td>\n",
       "      <td>-0.017373</td>\n",
       "      <td>0.002293</td>\n",
       "      <td>-0.002750</td>\n",
       "      <td>0.023918</td>\n",
       "      <td>0.020800</td>\n",
       "      <td>-0.003866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PP-3406</td>\n",
       "      <td>PP-3406</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.032787</td>\n",
       "      <td>-0.008313</td>\n",
       "      <td>-0.010611</td>\n",
       "      <td>0.017661</td>\n",
       "      <td>-0.010679</td>\n",
       "      <td>-0.007702</td>\n",
       "      <td>-0.004243</td>\n",
       "      <td>-0.004153</td>\n",
       "      <td>0.002486</td>\n",
       "      <td>0.002382</td>\n",
       "      <td>0.008385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PP-3407</td>\n",
       "      <td>PP-3407</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.540984</td>\n",
       "      <td>0.033813</td>\n",
       "      <td>-0.012696</td>\n",
       "      <td>0.020515</td>\n",
       "      <td>-0.000225</td>\n",
       "      <td>-0.008233</td>\n",
       "      <td>-0.003591</td>\n",
       "      <td>0.002511</td>\n",
       "      <td>-0.004819</td>\n",
       "      <td>0.002946</td>\n",
       "      <td>0.009602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>PP-40816</td>\n",
       "      <td>PP-40816</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.622951</td>\n",
       "      <td>0.032421</td>\n",
       "      <td>-0.004262</td>\n",
       "      <td>0.007910</td>\n",
       "      <td>-0.009565</td>\n",
       "      <td>-0.001786</td>\n",
       "      <td>-0.005201</td>\n",
       "      <td>0.002870</td>\n",
       "      <td>-0.002886</td>\n",
       "      <td>0.023377</td>\n",
       "      <td>0.010822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>PP-41801</td>\n",
       "      <td>PP-41801</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.770492</td>\n",
       "      <td>0.029534</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>-0.000886</td>\n",
       "      <td>0.016254</td>\n",
       "      <td>0.014111</td>\n",
       "      <td>-0.000023</td>\n",
       "      <td>0.008686</td>\n",
       "      <td>0.010129</td>\n",
       "      <td>0.026381</td>\n",
       "      <td>0.001977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>PP-42079</td>\n",
       "      <td>PP-42079</td>\n",
       "      <td>0.615385</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.540984</td>\n",
       "      <td>0.034438</td>\n",
       "      <td>-0.004636</td>\n",
       "      <td>-0.039376</td>\n",
       "      <td>-0.014328</td>\n",
       "      <td>-0.000155</td>\n",
       "      <td>-0.008150</td>\n",
       "      <td>-0.008431</td>\n",
       "      <td>-0.021045</td>\n",
       "      <td>-0.001832</td>\n",
       "      <td>-0.005066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>PP-50027</td>\n",
       "      <td>PP-50027</td>\n",
       "      <td>0.653846</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.426230</td>\n",
       "      <td>0.033116</td>\n",
       "      <td>0.013657</td>\n",
       "      <td>-0.003440</td>\n",
       "      <td>-0.009161</td>\n",
       "      <td>-0.008775</td>\n",
       "      <td>-0.002881</td>\n",
       "      <td>-0.000913</td>\n",
       "      <td>-0.003348</td>\n",
       "      <td>-0.017472</td>\n",
       "      <td>0.000134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>PP-55468</td>\n",
       "      <td>PP-55468</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.278689</td>\n",
       "      <td>0.034937</td>\n",
       "      <td>0.001731</td>\n",
       "      <td>0.017194</td>\n",
       "      <td>0.003374</td>\n",
       "      <td>-0.003817</td>\n",
       "      <td>0.001799</td>\n",
       "      <td>-0.005488</td>\n",
       "      <td>0.018886</td>\n",
       "      <td>0.001329</td>\n",
       "      <td>0.003024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>705 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          FID       IID  EDUCATION  FAMILY_HISTORY  MALE       AGE   PC1_gen  \\\n",
       "0     PP-3400   PP-3400   0.692308               0     0  0.098361 -0.004662   \n",
       "1     PP-3402   PP-3402   0.615385               0     1  0.213115  0.011903   \n",
       "2     PP-3403   PP-3403   0.692308               0     1  0.590164 -0.014213   \n",
       "3     PP-3406   PP-3406   0.692308               0     1  0.032787 -0.008313   \n",
       "4     PP-3407   PP-3407   0.615385               0     1  0.540984  0.033813   \n",
       "..        ...       ...        ...             ...   ...       ...       ...   \n",
       "700  PP-40816  PP-40816   0.692308               1     1  0.622951  0.032421   \n",
       "701  PP-41801  PP-41801   0.615385               0     1  0.770492  0.029534   \n",
       "702  PP-42079  PP-42079   0.615385               1     1  0.540984  0.034438   \n",
       "703  PP-50027  PP-50027   0.653846               1     1  0.426230  0.033116   \n",
       "704  PP-55468  PP-55468   0.500000               0     0  0.278689  0.034937   \n",
       "\n",
       "      PC2_gen   PC3_gen   PC4_gen   PC5_gen   PC6_gen   PC7_gen   PC8_gen  \\\n",
       "0   -0.023299  0.006330 -0.001153  0.003980 -0.002586 -0.000425 -0.007179   \n",
       "1   -0.023524 -0.001647  0.005045 -0.003863 -0.003192 -0.003624  0.000327   \n",
       "2   -0.025636 -0.002261 -0.003854 -0.017373  0.002293 -0.002750  0.023918   \n",
       "3   -0.010611  0.017661 -0.010679 -0.007702 -0.004243 -0.004153  0.002486   \n",
       "4   -0.012696  0.020515 -0.000225 -0.008233 -0.003591  0.002511 -0.004819   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "700 -0.004262  0.007910 -0.009565 -0.001786 -0.005201  0.002870 -0.002886   \n",
       "701  0.000108 -0.000886  0.016254  0.014111 -0.000023  0.008686  0.010129   \n",
       "702 -0.004636 -0.039376 -0.014328 -0.000155 -0.008150 -0.008431 -0.021045   \n",
       "703  0.013657 -0.003440 -0.009161 -0.008775 -0.002881 -0.000913 -0.003348   \n",
       "704  0.001731  0.017194  0.003374 -0.003817  0.001799 -0.005488  0.018886   \n",
       "\n",
       "      PC9_gen  PC10_gen  \n",
       "0   -0.020155 -0.000930  \n",
       "1    0.012216 -0.006821  \n",
       "2    0.020800 -0.003866  \n",
       "3    0.002382  0.008385  \n",
       "4    0.002946  0.009602  \n",
       "..        ...       ...  \n",
       "700  0.023377  0.010822  \n",
       "701  0.026381  0.001977  \n",
       "702 -0.001832 -0.005066  \n",
       "703 -0.017472  0.000134  \n",
       "704  0.001329  0.003024  \n",
       "\n",
       "[705 rows x 16 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "covariates_case = pd.read_csv(f\"{datadir}/te_covariates_case.txt\", sep='\\t')\n",
    "covariates_cont = pd.read_csv(f\"{datadir}/te_covariates_control.txt\", sep='\\t')\n",
    "covariates_case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "de9fdb95-c67f-4607-ad1a-da0a8c0bb16e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "% of MALE column in case (1=MALE, 0=FEMALE): \n",
      " 1    57.730496\n",
      "0    42.269504\n",
      "Name: MALE, dtype: float64\n",
      "% of MALE column in control (1=MALE, 0=FEMALE): \n",
      " 0    51.520913\n",
      "1    48.479087\n",
      "Name: MALE, dtype: float64\n",
      "% of FAMILY_HISTORY column in case (1=positive, 0=none): \n",
      " 0    73.900709\n",
      "1    26.099291\n",
      "Name: FAMILY_HISTORY, dtype: float64\n",
      "% of FAMILY_HISTORY column in control (1=positive, 0=none): \n",
      " 1    53.612167\n",
      "0    46.387833\n",
      "Name: FAMILY_HISTORY, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "#get percentage of MALE, FAMILY_HISTORY\n",
    "\n",
    "print('% of MALE column in case (1=MALE, 0=FEMALE): \\n', covariates_case['MALE'].value_counts(normalize=True) * 100)\n",
    "print('% of MALE column in control (1=MALE, 0=FEMALE): \\n', covariates_cont['MALE'].value_counts(normalize=True) * 100)\n",
    "\n",
    "print('% of FAMILY_HISTORY column in case (1=positive, 0=none): \\n', covariates_case['FAMILY_HISTORY'].value_counts(normalize=True) * 100)\n",
    "print('% of FAMILY_HISTORY column in control (1=positive, 0=none): \\n', covariates_cont['FAMILY_HISTORY'].value_counts(normalize=True) * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "809d86de-8eed-432a-956a-07a38fc62452",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean of AGE column in case: \n",
      " 63.589358799454295\n",
      "Standard Diviation of AGE column in case: \n",
      " 10.220211746349126\n",
      "Mean of AGE column in control: \n",
      " 58.28971962616822\n",
      "Standard Diviation of AGE column in control: \n",
      " 12.284097501215577\n"
     ]
    }
   ],
   "source": [
    "#get mean and SD of AGE\n",
    "\n",
    "print('Mean of AGE column in case: \\n', covar_df_case['AGE'].mean())\n",
    "print('Standard Diviation of AGE column in case: \\n', covar_df_case['AGE'].std())\n",
    "\n",
    "\n",
    "print('Mean of AGE column in control: \\n', covar_df_cont['AGE'].mean())\n",
    "print('Standard Diviation of AGE column in control: \\n', covar_df_cont['AGE'].std())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d5bc5b4-9fb3-42f5-95bd-73fc093277f3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.8",
   "language": "python",
   "name": "py3.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
