{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: tables in /usr/local/Anaconda/envs/py3.8/lib/python3.8/site-packages (3.6.1)\n",
      "Requirement already satisfied: numexpr>=2.6.2 in /usr/local/Anaconda/envs/py3.8/lib/python3.8/site-packages (from tables) (2.7.2)\n",
      "Requirement already satisfied: numpy>=1.9.3 in /usr/local/Anaconda/envs/py3.8/lib/python3.8/site-packages (from tables) (1.19.5)\n"
     ]
    }
   ],
   "source": [
    "#upload some libraries\n",
    "import os\n",
    "import pandas.testing as tm\n",
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import sys\n",
    "import joblib\n",
    "import subprocess\n",
    "import statsmodels.api as sm\n",
    "from scipy import stats\n",
    "\n",
    "! pip install --upgrade tables\n",
    "import tables\n",
    "\n",
    "# set paths\n",
    "basedir = '/data/songy4/twas'\n",
    "datadir = f'{basedir}/data_folder'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get Pheno (case vs control) data combined with expression data to select case only\n",
    "##this file should include FID and IID columns at the beginning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in pheno data: 1713\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD-PDAA503EF5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD-PDAA955TCY</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD-PDAB074CYQ</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD-PDAB411CTU</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD-PDAB549YWB</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID  PHENO\n",
       "0  PD-PDAA503EF5      1\n",
       "1  PD-PDAA955TCY      0\n",
       "2  PD-PDAB074CYQ      1\n",
       "3  PD-PDAB411CTU      0\n",
       "4  PD-PDAB549YWB      1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open 10PCs data file to extract pheno information\n",
    "\n",
    "pheno_df = pd.read_csv(f\"{datadir}/baseline_VST_proteincoding_omics_OG_PPMI_AND_PDBP_ALLTRANSCRIPTS_10PCs_loadings.csv\", engine='c')\n",
    "\n",
    "#count how manyID -- total number, both case and control\n",
    "print(\"Number of rows in pheno data:\", pheno_df.shape[0])\n",
    "\n",
    "#drop PCs from pheno_df\n",
    "pheno_id = pheno_df.copy()\n",
    "pheno_id.drop(pheno_id.iloc[:,2:12], axis=1, inplace=True)\n",
    "pheno_id.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows in expression data: 1713\n",
      "shape of case only expression data: (1138, 19687)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000285920</th>\n",
       "      <th>ENSG00000285937</th>\n",
       "      <th>ENSG00000285938</th>\n",
       "      <th>ENSG00000285946</th>\n",
       "      <th>ENSG00000285947</th>\n",
       "      <th>ENSG00000285950</th>\n",
       "      <th>ENSG00000285953</th>\n",
       "      <th>ENSG00000285975</th>\n",
       "      <th>ENSG00000285976</th>\n",
       "      <th>ENSG00000285982</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD-PDAA503EF5</td>\n",
       "      <td>PD-PDAA503EF5</td>\n",
       "      <td>1.412056</td>\n",
       "      <td>-0.523395</td>\n",
       "      <td>0.153490</td>\n",
       "      <td>-1.195026</td>\n",
       "      <td>-0.094739</td>\n",
       "      <td>0.278269</td>\n",
       "      <td>-1.606817</td>\n",
       "      <td>-1.231579</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.212292</td>\n",
       "      <td>0.085944</td>\n",
       "      <td>-0.274379</td>\n",
       "      <td>-0.065663</td>\n",
       "      <td>-0.126903</td>\n",
       "      <td>0.033090</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>0.059806</td>\n",
       "      <td>-1.298476</td>\n",
       "      <td>0.069994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD-PDAB074CYQ</td>\n",
       "      <td>PD-PDAB074CYQ</td>\n",
       "      <td>-0.856754</td>\n",
       "      <td>0.111547</td>\n",
       "      <td>0.741327</td>\n",
       "      <td>0.725518</td>\n",
       "      <td>1.231583</td>\n",
       "      <td>0.593704</td>\n",
       "      <td>-1.066773</td>\n",
       "      <td>0.544353</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.364937</td>\n",
       "      <td>0.051746</td>\n",
       "      <td>-0.786312</td>\n",
       "      <td>-0.003344</td>\n",
       "      <td>-0.166982</td>\n",
       "      <td>-0.059010</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>0.090575</td>\n",
       "      <td>-2.259028</td>\n",
       "      <td>0.097690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD-PDAB549YWB</td>\n",
       "      <td>PD-PDAB549YWB</td>\n",
       "      <td>1.118373</td>\n",
       "      <td>0.096341</td>\n",
       "      <td>0.427937</td>\n",
       "      <td>-1.107550</td>\n",
       "      <td>0.040538</td>\n",
       "      <td>-1.836332</td>\n",
       "      <td>-0.428239</td>\n",
       "      <td>1.349854</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.305483</td>\n",
       "      <td>-0.013945</td>\n",
       "      <td>-0.382129</td>\n",
       "      <td>-0.018791</td>\n",
       "      <td>-0.152562</td>\n",
       "      <td>0.084933</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>0.063530</td>\n",
       "      <td>-0.738222</td>\n",
       "      <td>0.053944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD-PDAB762PA3</td>\n",
       "      <td>PD-PDAB762PA3</td>\n",
       "      <td>-0.816394</td>\n",
       "      <td>-0.318743</td>\n",
       "      <td>-0.632429</td>\n",
       "      <td>-1.268311</td>\n",
       "      <td>0.100997</td>\n",
       "      <td>0.499943</td>\n",
       "      <td>-0.195359</td>\n",
       "      <td>1.295541</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.075632</td>\n",
       "      <td>0.096710</td>\n",
       "      <td>0.624501</td>\n",
       "      <td>0.031793</td>\n",
       "      <td>-0.143065</td>\n",
       "      <td>0.120146</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>-0.050857</td>\n",
       "      <td>-0.730364</td>\n",
       "      <td>-0.024675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD-PDAD926WRZ</td>\n",
       "      <td>PD-PDAD926WRZ</td>\n",
       "      <td>-0.248133</td>\n",
       "      <td>0.066066</td>\n",
       "      <td>-1.023469</td>\n",
       "      <td>-2.544938</td>\n",
       "      <td>-0.100176</td>\n",
       "      <td>0.210245</td>\n",
       "      <td>0.317189</td>\n",
       "      <td>-0.789634</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.149070</td>\n",
       "      <td>-0.037379</td>\n",
       "      <td>0.457722</td>\n",
       "      <td>-0.070049</td>\n",
       "      <td>-0.139220</td>\n",
       "      <td>0.148462</td>\n",
       "      <td>0.037965</td>\n",
       "      <td>-0.020618</td>\n",
       "      <td>-1.623932</td>\n",
       "      <td>0.020244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19687 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             FID            IID  ENSG00000000003  ENSG00000000005  \\\n",
       "0  PD-PDAA503EF5  PD-PDAA503EF5         1.412056        -0.523395   \n",
       "1  PD-PDAB074CYQ  PD-PDAB074CYQ        -0.856754         0.111547   \n",
       "2  PD-PDAB549YWB  PD-PDAB549YWB         1.118373         0.096341   \n",
       "3  PD-PDAB762PA3  PD-PDAB762PA3        -0.816394        -0.318743   \n",
       "4  PD-PDAD926WRZ  PD-PDAD926WRZ        -0.248133         0.066066   \n",
       "\n",
       "   ENSG00000000419  ENSG00000000457  ENSG00000000460  ENSG00000000938  \\\n",
       "0         0.153490        -1.195026        -0.094739         0.278269   \n",
       "1         0.741327         0.725518         1.231583         0.593704   \n",
       "2         0.427937        -1.107550         0.040538        -1.836332   \n",
       "3        -0.632429        -1.268311         0.100997         0.499943   \n",
       "4        -1.023469        -2.544938        -0.100176         0.210245   \n",
       "\n",
       "   ENSG00000000971  ENSG00000001036  ...  ENSG00000285920  ENSG00000285937  \\\n",
       "0        -1.606817        -1.231579  ...        -0.212292         0.085944   \n",
       "1        -1.066773         0.544353  ...        -0.364937         0.051746   \n",
       "2        -0.428239         1.349854  ...        -0.305483        -0.013945   \n",
       "3        -0.195359         1.295541  ...        -0.075632         0.096710   \n",
       "4         0.317189        -0.789634  ...        -0.149070        -0.037379   \n",
       "\n",
       "   ENSG00000285938  ENSG00000285946  ENSG00000285947  ENSG00000285950  \\\n",
       "0        -0.274379        -0.065663        -0.126903         0.033090   \n",
       "1        -0.786312        -0.003344        -0.166982        -0.059010   \n",
       "2        -0.382129        -0.018791        -0.152562         0.084933   \n",
       "3         0.624501         0.031793        -0.143065         0.120146   \n",
       "4         0.457722        -0.070049        -0.139220         0.148462   \n",
       "\n",
       "   ENSG00000285953  ENSG00000285975  ENSG00000285976  ENSG00000285982  \n",
       "0        -0.499505         0.059806        -1.298476         0.069994  \n",
       "1        -0.499505         0.090575        -2.259028         0.097690  \n",
       "2        -0.499505         0.063530        -0.738222         0.053944  \n",
       "3        -0.499505        -0.050857        -0.730364        -0.024675  \n",
       "4         0.037965        -0.020618        -1.623932         0.020244  \n",
       "\n",
       "[5 rows x 19687 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open normalized expression file PDBP_and_OG_PPMI_Combined_Clinical_Omics.csv\n",
    "\n",
    "expre_df = pd.read_csv(f\"{datadir}/PDBP_and_OG_PPMI_Transcriptomics_ALLTRANSCRIPTS_diffExp_ProteinCoding_ADJUSTED10PCs.csv\", engine='c')\n",
    "\n",
    "#count how manyID -- total number, both case and control\n",
    "print(\"Number of rows in expression data:\", expre_df.shape[0])\n",
    "\n",
    "#add PHENO column on omics_df by matching participant_id_x with ID \n",
    "expression = expre_df.merge(pheno_id, on='ID', how='left')\n",
    "\n",
    "#leave only PHENO=1 to only include only cases, not controls\n",
    "expression_case = expression.loc[expression['PHENO'] == 1.0]\n",
    "\n",
    "#drop Pheno column\n",
    "expression_case = expression_case.drop(['PHENO'], axis=1)\n",
    "\n",
    "#grab ID column and duplicate into two columns to compare with .fam file FID IID\n",
    "expression_case = expression_case.rename(columns={'ID':'FID'})\n",
    "\n",
    "#rearrange Gene_Symbol, Chr, Start, Stop columns right nextg to index column \n",
    "expression_case.insert(1, 'IID', expression_case['FID'])\n",
    "\n",
    "expression_case = expression_case.reset_index(drop=True)\n",
    "print('shape of case only expression data:', expression_case.shape)\n",
    "expression_case.head()\n",
    "\n",
    "#case only expression dataset is ready to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <th>ENSG00000233750</th>\n",
       "      <th>ENSG00000268903</th>\n",
       "      <th>ENSG00000241860</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>380928</td>\n",
       "      <td>380928</td>\n",
       "      <td>0.221062</td>\n",
       "      <td>-1.095390</td>\n",
       "      <td>-1.396011</td>\n",
       "      <td>-0.673701</td>\n",
       "      <td>0.562634</td>\n",
       "      <td>-0.340576</td>\n",
       "      <td>0.674828</td>\n",
       "      <td>-0.619243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>198723</td>\n",
       "      <td>198723</td>\n",
       "      <td>-2.083293</td>\n",
       "      <td>0.581032</td>\n",
       "      <td>2.296874</td>\n",
       "      <td>-2.466542</td>\n",
       "      <td>0.197751</td>\n",
       "      <td>0.894726</td>\n",
       "      <td>1.523388</td>\n",
       "      <td>0.032913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>109803</td>\n",
       "      <td>109803</td>\n",
       "      <td>-1.420241</td>\n",
       "      <td>1.381628</td>\n",
       "      <td>-0.035638</td>\n",
       "      <td>-1.426041</td>\n",
       "      <td>1.256708</td>\n",
       "      <td>-0.041527</td>\n",
       "      <td>0.470557</td>\n",
       "      <td>0.738527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1987391</td>\n",
       "      <td>1987391</td>\n",
       "      <td>-0.203393</td>\n",
       "      <td>-0.052965</td>\n",
       "      <td>-2.009493</td>\n",
       "      <td>-0.743823</td>\n",
       "      <td>0.337753</td>\n",
       "      <td>-0.270654</td>\n",
       "      <td>-1.097861</td>\n",
       "      <td>0.042117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>123414</td>\n",
       "      <td>123414</td>\n",
       "      <td>-2.099258</td>\n",
       "      <td>0.065376</td>\n",
       "      <td>0.069495</td>\n",
       "      <td>-0.238382</td>\n",
       "      <td>-2.003622</td>\n",
       "      <td>0.191987</td>\n",
       "      <td>-0.049066</td>\n",
       "      <td>-0.999551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       FID      IID  ENSG00000223972  ENSG00000227232  ENSG00000237613  \\\n",
       "0   380928   380928         0.221062        -1.095390        -1.396011   \n",
       "1   198723   198723        -2.083293         0.581032         2.296874   \n",
       "2   109803   109803        -1.420241         1.381628        -0.035638   \n",
       "3  1987391  1987391        -0.203393        -0.052965        -2.009493   \n",
       "4   123414   123414        -2.099258         0.065376         0.069495   \n",
       "\n",
       "   ENSG00000238009  ENSG00000239945  ENSG00000233750  ENSG00000268903  \\\n",
       "0        -0.673701         0.562634        -0.340576         0.674828   \n",
       "1        -2.466542         0.197751         0.894726         1.523388   \n",
       "2        -1.426041         1.256708        -0.041527         0.470557   \n",
       "3        -0.743823         0.337753        -0.270654        -1.097861   \n",
       "4        -0.238382        -2.003622         0.191987        -0.049066   \n",
       "\n",
       "   ENSG00000241860  \n",
       "0        -0.619243  \n",
       "1         0.032913  \n",
       "2         0.738527  \n",
       "3         0.042117  \n",
       "4        -0.999551  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Example_phenotype_file.txt from github\n",
    "example = 'https://raw.githubusercontent.com/opain/Calculating-FUSION-TWAS-weights-pipeline/master/Example_phenotype_file.txt'\n",
    "df_example = pd.read_csv(example, sep='\\s+')\n",
    "df_example.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to //data/songy4/twas/data_folder/PCA.log.\n",
      "Options in effect:\n",
      "  --bfile //data/songy4/twas/data_folder/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI\n",
      "  --out //data/songy4/twas/data_folder/PCA\n",
      "  --pca 10\n",
      "\n",
      "1547809 MB RAM detected; reserving 773904 MB for main workspace.\n",
      "23858199 variants loaded from .bim file.\n",
      "1997 people (1191 males, 806 females) loaded from .fam.\n",
      "1844 phenotype values loaded from .fam.\n",
      "Using up to 143 threads (change this with --threads).\n",
      "Before main variant filters, 1997 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Warning: 764392 het. haploid genotypes present (see\n",
      "//data/songy4/twas/data_folder/PCA.hh ); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999728.\n",
      "23858199 variants and 1997 people pass filters and QC.\n",
      "Among remaining phenotypes, 1231 are cases and 613 are controls.  (153\n",
      "phenotypes are missing.)\n",
      "Excluding 797420 variants on non-autosomes from relationship matrix calc.\n",
      "Relationship matrix calculation complete.\n",
      "--pca: Results saved to //data/songy4/twas/data_folder/PCA.eigenval and\n",
      "//data/songy4/twas/data_folder/PCA.eigenvec .\n"
     ]
    }
   ],
   "source": [
    "#extract genetic PCs from expression data \n",
    "!plink --bfile /{datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI --pca 10 --out /{datadir}/PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>PC1_gen</th>\n",
       "      <th>PC2_gen</th>\n",
       "      <th>PC3_gen</th>\n",
       "      <th>PC4_gen</th>\n",
       "      <th>PC5_gen</th>\n",
       "      <th>PC6_gen</th>\n",
       "      <th>PC7_gen</th>\n",
       "      <th>PC8_gen</th>\n",
       "      <th>PC9_gen</th>\n",
       "      <th>PC10_gen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD-PDAA503EF5</td>\n",
       "      <td>0.002470</td>\n",
       "      <td>-0.003959</td>\n",
       "      <td>-0.001653</td>\n",
       "      <td>0.000744</td>\n",
       "      <td>0.000853</td>\n",
       "      <td>-0.000336</td>\n",
       "      <td>0.000615</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>-0.000596</td>\n",
       "      <td>0.000886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD-PDAA852XL7</td>\n",
       "      <td>0.003624</td>\n",
       "      <td>-0.006965</td>\n",
       "      <td>-0.004234</td>\n",
       "      <td>-0.000906</td>\n",
       "      <td>0.001070</td>\n",
       "      <td>0.000792</td>\n",
       "      <td>0.001020</td>\n",
       "      <td>-0.001415</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.000866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD-PDAA955TCY</td>\n",
       "      <td>0.003665</td>\n",
       "      <td>-0.005951</td>\n",
       "      <td>-0.003591</td>\n",
       "      <td>0.000728</td>\n",
       "      <td>0.000924</td>\n",
       "      <td>-0.000524</td>\n",
       "      <td>-0.001126</td>\n",
       "      <td>-0.001011</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>-0.000057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD-PDAB074CYQ</td>\n",
       "      <td>0.003028</td>\n",
       "      <td>-0.005959</td>\n",
       "      <td>-0.004660</td>\n",
       "      <td>-0.001468</td>\n",
       "      <td>-0.001162</td>\n",
       "      <td>-0.000704</td>\n",
       "      <td>-0.005008</td>\n",
       "      <td>0.000753</td>\n",
       "      <td>0.000317</td>\n",
       "      <td>0.000158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD-PDAB411CTU</td>\n",
       "      <td>0.002395</td>\n",
       "      <td>-0.004831</td>\n",
       "      <td>-0.002924</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000973</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000754</td>\n",
       "      <td>-0.000845</td>\n",
       "      <td>0.000794</td>\n",
       "      <td>0.002811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1992</th>\n",
       "      <td>PP-4126</td>\n",
       "      <td>0.003207</td>\n",
       "      <td>-0.004924</td>\n",
       "      <td>-0.003516</td>\n",
       "      <td>0.000243</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>-0.000103</td>\n",
       "      <td>-0.001389</td>\n",
       "      <td>-0.000686</td>\n",
       "      <td>0.000260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1993</th>\n",
       "      <td>PP-4127</td>\n",
       "      <td>0.000741</td>\n",
       "      <td>-0.000308</td>\n",
       "      <td>-0.000660</td>\n",
       "      <td>0.001759</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>-0.000244</td>\n",
       "      <td>0.001068</td>\n",
       "      <td>-0.001482</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>-0.000034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1994</th>\n",
       "      <td>PP-4135</td>\n",
       "      <td>-0.021394</td>\n",
       "      <td>0.058516</td>\n",
       "      <td>0.051938</td>\n",
       "      <td>0.014075</td>\n",
       "      <td>0.000885</td>\n",
       "      <td>0.002686</td>\n",
       "      <td>0.012527</td>\n",
       "      <td>-0.005588</td>\n",
       "      <td>0.007664</td>\n",
       "      <td>0.001302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1995</th>\n",
       "      <td>PP-4136</td>\n",
       "      <td>0.003769</td>\n",
       "      <td>-0.006528</td>\n",
       "      <td>-0.003886</td>\n",
       "      <td>-0.000414</td>\n",
       "      <td>0.001516</td>\n",
       "      <td>-0.000477</td>\n",
       "      <td>-0.002527</td>\n",
       "      <td>0.000252</td>\n",
       "      <td>-0.001974</td>\n",
       "      <td>0.000660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1996</th>\n",
       "      <td>PP-4139</td>\n",
       "      <td>-0.022274</td>\n",
       "      <td>0.058211</td>\n",
       "      <td>0.052003</td>\n",
       "      <td>0.016919</td>\n",
       "      <td>0.000807</td>\n",
       "      <td>0.002815</td>\n",
       "      <td>0.012435</td>\n",
       "      <td>-0.003320</td>\n",
       "      <td>0.008108</td>\n",
       "      <td>0.001583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1997 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ID   PC1_gen   PC2_gen   PC3_gen   PC4_gen   PC5_gen  \\\n",
       "0     PD-PDAA503EF5  0.002470 -0.003959 -0.001653  0.000744  0.000853   \n",
       "1     PD-PDAA852XL7  0.003624 -0.006965 -0.004234 -0.000906  0.001070   \n",
       "2     PD-PDAA955TCY  0.003665 -0.005951 -0.003591  0.000728  0.000924   \n",
       "3     PD-PDAB074CYQ  0.003028 -0.005959 -0.004660 -0.001468 -0.001162   \n",
       "4     PD-PDAB411CTU  0.002395 -0.004831 -0.002924  0.000033  0.000973   \n",
       "...             ...       ...       ...       ...       ...       ...   \n",
       "1992        PP-4126  0.003207 -0.004924 -0.003516  0.000243  0.000029   \n",
       "1993        PP-4127  0.000741 -0.000308 -0.000660  0.001759  0.000649   \n",
       "1994        PP-4135 -0.021394  0.058516  0.051938  0.014075  0.000885   \n",
       "1995        PP-4136  0.003769 -0.006528 -0.003886 -0.000414  0.001516   \n",
       "1996        PP-4139 -0.022274  0.058211  0.052003  0.016919  0.000807   \n",
       "\n",
       "       PC6_gen   PC7_gen   PC8_gen   PC9_gen  PC10_gen  \n",
       "0    -0.000336  0.000615  0.000049 -0.000596  0.000886  \n",
       "1     0.000792  0.001020 -0.001415  0.000061  0.000866  \n",
       "2    -0.000524 -0.001126 -0.001011  0.000473 -0.000057  \n",
       "3    -0.000704 -0.005008  0.000753  0.000317  0.000158  \n",
       "4     0.000205  0.000754 -0.000845  0.000794  0.002811  \n",
       "...        ...       ...       ...       ...       ...  \n",
       "1992  0.000137 -0.000103 -0.001389 -0.000686  0.000260  \n",
       "1993 -0.000244  0.001068 -0.001482  0.000012 -0.000034  \n",
       "1994  0.002686  0.012527 -0.005588  0.007664  0.001302  \n",
       "1995 -0.000477 -0.002527  0.000252 -0.001974  0.000660  \n",
       "1996  0.002815  0.012435 -0.003320  0.008108  0.001583  \n",
       "\n",
       "[1997 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read genetic PCs\n",
    "pcs_df = pd.read_csv(f\"{datadir}/PCA.eigenvec\", sep=' ', usecols=[1,2,3,4,5,6,7,8,9,10,11], names=['ID','PC1_gen','PC2_gen','PC3_gen','PC4_gen','PC5_gen','PC6_gen','PC7_gen','PC8_gen','PC9_gen','PC10_gen'])\n",
    "pcs_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create fid_df.txt file to remove rows in .bim file if FID is missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of case only ID: (1138, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD-PDAA503EF5</td>\n",
       "      <td>PD-PDAA503EF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD-PDAB074CYQ</td>\n",
       "      <td>PD-PDAB074CYQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD-PDAB549YWB</td>\n",
       "      <td>PD-PDAB549YWB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD-PDAB762PA3</td>\n",
       "      <td>PD-PDAB762PA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD-PDAD926WRZ</td>\n",
       "      <td>PD-PDAD926WRZ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             FID            IID\n",
       "0  PD-PDAA503EF5  PD-PDAA503EF5\n",
       "1  PD-PDAB074CYQ  PD-PDAB074CYQ\n",
       "2  PD-PDAB549YWB  PD-PDAB549YWB\n",
       "3  PD-PDAB762PA3  PD-PDAB762PA3\n",
       "4  PD-PDAD926WRZ  PD-PDAD926WRZ"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#grab FID and IID to compare with .fam file FID IID\n",
    "case_id_df= expression_case[['FID', 'IID']].copy()\n",
    "print('shape of case only ID:', case_id_df.shape)\n",
    "case_id_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save id_df as txt  -- run only once for download\n",
    "case_id_df.to_csv(r'./data_folder/case_id_df.txt', sep='\\t' ,index=False)\n",
    "#id_df.to_csv(r'./data_folder/fid_df.csv' ,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plink file preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.bed\n",
      "AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.bim\n",
      "AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.fam\n",
      "AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.hh\n",
      "AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.log\n",
      "PCs\n"
     ]
    }
   ],
   "source": [
    "#look into genotype data folder to grab PLINK files\n",
    "! ls /data/LNG/makariousmb/Projects/GenoMLxAMP/clean_wgs/PDBP_AND_OG_PPMI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy bed bim fam files into data folder - run only once for download\n",
    "#these plink files are genotic data (genotype)\n",
    "! cp -R /data/LNG/makariousmb/Projects/GenoMLxAMP/clean_wgs/PDBP_AND_OG_PPMI/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.fam /{datadir}/\n",
    "! cp -R /data/LNG/makariousmb/Projects/GenoMLxAMP/clean_wgs/PDBP_AND_OG_PPMI/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.bim /{datadir}/\n",
    "! cp -R /data/LNG/makariousmb/Projects/GenoMLxAMP/clean_wgs/PDBP_AND_OG_PPMI/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.bed /{datadir}/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PD-PDAA503EF5 PD-PDAA503EF5 0 0 2 2\n",
      "PD-PDAA852XL7 PD-PDAA852XL7 0 0 1 -9\n",
      "PD-PDAA955TCY PD-PDAA955TCY 0 0 2 1\n",
      "PD-PDAB074CYQ PD-PDAB074CYQ 0 0 1 2\n",
      "PD-PDAB411CTU PD-PDAB411CTU 0 0 1 1\n",
      "PD-PDAB549YWB PD-PDAB549YWB 0 0 1 2\n",
      "PD-PDAB597LKL PD-PDAB597LKL 0 0 2 -9\n",
      "PD-PDAB729HWD PD-PDAB729HWD 0 0 1 2\n",
      "PD-PDAB762PA3 PD-PDAB762PA3 0 0 2 2\n",
      "PD-PDAB803VAR PD-PDAB803VAR 0 0 1 1\n"
     ]
    }
   ],
   "source": [
    "#check number of IDs in original .fam file\n",
    "!cat {datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.fam | wc -l\n",
    "\n",
    "#look at the fam file (fid, iid, mid, pid, gender, affection status)\n",
    "! head /data/LNG/makariousmb/Projects/GenoMLxAMP/clean_wgs/PDBP_AND_OG_PPMI/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.fam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23858199\n",
      "1\trs145427775\t0\t10291\tT\tC\n",
      "1\trs55998931\t0\t10492\tT\tC\n",
      "1\trs199896944\t0\t13504\tA\tG\n",
      "1\trs199856693\t0\t14933\tA\tG\n",
      "1\trs201855936\t0\t14948\tA\tG\n",
      "1\trs71252251\t0\t14976\tA\tG\n",
      "1\trs201045431\t0\t15029\tA\tG\n",
      "1\trs368345873\t0\t15208\tA\tG\n",
      "1\trs374029747\t0\t15774\tA\tG\n",
      "1\trs201330479\t0\t16792\tA\tG\n"
     ]
    }
   ],
   "source": [
    "#check number of IDs in original .bim file\n",
    "!cat {datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.bim | wc -l\n",
    "\n",
    "#look at the bim file (variance, chromosome and position, allele1 and allele2)\n",
    "! head {datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.bim "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /data/songy4/TWAS/data/id_comparison/qc_genotypes_twas.log.\n",
      "Options in effect:\n",
      "  --bfile /data/LNG/makariousmb/Projects/GenoMLxAMP/clean_wgs/PDBP_AND_OG_PPMI/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI\n",
      "  --keep /data/songy4/TWAS/data/id_df.txt\n",
      "  --make-bed\n",
      "  --out /data/songy4/TWAS/data/id_comparison/qc_genotypes_twas\n",
      "\n",
      "386449 MB RAM detected; reserving 193224 MB for main workspace.\n",
      "23858199 variants loaded from .bim file.\n",
      "1997 people (1191 males, 806 females) loaded from .fam.\n",
      "1844 phenotype values loaded from .fam.\n",
      "--keep: 1138 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1138 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Warning: 579385 het. haploid genotypes present (see\n",
      "/data/songy4/TWAS/data/id_comparison/qc_genotypes_twas.hh ); many commands\n",
      "treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate in remaining samples is 0.999725.\n",
      "23858199 variants and 1138 people pass filters and QC.\n",
      "Among remaining phenotypes, 1138 are cases and 0 are controls.\n",
      "--make-bed to /data/songy4/TWAS/data/id_comparison/qc_genotypes_twas.bed +\n",
      "/data/songy4/TWAS/data/id_comparison/qc_genotypes_twas.bim +\n",
      "/data/songy4/TWAS/data/id_comparison/qc_genotypes_twas.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "#filter sample id using fid_df.txt to reduce rows that are not in fid_df.txt\n",
    "##plink --bfile <genotype file> --keep <id file> --make-bed --out <outfile>\n",
    "!plink --bfile /{datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI --keep /{datadir}/case_id_df.txt  --make-bed --out /{datadir}/qc_genotypes_twas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#check the number of IDs in .fam file to make sure it is same numbers as number of rows in exp_df\n",
    "!cat /{datadir}/qc_genotypes_twas.fam | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23858199\n",
      "1\trs145427775\t0\t10291\tT\tC\n",
      "1\trs55998931\t0\t10492\tT\tC\n",
      "1\trs199896944\t0\t13504\tA\tG\n",
      "1\trs199856693\t0\t14933\tA\tG\n",
      "1\trs201855936\t0\t14948\tA\tG\n",
      "1\trs71252251\t0\t14976\tA\tG\n",
      "1\trs201045431\t0\t15029\tA\tG\n",
      "1\trs368345873\t0\t15208\tA\tG\n",
      "1\trs374029747\t0\t15774\tA\tG\n",
      "1\trs201330479\t0\t16792\tA\tG\n"
     ]
    }
   ],
   "source": [
    "#check number of IDs in original .bim file\n",
    "!cat {datadir}/qc_genotypes_twas.bim | wc -l\n",
    "\n",
    "#look at the bim file (variance, chromosome and position, allele1 and allele2)\n",
    "! head {datadir}/qc_genotypes_twas.bim "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to //data/songy4/twas/data_folder/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.log.\n",
      "Options in effect:\n",
      "  --bed //data/songy4/twas/data_folder/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.bed\n",
      "  --bfile //data/songy4/twas/data_folder/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI\n",
      "  --freq\n",
      "  --out //data/songy4/twas/data_folder/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI\n",
      "\n",
      "1547809 MB RAM detected; reserving 773904 MB for main workspace.\n",
      "23858199 variants loaded from .bim file.\n",
      "1997 people (1191 males, 806 females) loaded from .fam.\n",
      "1844 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1997 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Warning: 764392 het. haploid genotypes present (see\n",
      "//data/songy4/twas/data_folder/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.hh\n",
      "); many commands treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999728.\n",
      "--freq: Allele frequencies (founders only) written to\n",
      "//data/songy4/twas/data_folder/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.frq\n",
      ".\n"
     ]
    }
   ],
   "source": [
    "#get frequency from plink file -- original plink files (AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI)\n",
    "!plink --bfile /{datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI --bed /{datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.bed --freq --out /{datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " CHR           SNP   A1   A2          MAF  NCHROBS\n",
      "   1   rs145427775    T    C     0.003518     3980\n",
      "   1    rs55998931    T    C      0.03735     3962\n",
      "   1   rs199896944    A    G    0.0007784     3854\n",
      "   1   rs199856693    A    G      0.06831     3982\n",
      "   1   rs201855936    A    G      0.01882     3986\n",
      "   1    rs71252251    A    G      0.02851     3964\n",
      "   1   rs201045431    A    G     0.001265     3954\n",
      "   1   rs368345873    A    G     0.004082     3920\n",
      "   1   rs374029747    A    G     0.006513     3992\n"
     ]
    }
   ],
   "source": [
    "#read frq file from original plink files (AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI)\n",
    "! head {datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.frq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "#count if MAF is bigger than 0.5\n",
    "!cat {datadir}/AMP_Euro_sampleQC_variantQC_FINAL_PDBP_AND_OG_PPMI.frq | awk -F\"\\t\" '$5>0.5' | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to plink.log.\n",
      "Options in effect:\n",
      "  --bed //data/songy4/twas/data_folder/qc_genotypes_twas.bed\n",
      "  --bfile //data/songy4/twas/data_folder/qc_genotypes_twas\n",
      "  --freq\n",
      "\n",
      "1547809 MB RAM detected; reserving 773904 MB for main workspace.\n",
      "23858199 variants loaded from .bim file.\n",
      "1138 people (730 males, 408 females) loaded from .fam.\n",
      "1138 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1138 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Warning: 579385 het. haploid genotypes present (see plink.hh ); many commands\n",
      "treat these as missing.\n",
      "Warning: Nonmissing nonmale Y chromosome genotype(s) present; many commands\n",
      "treat these as missing.\n",
      "Total genotyping rate is 0.999725.\n",
      "--freq: Allele frequencies (founders only) written to plink.frq .\n"
     ]
    }
   ],
   "source": [
    "#get frequency from plink file -- qc_genotypes_twas\n",
    "!plink --bfile /{datadir}/qc_genotypes_twas  --bed /{datadir}/qc_genotypes_twas.bed --freq --out /{datadir}/qc_genotypes_twas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " CHR           SNP   A1   A2          MAF  NCHROBS\n",
      "   1   rs145427775    T    C     0.003521     2272\n",
      "   1    rs55998931    T    C      0.02396     2254\n",
      "   1   rs199896944    A    G    0.0004558     2194\n",
      "   1   rs199856693    A    G      0.06399     2266\n",
      "   1   rs201855936    A    G      0.01938     2270\n",
      "   1    rs71252251    A    G      0.02396     2254\n",
      "   1   rs201045431    A    G    0.0008905     2246\n",
      "   1   rs368345873    A    G     0.004061     2216\n",
      "   1   rs374029747    A    G     0.004837     2274\n"
     ]
    }
   ],
   "source": [
    "#read frq file from qc_genotypes_twas plink files\n",
    "! head /{datadir}/qc_genotypes_twas.frq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "#count if MAF is bigger than 0.5\n",
    "!cat /{datadir}/qc_genotypes_twas.frq | awk -F\"\\t\" '$5>0.5' | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v1.90b4.4 64-bit (21 May 2017)           www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2017 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to plink.log.\n",
      "Options in effect:\n",
      "  --bed //data/songy4/twas/data_folder/qc_genotypes_twas_hg19_lifted.bed\n",
      "  --bfile //data/songy4/twas/data_folder/qc_genotypes_twas_hg19_lifted\n",
      "  --freq\n",
      "\n",
      "1547809 MB RAM detected; reserving 773904 MB for main workspace.\n",
      "23060234 variants loaded from .bim file.\n",
      "1138 people (730 males, 408 females) loaded from .fam.\n",
      "1138 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 1138 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.999731.\n",
      "--freq: Allele frequencies (founders only) written to plink.frq .\n"
     ]
    }
   ],
   "source": [
    "#get frequency from plink file -- liftover version\n",
    "!plink --bfile /{datadir}/qc_genotypes_twas_hg19_lifted  --bed /{datadir}/qc_genotypes_twas_hg19_lifted.bed --freq --out /{datadir}/qc_genotypes_twas_hg19_lifted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " CHR           SNP   A1   A2          MAF  NCHROBS\n",
      "   1   rs145427775    T    C     0.003521     2272\n",
      "   1    rs55998931    T    C      0.02396     2254\n",
      "   1   rs199896944    A    G    0.0004558     2194\n",
      "   1   rs199856693    A    G      0.06399     2266\n",
      "   1   rs201855936    A    G      0.01938     2270\n",
      "   1    rs71252251    A    G      0.02396     2254\n",
      "   1   rs201045431    A    G    0.0008905     2246\n",
      "   1   rs368345873    A    G     0.004061     2216\n",
      "   1   rs374029747    A    G     0.004837     2274\n"
     ]
    }
   ],
   "source": [
    "#read frq file from qc_genotypes_twas plink files\n",
    "! head /{datadir}/qc_genotypes_twas_hg19_lifted.frq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "#count if MAF is bigger than 0.5 --> this is to check if minor allele is A1\n",
    "!cat /{datadir}/qc_genotypes_twas_hg19_lifted.frq | awk -F\"\\t\" '$5>0.5' | wc -l\n",
    "\n",
    "###if MAF value is bigger than 0.5 that is major allele and those alleles (A1 and A2) need to be flipped using plink --reference-allele\n",
    "### https://zzz.bwh.harvard.edu/plink/reference.shtml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Covariate data preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of covariate data: (3941, 7)\n",
      "              ID  AGE  MALE  EDUCATION  FAMILY_HISTORY  UPSIT  InfAJ\n",
      "0  PD-PDNE299YPT   74     1        2.0             0.0   16.0      0\n",
      "1  PD-PDYW828VAV   57     0        1.0             1.0   31.0      0\n",
      "2  PD-PDWV958JPG   68     1        2.0             0.0   11.0      0\n",
      "3  PD-PDRF387ZV3   68     1        1.0             0.0   28.0      0\n",
      "4  PD-PDAC268KWV   69     0        2.0             0.0   24.0      0\n",
      "shape of covariate data: (1138, 7)\n",
      "               ID  AGE  MALE  EDUCATION  FAMILY_HISTORY  UPSIT  InfAJ\n",
      "29  PD-PDKG339XM0   56     0        2.0             1.0   17.0      0\n",
      "30  PD-PDRM726FKG   63     1        2.0             0.0   18.0      0\n",
      "33  PD-PDEM487YZD   51     1        1.0             0.0   28.0      0\n",
      "34  PD-PDVW658HP7   70     1        2.0             1.0   28.0      0\n",
      "35  PD-PDVJ880MZU   66     1        2.0             0.0   13.0      0\n"
     ]
    }
   ],
   "source": [
    "#open covariate file clinicodemographic_3nov2020.csv\n",
    "covariate = pd.read_csv(f\"{datadir}/clinicodemographic_3nov2020.csv\", index_col=False )\n",
    "\n",
    "print('shape of covariate data:', covariate.shape)\n",
    "print(covariate.head())\n",
    "\n",
    "#drop cov_df ID not in exp_case ID\n",
    "covariate_case = covariate[covariate['ID'].isin(expression_case['FID'])]\n",
    "\n",
    "print('shape of covariate data:', covariate_case.shape)\n",
    "print(covariate_case.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of covariate data: (1138, 18)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>AGE</th>\n",
       "      <th>MALE</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>UPSIT</th>\n",
       "      <th>InfAJ</th>\n",
       "      <th>PC1_gen</th>\n",
       "      <th>PC2_gen</th>\n",
       "      <th>PC3_gen</th>\n",
       "      <th>PC4_gen</th>\n",
       "      <th>PC5_gen</th>\n",
       "      <th>PC6_gen</th>\n",
       "      <th>PC7_gen</th>\n",
       "      <th>PC8_gen</th>\n",
       "      <th>PC9_gen</th>\n",
       "      <th>PC10_gen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD-PDKG339XM0</td>\n",
       "      <td>PD-PDKG339XM0</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.002876</td>\n",
       "      <td>-0.007443</td>\n",
       "      <td>-0.005085</td>\n",
       "      <td>-0.000187</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>-0.000787</td>\n",
       "      <td>-0.002445</td>\n",
       "      <td>-0.000240</td>\n",
       "      <td>-0.001125</td>\n",
       "      <td>0.000828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD-PDRM726FKG</td>\n",
       "      <td>PD-PDRM726FKG</td>\n",
       "      <td>63</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>-0.007768</td>\n",
       "      <td>-0.005870</td>\n",
       "      <td>-0.000552</td>\n",
       "      <td>0.001536</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>-0.000336</td>\n",
       "      <td>0.001626</td>\n",
       "      <td>0.001068</td>\n",
       "      <td>0.001081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD-PDEM487YZD</td>\n",
       "      <td>PD-PDEM487YZD</td>\n",
       "      <td>51</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003569</td>\n",
       "      <td>-0.007966</td>\n",
       "      <td>-0.005733</td>\n",
       "      <td>-0.001055</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>-0.000491</td>\n",
       "      <td>0.001775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD-PDVW658HP7</td>\n",
       "      <td>PD-PDVW658HP7</td>\n",
       "      <td>70</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.005395</td>\n",
       "      <td>-0.010209</td>\n",
       "      <td>-0.007062</td>\n",
       "      <td>-0.002003</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>0.001092</td>\n",
       "      <td>-0.000483</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>-0.001346</td>\n",
       "      <td>0.000829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD-PDVJ880MZU</td>\n",
       "      <td>PD-PDVJ880MZU</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001514</td>\n",
       "      <td>-0.003853</td>\n",
       "      <td>-0.001543</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>-0.000692</td>\n",
       "      <td>-0.000884</td>\n",
       "      <td>-0.000402</td>\n",
       "      <td>0.000285</td>\n",
       "      <td>0.000432</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             FID            IID  AGE  MALE  EDUCATION  FAMILY_HISTORY  UPSIT  \\\n",
       "0  PD-PDKG339XM0  PD-PDKG339XM0   56     0        2.0             1.0   17.0   \n",
       "1  PD-PDRM726FKG  PD-PDRM726FKG   63     1        2.0             0.0   18.0   \n",
       "2  PD-PDEM487YZD  PD-PDEM487YZD   51     1        1.0             0.0   28.0   \n",
       "3  PD-PDVW658HP7  PD-PDVW658HP7   70     1        2.0             1.0   28.0   \n",
       "4  PD-PDVJ880MZU  PD-PDVJ880MZU   66     1        2.0             0.0   13.0   \n",
       "\n",
       "   InfAJ   PC1_gen   PC2_gen   PC3_gen   PC4_gen   PC5_gen   PC6_gen  \\\n",
       "0      0  0.002876 -0.007443 -0.005085 -0.000187  0.000608 -0.000787   \n",
       "1      0  0.003366 -0.007768 -0.005870 -0.000552  0.001536  0.000039   \n",
       "2      0  0.003569 -0.007966 -0.005733 -0.001055  0.000629  0.000100   \n",
       "3      0  0.005395 -0.010209 -0.007062 -0.002003  0.001324  0.001092   \n",
       "4      0  0.001514 -0.003853 -0.001543  0.000162  0.000673 -0.000692   \n",
       "\n",
       "    PC7_gen   PC8_gen   PC9_gen  PC10_gen  \n",
       "0 -0.002445 -0.000240 -0.001125  0.000828  \n",
       "1 -0.000336  0.001626  0.001068  0.001081  \n",
       "2  0.000309  0.000120 -0.000491  0.001775  \n",
       "3 -0.000483  0.000103 -0.001346  0.000829  \n",
       "4 -0.000884 -0.000402  0.000285  0.000432  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge pcs_df into covariate_df\n",
    "covar_df_case = covariate_case.merge(pcs_df, on='ID', how='left')\n",
    "\n",
    "#rename ID to FID\n",
    "covariate_case = covar_df_case.rename(columns={'ID':'FID'})\n",
    "\n",
    "#rearrange Gene_Symbol, Chr, Start, Stop columns right nextg to index column \n",
    "covariate_case.insert(1, 'IID', covariate_case['FID'])\n",
    "\n",
    "covariate_case = covariate_case.reset_index(drop=True)\n",
    "print('shape of covariate data:', covariate_case.shape)\n",
    "\n",
    "covariate_case.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique value of AGE: [56 63 51 70 66 41 71 78 35 67 43 58 77 72 62 53 69 65 52 50 75 60 68 61\n",
      " 59 80 86 73 81 74 48 46 64 55 54 57 87 49 76 79 45 85 42 82 84 34 38 44\n",
      " 37 47 83 39 40]\n",
      "Unique value of MALE: [0 1]\n",
      "Unique value of EDUCATION: [2. 1. 0.]\n",
      "Unique value of FAMILY_HISTORY: [ 1.  0. nan]\n",
      "Unique value of UPSIT: [17. 18. 28. 13. 23. 32. 29. 21.  6.  9. 31. 15. 34. 14. 22. nan 25. 40.\n",
      " 16. 33. 20. 19. 37. 10. 11. 12. 26.  8. 36. 35. 24. 30.  7.  4. 27.  1.\n",
      " 39. 38.  5.]\n",
      "Unique value of InfAJ: [0 1]\n",
      "null value in each column: \n",
      " FID                0\n",
      "IID                0\n",
      "AGE                0\n",
      "MALE               0\n",
      "EDUCATION          0\n",
      "FAMILY_HISTORY     4\n",
      "UPSIT             56\n",
      "InfAJ              0\n",
      "PC1_gen            0\n",
      "PC2_gen            0\n",
      "PC3_gen            0\n",
      "PC4_gen            0\n",
      "PC5_gen            0\n",
      "PC6_gen            0\n",
      "PC7_gen            0\n",
      "PC8_gen            0\n",
      "PC9_gen            0\n",
      "PC10_gen           0\n",
      "dtype: int64\n",
      "data types in each column: \n",
      " FID                object\n",
      "IID                object\n",
      "AGE                 int64\n",
      "MALE                int64\n",
      "EDUCATION         float64\n",
      "FAMILY_HISTORY    float64\n",
      "UPSIT             float64\n",
      "InfAJ               int64\n",
      "PC1_gen           float64\n",
      "PC2_gen           float64\n",
      "PC3_gen           float64\n",
      "PC4_gen           float64\n",
      "PC5_gen           float64\n",
      "PC6_gen           float64\n",
      "PC7_gen           float64\n",
      "PC8_gen           float64\n",
      "PC9_gen           float64\n",
      "PC10_gen          float64\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "#check unique values in each column\n",
    "for i in covariate_case.columns[2:8]:\n",
    "    print('Unique value of ' + i + ':', covariate_case[i].unique())\n",
    "\n",
    "#check nan numbers in each column \n",
    "print('null value in each column: \\n', covariate_case.isna().sum())\n",
    "\n",
    "#check data type in all column\n",
    "print('data types in each column: \\n',covariate_case.dtypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null value in each column: \n",
      " FID               0\n",
      "IID               0\n",
      "AGE               0\n",
      "MALE              0\n",
      "EDUCATION         0\n",
      "FAMILY_HISTORY    0\n",
      "UPSIT             0\n",
      "InfAJ             0\n",
      "PC1_gen           0\n",
      "PC2_gen           0\n",
      "PC3_gen           0\n",
      "PC4_gen           0\n",
      "PC5_gen           0\n",
      "PC6_gen           0\n",
      "PC7_gen           0\n",
      "PC8_gen           0\n",
      "PC9_gen           0\n",
      "PC10_gen          0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#replace NaN with mean for nonbinary and median for binary column\n",
    "covariate_case['FAMILY_HISTORY'] = covariate_case['FAMILY_HISTORY'].fillna(covariate_case['FAMILY_HISTORY'].median())\n",
    "covariate_case['UPSIT'] = covariate_case['UPSIT'].fillna(covariate_case['UPSIT'].mean())\n",
    "\n",
    "#check nan numbers in each column \n",
    "print('null value in each column: \\n', covariate_case.isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#change float to int in column EDUCATION, FAMILY_HISTORY, and UPSIT\n",
    "\n",
    "##col=(covariate.dtypes=='float64')\n",
    "col_case=covariate_case.columns[4:7]\n",
    "covariate_case.loc[:,col_case]=covariate_case.loc[:,col_case].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of covariate data: (1138, 18)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>AGE</th>\n",
       "      <th>MALE</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>UPSIT</th>\n",
       "      <th>InfAJ</th>\n",
       "      <th>PC1_gen</th>\n",
       "      <th>PC2_gen</th>\n",
       "      <th>PC3_gen</th>\n",
       "      <th>PC4_gen</th>\n",
       "      <th>PC5_gen</th>\n",
       "      <th>PC6_gen</th>\n",
       "      <th>PC7_gen</th>\n",
       "      <th>PC8_gen</th>\n",
       "      <th>PC9_gen</th>\n",
       "      <th>PC10_gen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD-PDKG339XM0</td>\n",
       "      <td>PD-PDKG339XM0</td>\n",
       "      <td>0.415094</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.410256</td>\n",
       "      <td>0</td>\n",
       "      <td>0.002876</td>\n",
       "      <td>-0.007443</td>\n",
       "      <td>-0.005085</td>\n",
       "      <td>-0.000187</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>-0.000787</td>\n",
       "      <td>-0.002445</td>\n",
       "      <td>-0.000240</td>\n",
       "      <td>-0.001125</td>\n",
       "      <td>0.000828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD-PDRM726FKG</td>\n",
       "      <td>PD-PDRM726FKG</td>\n",
       "      <td>0.547170</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.435897</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>-0.007768</td>\n",
       "      <td>-0.005870</td>\n",
       "      <td>-0.000552</td>\n",
       "      <td>0.001536</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>-0.000336</td>\n",
       "      <td>0.001626</td>\n",
       "      <td>0.001068</td>\n",
       "      <td>0.001081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD-PDEM487YZD</td>\n",
       "      <td>PD-PDEM487YZD</td>\n",
       "      <td>0.320755</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003569</td>\n",
       "      <td>-0.007966</td>\n",
       "      <td>-0.005733</td>\n",
       "      <td>-0.001055</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>-0.000491</td>\n",
       "      <td>0.001775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD-PDVW658HP7</td>\n",
       "      <td>PD-PDVW658HP7</td>\n",
       "      <td>0.679245</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.005395</td>\n",
       "      <td>-0.010209</td>\n",
       "      <td>-0.007062</td>\n",
       "      <td>-0.002003</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>0.001092</td>\n",
       "      <td>-0.000483</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>-0.001346</td>\n",
       "      <td>0.000829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD-PDVJ880MZU</td>\n",
       "      <td>PD-PDVJ880MZU</td>\n",
       "      <td>0.603774</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001514</td>\n",
       "      <td>-0.003853</td>\n",
       "      <td>-0.001543</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>-0.000692</td>\n",
       "      <td>-0.000884</td>\n",
       "      <td>-0.000402</td>\n",
       "      <td>0.000285</td>\n",
       "      <td>0.000432</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             FID            IID       AGE  MALE  EDUCATION  FAMILY_HISTORY  \\\n",
       "0  PD-PDKG339XM0  PD-PDKG339XM0  0.415094     0        1.0               1   \n",
       "1  PD-PDRM726FKG  PD-PDRM726FKG  0.547170     1        1.0               0   \n",
       "2  PD-PDEM487YZD  PD-PDEM487YZD  0.320755     1        0.5               0   \n",
       "3  PD-PDVW658HP7  PD-PDVW658HP7  0.679245     1        1.0               1   \n",
       "4  PD-PDVJ880MZU  PD-PDVJ880MZU  0.603774     1        1.0               0   \n",
       "\n",
       "      UPSIT  InfAJ   PC1_gen   PC2_gen   PC3_gen   PC4_gen   PC5_gen  \\\n",
       "0  0.410256      0  0.002876 -0.007443 -0.005085 -0.000187  0.000608   \n",
       "1  0.435897      0  0.003366 -0.007768 -0.005870 -0.000552  0.001536   \n",
       "2  0.692308      0  0.003569 -0.007966 -0.005733 -0.001055  0.000629   \n",
       "3  0.692308      0  0.005395 -0.010209 -0.007062 -0.002003  0.001324   \n",
       "4  0.307692      0  0.001514 -0.003853 -0.001543  0.000162  0.000673   \n",
       "\n",
       "    PC6_gen   PC7_gen   PC8_gen   PC9_gen  PC10_gen  \n",
       "0 -0.000787 -0.002445 -0.000240 -0.001125  0.000828  \n",
       "1  0.000039 -0.000336  0.001626  0.001068  0.001081  \n",
       "2  0.000100  0.000309  0.000120 -0.000491  0.001775  \n",
       "3  0.001092 -0.000483  0.000103 -0.001346  0.000829  \n",
       "4 -0.000692 -0.000884 -0.000402  0.000285  0.000432  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#normalize covariate values for AGE, UPSIT, EDUCATION columns with min_max_scaling\n",
    "\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "scaler = MinMaxScaler()\n",
    "covariate_scale_case = covariate_case.copy()\n",
    "covariate_scale_case[['AGE', 'UPSIT', 'EDUCATION']] = scaler.fit_transform(covariate_scale_case[['AGE', 'UPSIT', 'EDUCATION']])\n",
    "\n",
    "\n",
    "\n",
    "# apply the min-max scaling in Pandas using the .min() and .max() methods\n",
    "#def min_max_scaling(df):\n",
    "    # copy the dataframe\n",
    " #   df_norm = df.copy()\n",
    "    # apply min-max scaling\n",
    "  #  for column in df_norm.iloc[:, [2, 4, 6]]:\n",
    "   #     df_norm[column] = (df_norm[column] - df_norm[column].min()) / (df_norm[column].max() - df_norm[column].min()) \n",
    "        \n",
    "    #return df_norm\n",
    "    \n",
    "# call the min_max_scaling function\n",
    "#covariate_scale_case = min_max_scaling(covariate_case)\n",
    "print('shape of covariate data:', covariate_scale_case.shape)\n",
    "covariate_scale_case.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save covariate as txt -- run only once for download\n",
    "covariate_scale_case.to_csv(r'./data_folder/covariates_case.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Coordinate data preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of chromosom position data: (57905, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14363</td>\n",
       "      <td>29806</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>52473</td>\n",
       "      <td>54936</td>\n",
       "      <td>ENSG00000268020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  X.Chr  start    end               ID\n",
       "0     1  11869  14412  ENSG00000223972\n",
       "1     1  14363  29806  ENSG00000227232\n",
       "2     1  29554  31109  ENSG00000243485\n",
       "3     1  34554  36081  ENSG00000237613\n",
       "4     1  52473  54936  ENSG00000268020"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open gene_chr_start_stop.tab with column names -- this has ch37 chr and position info\n",
    "##the file \"gene_chr_start_stop.tab\" is in GP2_WG/working groups/CD_DAWG/scratch/example_TWAS\n",
    "chr_pos_df = pd.read_csv(f\"{datadir}/gene_chr_start_stop.tab\", sep='\\t', header=None, index_col=False, names=('ID', 'X.Chr', 'start', 'end'))\n",
    "\n",
    "#rearrange the columns \n",
    "chr_pos_df = chr_pos_df[['X.Chr','start','end','ID']]\n",
    "\n",
    "print('shape of chromosom position data:', chr_pos_df.shape)\n",
    "chr_pos_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unique chromosom: ['1' '2' '3' '4' '5' '6' '7' 'X' '8' '9' '10' '11' '12' '13' '14' '15'\n",
      " '16' '17' '18' '20' '19' 'Y' '22' '21' 'MT' 'GL000192.1' 'GL000225.1'\n",
      " 'GL000194.1' 'GL000193.1' 'GL000222.1' 'GL000212.1' 'GL000195.1'\n",
      " 'GL000223.1' 'GL000224.1' 'GL000219.1' 'GL000205.1' 'GL000215.1'\n",
      " 'GL000216.1' 'GL000199.1' 'GL000211.1' 'GL000213.1' 'GL000220.1'\n",
      " 'GL000218.1' 'GL000209.1' 'GL000221.1' 'GL000228.1' 'GL000191.1'\n",
      " 'GL000204.1' 'GL000233.1' 'GL000237.1' 'GL000230.1' 'GL000242.1'\n",
      " 'GL000243.1' 'GL000241.1' 'GL000236.1' 'GL000240.1' 'GL000196.1'\n",
      " 'GL000247.1' 'GL000201.1' 'GL000231.1' 'GL000229.1']\n",
      "number of unique chromosom: 61\n"
     ]
    }
   ],
   "source": [
    "#checke chromosome inputs in chr_pos_df\n",
    "unique_chromosom = chr_pos_df['X.Chr'].unique()\n",
    "print('unique chromosom:', unique_chromosom)\n",
    "\n",
    "#checke ID inputs in chr_pos_df -- there is no overlap in ID\n",
    "nunique_chromosom = chr_pos_df['X.Chr'].nunique()\n",
    "print('number of unique chromosom:', nunique_chromosom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19685"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make exp_case column names as list without FID and IID\n",
    "##this will be all ENSG transcripts numbers in our expression data\n",
    "trans = expression_case.columns[2:19688].tolist()\n",
    "len(trans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of columns in expression data: 19685\n",
      "number of columns not in expression data: 602\n"
     ]
    }
   ],
   "source": [
    "#make trans list as a dataframe\n",
    "fid = pd.DataFrame(trans, columns=[\"ID\"])\n",
    "print(\"total number of columns in expression data:\", fid.shape[0])\n",
    "\n",
    "#check how many ENSG are missing from FID that are not in ID\n",
    "mis_coord = fid[~fid['ID'].isin(chr_pos_df['ID'])]\n",
    "print(\"number of columns not in expression data:\", mis_coord.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19083"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop rows of chr_pos_df where chr_pos_df ID is not match with exp_case column name list\n",
    "coord = chr_pos_df[chr_pos_df['ID'].isin(trans)]\n",
    "\n",
    "#create pre_gene_list as list for swarm job later\n",
    "pre_gene_list = coord['ID'].tolist()\n",
    "len(pre_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of MT chromosom in coordinate data: 13\n",
      "Number of Y chromosom in coordinate data: 46\n",
      "Number of X chromosom in coordinate data: 783\n",
      "Number of GL chromosom in coordinate data: 0\n"
     ]
    }
   ],
   "source": [
    "#count how many MT are in coord\n",
    "mt_coord = coord[coord['X.Chr'].str.contains('MT')]\n",
    "print(\"Number of MT chromosom in coordinate data:\" , mt_coord['X.Chr'].count())\n",
    "\n",
    "#count how many X are in coord\n",
    "y_coord = coord[coord['X.Chr'].str.contains('Y')]\n",
    "print(\"Number of Y chromosom in coordinate data:\" , y_coord['X.Chr'].count())\n",
    "\n",
    "#count how many X are in coord\n",
    "x_coord = coord[coord['X.Chr'].str.contains('X')]\n",
    "print(\"Number of X chromosom in coordinate data:\" , x_coord['X.Chr'].count())\n",
    "\n",
    "#count how many X are in coord\n",
    "gl_coord = coord[coord['X.Chr'].str.contains('GL')]\n",
    "print(\"Number of GL chromosom in coordinate data:\" , gl_coord['X.Chr'].count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape of coord_df: (18241, 4)\n",
      "  X.Chr   start     end               ID\n",
      "0     1   69091   70008  ENSG00000186092\n",
      "1     1  860260  879955  ENSG00000187634\n",
      "2     1  879584  894689  ENSG00000188976\n",
      "3     1  895967  901095  ENSG00000187961\n",
      "4     1  901877  911245  ENSG00000187583\n",
      "number of missing genes: 602\n"
     ]
    }
   ],
   "source": [
    "#drop all the rows contain MT, X, Y, GL in X.Chr \n",
    "word_list = [\"MT\", \"X\", \"Y\", \"GL\"]\n",
    "coord_df = coord[~coord['X.Chr'].str.contains('|'.join(word_list))]\n",
    "coord_df = coord_df.reset_index(drop=True)\n",
    "print(\"shape of coord_df:\", coord_df.shape)\n",
    "print(coord_df.head())\n",
    "\n",
    "#check miscoordinate numbers are same 602\n",
    "no_coord = mis_coord['ID'].tolist()\n",
    "print(\"number of missing genes:\", len(no_coord))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes in gene_id: 18241\n",
      "number of genes in gene_list: 18241\n"
     ]
    }
   ],
   "source": [
    "#read gene_list from liftover coordinate data\n",
    "gene_id = coord_df['ID']\n",
    "print(\"number of genes in gene_id:\",len(gene_id))\n",
    "#create gene_list as list to drop columns in exp_case \n",
    "gene_list = coord_df['ID'].tolist()\n",
    "print(\"number of genes in gene_list:\", len(gene_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save coord_df as txt -- run only once for download\n",
    "coord_df.to_csv(r'./data_folder/twas_coordinate.txt', sep='\\t' ,index=False)\n",
    "\n",
    "#save gene_list as txt -- run only once for download\n",
    "gene_id.to_csv(r'./data_folder/gene_list.txt', sep='\\t' ,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000273088</th>\n",
       "      <th>ENSG00000273154</th>\n",
       "      <th>ENSG00000273155</th>\n",
       "      <th>ENSG00000273171</th>\n",
       "      <th>ENSG00000273173</th>\n",
       "      <th>ENSG00000273213</th>\n",
       "      <th>ENSG00000273217</th>\n",
       "      <th>ENSG00000273238</th>\n",
       "      <th>ENSG00000273274</th>\n",
       "      <th>ENSG00000273291</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD-PDAA503EF5</td>\n",
       "      <td>PD-PDAA503EF5</td>\n",
       "      <td>0.153490</td>\n",
       "      <td>-1.195026</td>\n",
       "      <td>-0.094739</td>\n",
       "      <td>0.278269</td>\n",
       "      <td>-1.606817</td>\n",
       "      <td>-1.231579</td>\n",
       "      <td>-0.471478</td>\n",
       "      <td>0.480406</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>0.519157</td>\n",
       "      <td>-0.078847</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>-0.265569</td>\n",
       "      <td>0.662116</td>\n",
       "      <td>-0.321552</td>\n",
       "      <td>-0.492056</td>\n",
       "      <td>-0.895260</td>\n",
       "      <td>0.836818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD-PDAB074CYQ</td>\n",
       "      <td>PD-PDAB074CYQ</td>\n",
       "      <td>0.741327</td>\n",
       "      <td>0.725518</td>\n",
       "      <td>1.231583</td>\n",
       "      <td>0.593704</td>\n",
       "      <td>-1.066773</td>\n",
       "      <td>0.544353</td>\n",
       "      <td>-1.111941</td>\n",
       "      <td>0.219180</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>-1.235525</td>\n",
       "      <td>-0.291322</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>0.481206</td>\n",
       "      <td>0.671036</td>\n",
       "      <td>-0.282876</td>\n",
       "      <td>-0.326137</td>\n",
       "      <td>-0.624609</td>\n",
       "      <td>-0.643441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD-PDAB549YWB</td>\n",
       "      <td>PD-PDAB549YWB</td>\n",
       "      <td>0.427937</td>\n",
       "      <td>-1.107550</td>\n",
       "      <td>0.040538</td>\n",
       "      <td>-1.836332</td>\n",
       "      <td>-0.428239</td>\n",
       "      <td>1.349854</td>\n",
       "      <td>-1.249725</td>\n",
       "      <td>-0.582143</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>0.628770</td>\n",
       "      <td>-0.542193</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>0.347083</td>\n",
       "      <td>-1.752264</td>\n",
       "      <td>-0.178388</td>\n",
       "      <td>-0.615766</td>\n",
       "      <td>0.517863</td>\n",
       "      <td>0.938400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD-PDAB762PA3</td>\n",
       "      <td>PD-PDAB762PA3</td>\n",
       "      <td>-0.632429</td>\n",
       "      <td>-1.268311</td>\n",
       "      <td>0.100997</td>\n",
       "      <td>0.499943</td>\n",
       "      <td>-0.195359</td>\n",
       "      <td>1.295541</td>\n",
       "      <td>0.593312</td>\n",
       "      <td>-1.838987</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>0.833768</td>\n",
       "      <td>0.860871</td>\n",
       "      <td>-0.499505</td>\n",
       "      <td>-0.738382</td>\n",
       "      <td>0.277839</td>\n",
       "      <td>-0.103466</td>\n",
       "      <td>-0.259155</td>\n",
       "      <td>-0.890719</td>\n",
       "      <td>-0.630759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD-PDAD926WRZ</td>\n",
       "      <td>PD-PDAD926WRZ</td>\n",
       "      <td>-1.023469</td>\n",
       "      <td>-2.544938</td>\n",
       "      <td>-0.100176</td>\n",
       "      <td>0.210245</td>\n",
       "      <td>0.317189</td>\n",
       "      <td>-0.789634</td>\n",
       "      <td>-0.752465</td>\n",
       "      <td>0.179003</td>\n",
       "      <td>...</td>\n",
       "      <td>0.037965</td>\n",
       "      <td>0.657849</td>\n",
       "      <td>0.597915</td>\n",
       "      <td>0.037965</td>\n",
       "      <td>0.003539</td>\n",
       "      <td>-0.053181</td>\n",
       "      <td>-0.095052</td>\n",
       "      <td>-0.571399</td>\n",
       "      <td>-1.044443</td>\n",
       "      <td>1.465210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>PP-4125</td>\n",
       "      <td>PP-4125</td>\n",
       "      <td>0.313107</td>\n",
       "      <td>0.175639</td>\n",
       "      <td>-0.246924</td>\n",
       "      <td>-0.490545</td>\n",
       "      <td>-2.050227</td>\n",
       "      <td>-0.381930</td>\n",
       "      <td>0.182633</td>\n",
       "      <td>0.045561</td>\n",
       "      <td>...</td>\n",
       "      <td>0.575435</td>\n",
       "      <td>-0.300371</td>\n",
       "      <td>-0.950092</td>\n",
       "      <td>0.575435</td>\n",
       "      <td>0.446657</td>\n",
       "      <td>-1.066448</td>\n",
       "      <td>-0.236436</td>\n",
       "      <td>-2.419127</td>\n",
       "      <td>-1.431688</td>\n",
       "      <td>-2.650735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1134</th>\n",
       "      <td>PP-4126</td>\n",
       "      <td>PP-4126</td>\n",
       "      <td>-0.509980</td>\n",
       "      <td>-1.396533</td>\n",
       "      <td>-0.740332</td>\n",
       "      <td>0.691257</td>\n",
       "      <td>0.945632</td>\n",
       "      <td>-1.188982</td>\n",
       "      <td>1.400997</td>\n",
       "      <td>1.025471</td>\n",
       "      <td>...</td>\n",
       "      <td>0.037965</td>\n",
       "      <td>1.070220</td>\n",
       "      <td>0.152859</td>\n",
       "      <td>0.037965</td>\n",
       "      <td>1.173718</td>\n",
       "      <td>-0.667721</td>\n",
       "      <td>-0.335580</td>\n",
       "      <td>0.467387</td>\n",
       "      <td>0.724382</td>\n",
       "      <td>0.816589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1135</th>\n",
       "      <td>PP-4127</td>\n",
       "      <td>PP-4127</td>\n",
       "      <td>-0.942894</td>\n",
       "      <td>1.153752</td>\n",
       "      <td>0.699754</td>\n",
       "      <td>0.230091</td>\n",
       "      <td>-0.138720</td>\n",
       "      <td>0.958663</td>\n",
       "      <td>-0.684057</td>\n",
       "      <td>-0.556005</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.036976</td>\n",
       "      <td>-0.830943</td>\n",
       "      <td>-0.450324</td>\n",
       "      <td>-1.036976</td>\n",
       "      <td>0.675392</td>\n",
       "      <td>0.377462</td>\n",
       "      <td>-0.249957</td>\n",
       "      <td>-0.336044</td>\n",
       "      <td>0.367930</td>\n",
       "      <td>-0.612752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1136</th>\n",
       "      <td>PP-4135</td>\n",
       "      <td>PP-4135</td>\n",
       "      <td>0.334858</td>\n",
       "      <td>0.571822</td>\n",
       "      <td>-0.503539</td>\n",
       "      <td>-2.049354</td>\n",
       "      <td>3.221209</td>\n",
       "      <td>-0.065807</td>\n",
       "      <td>5.463260</td>\n",
       "      <td>1.003958</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.036976</td>\n",
       "      <td>0.635543</td>\n",
       "      <td>-1.422952</td>\n",
       "      <td>-1.036976</td>\n",
       "      <td>0.783682</td>\n",
       "      <td>-0.779745</td>\n",
       "      <td>-0.131228</td>\n",
       "      <td>1.435043</td>\n",
       "      <td>0.387422</td>\n",
       "      <td>-0.793902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1137</th>\n",
       "      <td>PP-4136</td>\n",
       "      <td>PP-4136</td>\n",
       "      <td>0.123918</td>\n",
       "      <td>-0.215251</td>\n",
       "      <td>-1.094800</td>\n",
       "      <td>0.726608</td>\n",
       "      <td>0.102289</td>\n",
       "      <td>-2.085580</td>\n",
       "      <td>0.834945</td>\n",
       "      <td>0.653457</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.036976</td>\n",
       "      <td>1.924620</td>\n",
       "      <td>0.689365</td>\n",
       "      <td>-1.036976</td>\n",
       "      <td>0.242921</td>\n",
       "      <td>-1.435439</td>\n",
       "      <td>-0.087193</td>\n",
       "      <td>0.978051</td>\n",
       "      <td>-0.859678</td>\n",
       "      <td>-0.434820</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1138 rows × 18243 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                FID            IID  ENSG00000000419  ENSG00000000457  \\\n",
       "0     PD-PDAA503EF5  PD-PDAA503EF5         0.153490        -1.195026   \n",
       "1     PD-PDAB074CYQ  PD-PDAB074CYQ         0.741327         0.725518   \n",
       "2     PD-PDAB549YWB  PD-PDAB549YWB         0.427937        -1.107550   \n",
       "3     PD-PDAB762PA3  PD-PDAB762PA3        -0.632429        -1.268311   \n",
       "4     PD-PDAD926WRZ  PD-PDAD926WRZ        -1.023469        -2.544938   \n",
       "...             ...            ...              ...              ...   \n",
       "1133        PP-4125        PP-4125         0.313107         0.175639   \n",
       "1134        PP-4126        PP-4126        -0.509980        -1.396533   \n",
       "1135        PP-4127        PP-4127        -0.942894         1.153752   \n",
       "1136        PP-4135        PP-4135         0.334858         0.571822   \n",
       "1137        PP-4136        PP-4136         0.123918        -0.215251   \n",
       "\n",
       "      ENSG00000000460  ENSG00000000938  ENSG00000000971  ENSG00000001036  \\\n",
       "0           -0.094739         0.278269        -1.606817        -1.231579   \n",
       "1            1.231583         0.593704        -1.066773         0.544353   \n",
       "2            0.040538        -1.836332        -0.428239         1.349854   \n",
       "3            0.100997         0.499943        -0.195359         1.295541   \n",
       "4           -0.100176         0.210245         0.317189        -0.789634   \n",
       "...               ...              ...              ...              ...   \n",
       "1133        -0.246924        -0.490545        -2.050227        -0.381930   \n",
       "1134        -0.740332         0.691257         0.945632        -1.188982   \n",
       "1135         0.699754         0.230091        -0.138720         0.958663   \n",
       "1136        -0.503539        -2.049354         3.221209        -0.065807   \n",
       "1137        -1.094800         0.726608         0.102289        -2.085580   \n",
       "\n",
       "      ENSG00000001084  ENSG00000001167  ...  ENSG00000273088  ENSG00000273154  \\\n",
       "0           -0.471478         0.480406  ...        -0.499505         0.519157   \n",
       "1           -1.111941         0.219180  ...        -0.499505        -1.235525   \n",
       "2           -1.249725        -0.582143  ...        -0.499505         0.628770   \n",
       "3            0.593312        -1.838987  ...        -0.499505         0.833768   \n",
       "4           -0.752465         0.179003  ...         0.037965         0.657849   \n",
       "...               ...              ...  ...              ...              ...   \n",
       "1133         0.182633         0.045561  ...         0.575435        -0.300371   \n",
       "1134         1.400997         1.025471  ...         0.037965         1.070220   \n",
       "1135        -0.684057        -0.556005  ...        -1.036976        -0.830943   \n",
       "1136         5.463260         1.003958  ...        -1.036976         0.635543   \n",
       "1137         0.834945         0.653457  ...        -1.036976         1.924620   \n",
       "\n",
       "      ENSG00000273155  ENSG00000273171  ENSG00000273173  ENSG00000273213  \\\n",
       "0           -0.078847        -0.499505        -0.265569         0.662116   \n",
       "1           -0.291322        -0.499505         0.481206         0.671036   \n",
       "2           -0.542193        -0.499505         0.347083        -1.752264   \n",
       "3            0.860871        -0.499505        -0.738382         0.277839   \n",
       "4            0.597915         0.037965         0.003539        -0.053181   \n",
       "...               ...              ...              ...              ...   \n",
       "1133        -0.950092         0.575435         0.446657        -1.066448   \n",
       "1134         0.152859         0.037965         1.173718        -0.667721   \n",
       "1135        -0.450324        -1.036976         0.675392         0.377462   \n",
       "1136        -1.422952        -1.036976         0.783682        -0.779745   \n",
       "1137         0.689365        -1.036976         0.242921        -1.435439   \n",
       "\n",
       "      ENSG00000273217  ENSG00000273238  ENSG00000273274  ENSG00000273291  \n",
       "0           -0.321552        -0.492056        -0.895260         0.836818  \n",
       "1           -0.282876        -0.326137        -0.624609        -0.643441  \n",
       "2           -0.178388        -0.615766         0.517863         0.938400  \n",
       "3           -0.103466        -0.259155        -0.890719        -0.630759  \n",
       "4           -0.095052        -0.571399        -1.044443         1.465210  \n",
       "...               ...              ...              ...              ...  \n",
       "1133        -0.236436        -2.419127        -1.431688        -2.650735  \n",
       "1134        -0.335580         0.467387         0.724382         0.816589  \n",
       "1135        -0.249957        -0.336044         0.367930        -0.612752  \n",
       "1136        -0.131228         1.435043         0.387422        -0.793902  \n",
       "1137        -0.087193         0.978051        -0.859678        -0.434820  \n",
       "\n",
       "[1138 rows x 18243 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop columns if not in gene_list_1\n",
    "final_pheno_case = expression_case.drop(columns=[col for col in expression_case[expression_case.columns[2:19687]] if col not in gene_list])\n",
    "final_pheno_case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save expression matrix as txt --run only once\n",
    "final_pheno_case.to_csv(r'./data_folder/twas_expression_case.txt', sep= '\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plink files (genomic data), expression data(phenomic data), coordinate data, covariate data, and gene list are ready to use\n",
    "\n",
    "geno_path = f'{datadir}/qc_genotypes_twas'\n",
    "#plink files to use: {datadir}/qc_genotypes_twas_hg19_lifted\n",
    "gene_list_path = f'{datadir}/gene_list.txt'\n",
    "pheno_path = f'{datadir}/expression_matrix_final.txt'\n",
    "coord_path = f'{datadir}/twas_coordinate.txt'\n",
    "covar_path = f'{datadir}/covariates.txt'\n",
    "\n",
    "GWAS summery stat will be preprocessed after 1_liftover process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Variant_ID</th>\n",
       "      <th>Position</th>\n",
       "      <th>Basepair coordinate</th>\n",
       "      <th>Allele_1</th>\n",
       "      <th>Allele_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs145427775</td>\n",
       "      <td>0</td>\n",
       "      <td>10291</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>rs55998931</td>\n",
       "      <td>0</td>\n",
       "      <td>10492</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199896944</td>\n",
       "      <td>0</td>\n",
       "      <td>13504</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>rs199856693</td>\n",
       "      <td>0</td>\n",
       "      <td>14933</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>rs201855936</td>\n",
       "      <td>0</td>\n",
       "      <td>14948</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chr   Variant_ID  Position  Basepair coordinate Allele_1 Allele_2\n",
       "0    1  rs145427775         0                10291        T        C\n",
       "1    1   rs55998931         0                10492        T        C\n",
       "2    1  rs199896944         0                13504        A        G\n",
       "3    1  rs199856693         0                14933        A        G\n",
       "4    1  rs201855936         0                14948        A        G"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open bim file \n",
    "bim_df = pd.read_csv(f\"{datadir}/qc_genotypes_twas_hg19_lifted.bim\", sep='\\t', header=None)\n",
    "bim_df.columns = ['Chr', 'Variant_ID', 'Position', 'Basepair coordinate', 'Allele_1', 'Allele_2']\n",
    "bim_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Allele1</th>\n",
       "      <th>Allele2</th>\n",
       "      <th>Freq1</th>\n",
       "      <th>FreqSE</th>\n",
       "      <th>MinFreq</th>\n",
       "      <th>MaxFreq</th>\n",
       "      <th>Effect</th>\n",
       "      <th>StdErr</th>\n",
       "      <th>P-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>t</td>\n",
       "      <td>c</td>\n",
       "      <td>0.9905</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0.9882</td>\n",
       "      <td>0.9971</td>\n",
       "      <td>0.0238</td>\n",
       "      <td>0.0594</td>\n",
       "      <td>0.6889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9375</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.9221</td>\n",
       "      <td>0.9473</td>\n",
       "      <td>0.0185</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>0.3801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.2112</td>\n",
       "      <td>0.0070</td>\n",
       "      <td>0.1703</td>\n",
       "      <td>0.2316</td>\n",
       "      <td>-0.0054</td>\n",
       "      <td>0.0123</td>\n",
       "      <td>0.6598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>a</td>\n",
       "      <td>g</td>\n",
       "      <td>0.9963</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.9961</td>\n",
       "      <td>0.9984</td>\n",
       "      <td>-0.0228</td>\n",
       "      <td>0.1356</td>\n",
       "      <td>0.8663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:97895884</td>\n",
       "      <td>c</td>\n",
       "      <td>g</td>\n",
       "      <td>0.0587</td>\n",
       "      <td>0.0035</td>\n",
       "      <td>0.0361</td>\n",
       "      <td>0.0685</td>\n",
       "      <td>-0.0219</td>\n",
       "      <td>0.0217</td>\n",
       "      <td>0.3148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MarkerName Allele1 Allele2   Freq1  FreqSE  MinFreq  MaxFreq  Effect  \\\n",
       "0  chr11:88249377       t       c  0.9905  0.0026   0.9882   0.9971  0.0238   \n",
       "1   chr1:60320992       a       g  0.9375  0.0049   0.9221   0.9473  0.0185   \n",
       "2  chr8:135908647       a       g  0.2112  0.0070   0.1703   0.2316 -0.0054   \n",
       "3  chr16:77148858       a       g  0.9963  0.0005   0.9961   0.9984 -0.0228   \n",
       "4  chr11:97895884       c       g  0.0587  0.0035   0.0361   0.0685 -0.0219   \n",
       "\n",
       "   StdErr  P-value  \n",
       "0  0.0594   0.6889  \n",
       "1  0.0210   0.3801  \n",
       "2  0.0123   0.6598  \n",
       "3  0.1356   0.8663  \n",
       "4  0.0217   0.3148  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check sumstat.csv\n",
    "meta = pd.read_csv(\"sumstat.csv\", sep=' ')\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "data statistics (to write a paper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>AGE</th>\n",
       "      <th>MALE</th>\n",
       "      <th>EDUCATION</th>\n",
       "      <th>FAMILY_HISTORY</th>\n",
       "      <th>UPSIT</th>\n",
       "      <th>InfAJ</th>\n",
       "      <th>PC1_gen</th>\n",
       "      <th>PC2_gen</th>\n",
       "      <th>PC3_gen</th>\n",
       "      <th>PC4_gen</th>\n",
       "      <th>PC5_gen</th>\n",
       "      <th>PC6_gen</th>\n",
       "      <th>PC7_gen</th>\n",
       "      <th>PC8_gen</th>\n",
       "      <th>PC9_gen</th>\n",
       "      <th>PC10_gen</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD-PDKG339XM0</td>\n",
       "      <td>PD-PDKG339XM0</td>\n",
       "      <td>0.415094</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.410256</td>\n",
       "      <td>0</td>\n",
       "      <td>0.002876</td>\n",
       "      <td>-0.007443</td>\n",
       "      <td>-0.005085</td>\n",
       "      <td>-0.000187</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>-0.000787</td>\n",
       "      <td>-0.002445</td>\n",
       "      <td>-0.000240</td>\n",
       "      <td>-0.001125</td>\n",
       "      <td>0.000828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD-PDRM726FKG</td>\n",
       "      <td>PD-PDRM726FKG</td>\n",
       "      <td>0.547170</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.435897</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>-0.007768</td>\n",
       "      <td>-0.005870</td>\n",
       "      <td>-0.000552</td>\n",
       "      <td>0.001536</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>-0.000336</td>\n",
       "      <td>0.001626</td>\n",
       "      <td>0.001068</td>\n",
       "      <td>0.001081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD-PDEM487YZD</td>\n",
       "      <td>PD-PDEM487YZD</td>\n",
       "      <td>0.320755</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003569</td>\n",
       "      <td>-0.007966</td>\n",
       "      <td>-0.005733</td>\n",
       "      <td>-0.001055</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>-0.000491</td>\n",
       "      <td>0.001775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD-PDVW658HP7</td>\n",
       "      <td>PD-PDVW658HP7</td>\n",
       "      <td>0.679245</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.692308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.005395</td>\n",
       "      <td>-0.010209</td>\n",
       "      <td>-0.007062</td>\n",
       "      <td>-0.002003</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>0.001092</td>\n",
       "      <td>-0.000483</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>-0.001346</td>\n",
       "      <td>0.000829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD-PDVJ880MZU</td>\n",
       "      <td>PD-PDVJ880MZU</td>\n",
       "      <td>0.603774</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0</td>\n",
       "      <td>0.001514</td>\n",
       "      <td>-0.003853</td>\n",
       "      <td>-0.001543</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.000673</td>\n",
       "      <td>-0.000692</td>\n",
       "      <td>-0.000884</td>\n",
       "      <td>-0.000402</td>\n",
       "      <td>0.000285</td>\n",
       "      <td>0.000432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>PP-3089</td>\n",
       "      <td>PP-3089</td>\n",
       "      <td>0.415094</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.794872</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.011536</td>\n",
       "      <td>0.025762</td>\n",
       "      <td>0.021367</td>\n",
       "      <td>0.003645</td>\n",
       "      <td>0.001353</td>\n",
       "      <td>0.000578</td>\n",
       "      <td>0.002760</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>-0.004029</td>\n",
       "      <td>-0.000709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1134</th>\n",
       "      <td>PP-3288</td>\n",
       "      <td>PP-3288</td>\n",
       "      <td>0.245283</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.948718</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003412</td>\n",
       "      <td>-0.004851</td>\n",
       "      <td>-0.003367</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.001136</td>\n",
       "      <td>-0.000175</td>\n",
       "      <td>-0.000127</td>\n",
       "      <td>-0.000302</td>\n",
       "      <td>0.000876</td>\n",
       "      <td>0.003059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1135</th>\n",
       "      <td>PP-3290</td>\n",
       "      <td>PP-3290</td>\n",
       "      <td>0.547170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0</td>\n",
       "      <td>0.002941</td>\n",
       "      <td>-0.004904</td>\n",
       "      <td>-0.005817</td>\n",
       "      <td>-0.001479</td>\n",
       "      <td>0.000497</td>\n",
       "      <td>-0.000874</td>\n",
       "      <td>-0.004521</td>\n",
       "      <td>0.000362</td>\n",
       "      <td>0.000367</td>\n",
       "      <td>-0.005127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1136</th>\n",
       "      <td>PD-PDVN779VZN</td>\n",
       "      <td>PD-PDVN779VZN</td>\n",
       "      <td>0.283019</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.512821</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003096</td>\n",
       "      <td>-0.006337</td>\n",
       "      <td>-0.005133</td>\n",
       "      <td>-0.001075</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>-0.000924</td>\n",
       "      <td>-0.000375</td>\n",
       "      <td>0.000356</td>\n",
       "      <td>-0.000781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1137</th>\n",
       "      <td>PD-PDZT881XTK</td>\n",
       "      <td>PD-PDZT881XTK</td>\n",
       "      <td>0.377358</td>\n",
       "      <td>0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.512821</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.022171</td>\n",
       "      <td>0.062697</td>\n",
       "      <td>0.052898</td>\n",
       "      <td>0.016910</td>\n",
       "      <td>-0.000252</td>\n",
       "      <td>0.004395</td>\n",
       "      <td>0.014195</td>\n",
       "      <td>-0.004681</td>\n",
       "      <td>0.008428</td>\n",
       "      <td>0.002533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1138 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                FID            IID       AGE  MALE  EDUCATION  FAMILY_HISTORY  \\\n",
       "0     PD-PDKG339XM0  PD-PDKG339XM0  0.415094     0        1.0               1   \n",
       "1     PD-PDRM726FKG  PD-PDRM726FKG  0.547170     1        1.0               0   \n",
       "2     PD-PDEM487YZD  PD-PDEM487YZD  0.320755     1        0.5               0   \n",
       "3     PD-PDVW658HP7  PD-PDVW658HP7  0.679245     1        1.0               1   \n",
       "4     PD-PDVJ880MZU  PD-PDVJ880MZU  0.603774     1        1.0               0   \n",
       "...             ...            ...       ...   ...        ...             ...   \n",
       "1133        PP-3089        PP-3089  0.415094     0        1.0               0   \n",
       "1134        PP-3288        PP-3288  0.245283     0        1.0               0   \n",
       "1135        PP-3290        PP-3290  0.547170     1        0.5               0   \n",
       "1136  PD-PDVN779VZN  PD-PDVN779VZN  0.283019     0        0.5               0   \n",
       "1137  PD-PDZT881XTK  PD-PDZT881XTK  0.377358     0        0.5               0   \n",
       "\n",
       "         UPSIT  InfAJ   PC1_gen   PC2_gen   PC3_gen   PC4_gen   PC5_gen  \\\n",
       "0     0.410256      0  0.002876 -0.007443 -0.005085 -0.000187  0.000608   \n",
       "1     0.435897      0  0.003366 -0.007768 -0.005870 -0.000552  0.001536   \n",
       "2     0.692308      0  0.003569 -0.007966 -0.005733 -0.001055  0.000629   \n",
       "3     0.692308      0  0.005395 -0.010209 -0.007062 -0.002003  0.001324   \n",
       "4     0.307692      0  0.001514 -0.003853 -0.001543  0.000162  0.000673   \n",
       "...        ...    ...       ...       ...       ...       ...       ...   \n",
       "1133  0.794872      0 -0.011536  0.025762  0.021367  0.003645  0.001353   \n",
       "1134  0.948718      0  0.003412 -0.004851 -0.003367  0.000105  0.001136   \n",
       "1135  0.769231      0  0.002941 -0.004904 -0.005817 -0.001479  0.000497   \n",
       "1136  0.512821      0  0.003096 -0.006337 -0.005133 -0.001075  0.000205   \n",
       "1137  0.512821      1 -0.022171  0.062697  0.052898  0.016910 -0.000252   \n",
       "\n",
       "       PC6_gen   PC7_gen   PC8_gen   PC9_gen  PC10_gen  \n",
       "0    -0.000787 -0.002445 -0.000240 -0.001125  0.000828  \n",
       "1     0.000039 -0.000336  0.001626  0.001068  0.001081  \n",
       "2     0.000100  0.000309  0.000120 -0.000491  0.001775  \n",
       "3     0.001092 -0.000483  0.000103 -0.001346  0.000829  \n",
       "4    -0.000692 -0.000884 -0.000402  0.000285  0.000432  \n",
       "...        ...       ...       ...       ...       ...  \n",
       "1133  0.000578  0.002760  0.000935 -0.004029 -0.000709  \n",
       "1134 -0.000175 -0.000127 -0.000302  0.000876  0.003059  \n",
       "1135 -0.000874 -0.004521  0.000362  0.000367 -0.005127  \n",
       "1136  0.000142 -0.000924 -0.000375  0.000356 -0.000781  \n",
       "1137  0.004395  0.014195 -0.004681  0.008428  0.002533  \n",
       "\n",
       "[1138 rows x 18 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "covariates_case = pd.read_csv(f\"{datadir}/covariates_case.txt\", sep='\\t')\n",
    "covariates_cont = pd.read_csv(f\"{datadir}/covariates_control.txt\", sep='\\t')\n",
    "covariates_case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "% of MALE column in case (1=MALE, 0=FEMALE): \n",
      " 1    64.147627\n",
      "0    35.852373\n",
      "Name: MALE, dtype: float64\n",
      "% of MALE column in control (1=MALE, 0=FEMALE): \n",
      " 1    51.130435\n",
      "0    48.869565\n",
      "Name: MALE, dtype: float64\n",
      "% of FAMILY_HISTORY column in case (1=positive, 0=none): \n",
      " 0    75.13181\n",
      "1    24.86819\n",
      "Name: FAMILY_HISTORY, dtype: float64\n",
      "% of FAMILY_HISTORY column in control (1=positive, 0=none): \n",
      " 0    92.0\n",
      "1     8.0\n",
      "Name: FAMILY_HISTORY, dtype: float64\n",
      "% of InfAJ column in case (1=Ashkenazi, 0=no): \n",
      " 0    95.43058\n",
      "1     4.56942\n",
      "Name: InfAJ, dtype: float64\n",
      "% of InfAJ column in control (1=Ashkenazi, 0=no): \n",
      " 0    94.086957\n",
      "1     5.913043\n",
      "Name: InfAJ, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "#get percentage of MALE, FAMILY_HISTORY, and InfAJ\n",
    "\n",
    "print('% of MALE column in case (1=MALE, 0=FEMALE): \\n', covariates_case['MALE'].value_counts(normalize=True) * 100)\n",
    "print('% of MALE column in control (1=MALE, 0=FEMALE): \\n', covariates_cont['MALE'].value_counts(normalize=True) * 100)\n",
    "\n",
    "print('% of FAMILY_HISTORY column in case (1=positive, 0=none): \\n', covariates_case['FAMILY_HISTORY'].value_counts(normalize=True) * 100)\n",
    "print('% of FAMILY_HISTORY column in control (1=positive, 0=none): \\n', covariates_cont['FAMILY_HISTORY'].value_counts(normalize=True) * 100)\n",
    "\n",
    "print('% of InfAJ column in case (1=Ashkenazi, 0=no): \\n', covariates_case['InfAJ'].value_counts(normalize=True) * 100)\n",
    "print('% of InfAJ column in control (1=Ashkenazi, 0=no): \\n', covariates_cont['InfAJ'].value_counts(normalize=True) * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean of AGE column in case: \n",
      " 63.608084358523726\n",
      "Standard Diviation of AGE column in case: \n",
      " 9.429514734602233\n",
      "Mean of UPSIT column in case: \n",
      " 21.18541300527241\n",
      "Standard Diviation of UPSIT column in case: \n",
      " 8.022389396150745\n",
      "Mean of AGE column in control: \n",
      " 66.60577840112202\n",
      "Standard Diviation of AGE column in control: \n",
      " 9.213826966553222\n",
      "Mean of UPSIT column in control: \n",
      " 31.228555399719475\n",
      "Standard Diviation of UPSIT column in control: \n",
      " 6.759718262779698\n"
     ]
    }
   ],
   "source": [
    "#get inverse min-max scaling from covariate data and get mean and SD of AGE, UPSIT\n",
    "\n",
    "covariates_case[['AGE', 'UPSIT', 'EDUCATION']] = scaler.inverse_transform(covariates_case[['AGE', 'UPSIT', 'EDUCATION']])\n",
    "covariates_cont[['AGE', 'UPSIT', 'EDUCATION']] = scaler.inverse_transform(covariates_cont[['AGE', 'UPSIT', 'EDUCATION']])\n",
    "\n",
    "\n",
    "print('Mean of AGE column in case: \\n', covariates_case['AGE'].mean())\n",
    "print('Standard Diviation of AGE column in case: \\n', covariates_case['AGE'].std())\n",
    "print('Mean of UPSIT column in case: \\n', covariates_case['UPSIT'].mean())\n",
    "print('Standard Diviation of UPSIT column in case: \\n', covariates_case['UPSIT'].std())\n",
    "\n",
    "print('Mean of AGE column in control: \\n', covariates_cont['AGE'].mean())\n",
    "print('Standard Diviation of AGE column in control: \\n', covariates_cont['AGE'].std())\n",
    "print('Mean of UPSIT column in control: \\n', covariates_cont['UPSIT'].mean())\n",
    "print('Standard Diviation of UPSIT column in control: \\n', covariates_cont['UPSIT'].std())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python/3.8",
   "language": "python",
   "name": "py3.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
